---
title: "WBC depletion methods - analysis"
output:
  html_document:
    df_print: paged
    toc: true
    toc_float: true
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = here::here())
```

```{r load-libraries, include=FALSE}
# Since some of the required packages are only available in Bioconductor,
# renv needs to be initialised as follows:
# renv::init(bioconductor = "3.19")

library(BiocManager)
library(here)
library(ggplot2)
library(viridis)
library(DESeq2)
library(tximport)
library(txdbmaker)
library(vsn)
library(hexbin)
library(pheatmap)
library(dplyr)
library(tidyr)
library(readr)
library(purrr)
library(stringr)
library(GGally)
library(VennDiagram)
library(UpSetR)
library(scales)
# library(GenomicFeatures)

ggplot2::theme_set(theme_bw())
```

# Data import

If the RDS objects have already been created in a previous run, the following cell can be uncommented and run, and the remaining cells in the data import section can be skipped. See [RDS object section](#rds-objects)).

```{r data-import, message = FALSE}
# data_list <- list.files(path = here("results-refseq/r_objects"),
#                         pattern = ".RDS",
#                         recursive = TRUE, full.names = TRUE) %>%
#   set_names(., str_remove(basename(.), ".RDS")) %>%
#   purrr::imap(~ readRDS(file = .x ))
# 
# list2env(data_list, .GlobalEnv)
# rm(data_list)
```

## Read in fastq-screen and mapping statistics

```{r file-import-fastqscreen-mapping, message=FALSE}
# read in fastq screen files
fastq_screen <- list.files(
  path = here("results-refseq/fastq-screen"), pattern = "_screen.txt",
  recursive = TRUE, full.names = TRUE
) %>%
  read_delim(skip = 1, delim = "\t", n_max = 2, id = "read", show_col_types = FALSE) %>%
  rename("#Multiple_hits_multiple_genomes" = Multiple_hits_multiple_genomes) %>%
  mutate(read = str_remove(basename(read), "_screen.txt")) %>%
  mutate(sample = as.factor(str_remove(read, "_R[1,2].*"))) %>%
  mutate(Genome = as.factor(Genome))

# combine R1 and R2 stats per species and recalculate percentages
fastq_screen <- fastq_screen %>%
  select(!starts_with("%")) %>%
  group_by(sample, Genome) %>%
  summarise(across(.cols = `#Reads_processed`:`#Multiple_hits_multiple_genomes`, .fns = ~ sum(.x)), .groups = "drop_last") %>%
  mutate(across(.cols = `#Unmapped`:`#Multiple_hits_multiple_genomes`, .fns = ~ .x / `#Reads_processed`, .names = "%_{.col}")) %>%
  rename_with(~ str_replace(.x, "#", ""), .cols = starts_with("%"))

# read in samtools flagstat and metadata files
mapping_stats_df <- read_delim((here("results-refseq/mapping_stats.csv")), col_types = list(sample = "factor")) %>%
  mutate(primary_mapped_human = primary_mapped_all - primary_mapped_pk)

# merge dataframes
fastq_screen_mapping_stats_df <- mapping_stats_df %>%
  left_join(fastq_screen, by = "sample")

# Picard PCR duplicates
pcr_dups <- read_delim(here("results-refseq/multiqc/star_salmon/multiqc_report_data/multiqc_picard_dups.txt"), delim = "\t", col_types = list(Sample = "factor"))

# STAR mapping statistics
star_stats <- read_delim(here("results-refseq/multiqc/star_salmon/multiqc_report_data/mqc_star_alignment_plot_1.txt"), delim = "\t", col_types = list(Sample = "factor"))
```

## Read in `salmon` gene counts and tximport gene counts/TPM abundances

Import the following sets of count files:

- Gene-level counts (created in `nf-core/rnaseq` pipeline through `tximport` of Salmon transcript-level counts )
- Gene-level TPM (created in `nf-core/rnaseq` pipeline through `tximport` of Salmon transcript-level counts )
- Gene-level counts inside DESeq/SummarizedExperiment object (created by manually importing Salmon `quant.sf` files via `tximport`. Values should correspond exactly to the other gene-level counts). Used for visualisation.

Background info on the different types of counts and how to import them:

> `salmon` outputs two types of counts: transcript-level (`quant.sf`) and gene-level (`quant.genes.sf`). Both offer feature length, effective length, TPM, and number of reads. However, when importing counts with `tximport` for downstream differential expression analysis, transcript-level is preferred, because `tximport` can perform gene-level aggregation across all samples, instead of the per-sample aggregation that is done for the `quant.genes.sf` files. See: https://github.com/COMBINE-lab/salmon/issues/437.

> For `tximport`, there are several different ways of creating count matrices (i.e., using raw counts with length offset or by scaling the TPM abundances), but this should only be relevant for downstream differential expression testing) See the description of different options provided by the nf-core/rnaseq pipeline: https://nf-co.re/rnaseq/3.14.0/docs/output/#pseudoalignment and the `tximport` docs here: https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html.

> The Salmon transcript-level quant files can be imported using tximport and then converted into a DESeq object (~SummarizedExperiment). This approach provides the same count matrix as the tximport gene counts csv file, but it enables the DESeq2 approach for visualisation, namely transforming the counts using `vst` (variance stabilizing transformation) or `rlog` in order to remove the dependence of the variance on the mean. See [DESeq2 vignette](https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#data-transformations-and-visualization).

```{r file-import-counts, message=FALSE, cache=TRUE}
# read in sample metadata
samples <- read_csv(here("./data/samplesheet.csv"), 
                    col_select = c("sample", "condition", "RNA_type", "library_kit", "removal", "WBC_depletion", "kit"), 
                    col_types =cols(.default = "f")) %>% 
  mutate(RNA_type = str_to_title(RNA_type)) %>% 
  mutate(filter_name = factor(WBC_depletion, 
                         levels = c("nofilter", "centpip", "plasmo", "pmacs", "cellulose"), 
                         labels = c("no filter", "cent/pip", "plasmodipur", "PMACS", "cellulose"))) %>% 
  # mutate(kit = as.factor(str_replace_all(kit, "_", " "))) %>% 
  mutate(kit_name = factor(kit,
                      levels = c("mRNA_qiagen_FSglob", "mRNA_qiagen_FSglobH", "tRNA_qiagen_FSglobH", "mRNA_illumina"),
                      labels = c("mRNA Qiagen FS globin", "mRNA Qiagen FS globin/rRNA", "tRNA Qiagen FS globin/rRNA", "mRNA Illumina"))) %>% 
  mutate(sample_name = as.factor(str_replace(str_replace(sample, "104974-001-0", "Sample "), "_S1_L001", paste(" -", kit_name, "-", filter_name)))) %>% 
  mutate(sample_short = as.factor(str_replace(str_replace(sample, "104974-001-0", "Sample "), "_S1_L001", "")))

# read in transcript info - transcript id, gene id and genes names
transcript_info <- read.csv(here("results-refseq/star_salmon/tx2gene.tsv"),
  sep = "\t", header = FALSE,
  col.names = c("tx", "gene_id", "gene_name")
)
transcript_info <- list(
  transcript = transcript_info,
  gene = unique(transcript_info[, 2:3]),
  tx2gene = transcript_info[, 1:2]
)

# txi gene-level counts - same as using tximport and DESeqDataSetFromTximport except rounding needs to be done manually
# gene-level aggregation across all samples
gene_counts_txi <- read_delim(here("results-refseq/star_salmon/salmon.merged.gene_counts.tsv")) %>%
  mutate(across(starts_with("X"), \(x) as.integer(round(x)))) %>%
  rename_with(~ str_replace_all(.x, pattern = "\\.", replacement = "-"), .cols = `X104974.001.001_S1_L001`:`X104974.001.016_S1_L001`) %>%
  rename_with(~ str_sub(.x, 2, -1), .cols = `X104974-001-001_S1_L001`:`X104974-001-016_S1_L001`)

# txi gene-level TPM abundances
gene_tpm <- read_delim(here("results-refseq/star_salmon/salmon.merged.gene_tpm.tsv")) %>%
  rename_with(~ str_replace_all(.x, pattern = "\\.", replacement = "-"), .cols = `X104974.001.001_S1_L001`:`X104974.001.016_S1_L001`) %>%
  rename_with(~ str_sub(.x, 2, -1), .cols = `X104974-001-001_S1_L001`:`X104974-001-016_S1_L001`)

# DESeqDataSetFromTximport - original counts + offset tximport
# to be used with DESeq2's vst/rlog transform for visualisations
salmon_quant_files <- paste0("./results-refseq/star_salmon/", list.files(path = "./results-refseq/star_salmon/", pattern = "quant.sf", recursive = TRUE))
txi <- tximport(salmon_quant_files, type = "salmon", tx2gene = transcript_info$tx2gene, txOut = FALSE)
gse <- DESeqDataSetFromTximport(txi, colData = samples, design = ~1)

# clean up
rm(txi)
```

## Collect gene annotation

- Add species information to genes by gathering list of *P. knowlesi* genes from `gff` file.
- Add `gene_biotype` info for *P. knowlesi* genes (protein coding and r/t/nc/sn/snoRNA).
- Add globin and rRNA gene annotation for both species.

```{r gene-annotation, message=FALSE}
# Note: not all Pk genes start with the prefix PKNH, so it is safer to read genes from GFF instead
gff <- read_delim(here("data/ref-refseq-refseq/GCF_000006355.2_GCA_000006355.2_genomic.gff.gz"), comment = "#", col_names = c("seqname", "source", "feature", "start", "end", "score", "strand", "frame", "attribute")) %>%
  # only keep genes, since we use aggregated gene counts
  filter(feature=="gene") %>%
  mutate(id = row_number()) %>%
  separate_rows(attribute, sep = ";") %>%
  separate(attribute, c("attribute", "attribute_value"), sep = "=") %>%
  pivot_wider(names_from = attribute, values_from = attribute_value) %>%
  select(c(gene_biotype, ID)) %>%
  rename(gene_id = ID) %>%
  mutate(species = "pk")

# check to make sure there are no genes in gff that do not appear in transcript info file
stopifnot(all(gff$gene_id %in% transcript_info$gene$gene_id))

# create gene annotation df (based on one of the count matrices, not tx2gene
# because the latter contains more features, namely pseudogenes)
gene_annotation <- gene_counts_txi %>% 
  select(c(gene_id, gene_name)) %>% 
  left_join(gff, by = "gene_id") %>%
  mutate(species = case_when(is.na(species) ~ "human", .default = species))

# check to make sure all genes were assigned a species
stopifnot(!any(is.na(gene_annotation$species)))

# naive check to make sure "most" Pk genes are assigned the correct species
stopifnot((gene_annotation %>% filter(str_detect(gene_id, "PKNH")) %>% select(species) %>% unique()) == "pk")
```

#### Annotate rRNA and globin genes

##### rRNA {.tabset}

The RefSeq `gff` files contains info on `rRNA` in the `gene_biotype` attribute of gene entries (including mitochondrial rRNA) for both species (i.e., we collected featured labelled as `gene` in the 3rd column of the gff, not rRNA directly, since counts were aggregated on the gene level).

```
NC_011902.2	RefSeq	gene	825471	825594	.	+	.	ID=gene-PKNH_0117700;Dbxref=GeneID:7318405;Name=PKNH_0117700;end_range=825594,.;gbkey=Gene;gene_biotype=rRNA;locus_tag=PKNH_0117700;old_locus_tag=PKH_011712;partial=true;start_range=.,825471

NC_011902.2	RefSeq	rRNA	825471	825594	.	+	.	ID=rna-XR_002461722.1;Parent=gene-PKNH_0117700;Dbxref=GeneID:7318405,GenBank:XR_002461722.1;Name=XR_002461722.1;end_range=825594,.;gbkey=rRNA;locus_tag=PKNH_0117700;orig_transcript_id=gnl|WGS:CADCXE|mrna.PKNH_0117700-RA;partial=true;product=5.8S ribosomal RNA;start_range=.,825471;transcript_id=XR_002461722.1

NC_011902.2	RefSeq	exon	825471	825594	.	+	.	ID=exon-XR_002461722.1-1;Parent=rna-XR_002461722.1;Dbxref=GeneID:7318405,GenBank:XR_002461722.1;end_range=825594,.;gbkey=rRNA;locus_tag=PKNH_0117700;orig_transcript_id=gnl|WGS:CADCXE|mrna.PKNH_0117700-RA;partial=true;product=5.8S ribosomal RNA;start_range=.,825471;transcript_id=XR_002461722.1
```

The bash script in [`scripts/2_expression_analysis/0_gene_annotations-refseq.sh`](./scripts/2_expression_analysis/0_gene_annotations-refseq.sh) was used to retrieve the rRNA genes from both annotation files.

```{bash}
# zcat "${ref_human}" | awk '$3 ~ /gene/' | cut -f9 | grep -E "gene_biotype=rRNA" | grep -v "rRNA_pseudo" | grep -o "ID=[^;]*" | sed 's/ID=/human_rRNA;/' >> "${output_file}"
# zcat "${ref_pk}" | cut -f9 | grep -E "gene_biotype=rRNA" | grep -o "ID=[^;]*" | sed 's/ID=/pk_rRNA;/' >> "${output_file}"
```

```{r message=FALSE}
# read in gene type table
gene_types_df <- read_csv2(here("data/ref-refseq-refseq/gene_types.csv")) %>% left_join(transcript_info$gene, by = "gene_id")
```

**Note that the GENCODE and RefSeq annotation file contain different subsets of annotated rRNA genes!** See semi-comprehensive list of human rRNA genes here: https://www.genenames.org/data/genegroup/#!/group/848. Some are not included in any reference assembly, but others appear to be exclusive to either GENCODE or RefSeq. Some examples include RNA18S and 28S being present in RefSeq only. Moreover, for GENCODE, the majority of rRNA counts appear to be zero across the board.

See tables below for a full breakdown:

<!-- (Checked using `awk -F';' '$1 ~ /human_rRNA/ {print $2}' data/ref-refseq-refseq/gene_types.csv | while read -r line; do echo "$line"; grep "${line}" results-refseq/star_salmon/tx2gene.tsv; grep "$line" results-refseq/star_salmon/**/quant.gene*; done`) -->

```{r message=FALSE}
knitr::kable(
  gene_types_df %>%
    filter((gene_type == "human_rRNA") | (gene_type == "Mt_rRNA")) %>%
    left_join(gene_tpm, by = c("gene_id", "gene_name")) %>%
    filter(rowSums(across(.cols = `104974-001-001_S1_L001`:`104974-001-016_S1_L001`)) != 0),
  caption = "Non-zero counts of human rRNA genes annotated in RefSeq"
)

# check if GENCODE results are also present in directory
if (file.exists(here("data/ref/gene_types.csv"))) {
  transcript_info_gencode <- read.csv(here("results/star_salmon/tx2gene.tsv"),
    sep = "\t", header = FALSE,
    col.names = c("tx", "gene_id", "gene_name")
  )
  transcript_info_gencode <- list(
    transcript = transcript_info_gencode,
    gene = unique(transcript_info_gencode[, 2:3]),
    tx2gene = transcript_info_gencode[, 1:2]
  )
  gene_types_gencode_df <- read_csv2(here("data/ref/gene_types.csv")) %>%
    left_join(transcript_info_gencode$gene, by = "gene_id")

  gene_tpm_gencode <- read_delim(here("results/star_salmon/salmon.merged.gene_tpm.tsv"))

  knitr::kable(
    gene_types_gencode_df %>%
      filter((gene_type == "human_rRNA") | (gene_type == "Mt_rRNA")) %>%
      left_join(gene_tpm_gencode, by = c("gene_id", "gene_name")) %>%
      filter(rowSums(across(starts_with("X"))) != 0),
    caption = "Non-zero counts of human rRNA genes annotated in GENCODE"
  )
}
```

```{r message=FALSE, fig.height=18, out.width="100%"}
# gene_types_df %>%
#   left_join(gene_tpm, by = c("gene_id", "gene_name")) %>% 
#   filter(rowSums(across(starts_with("X"))) != 0) %>% 
#   pivot_longer(cols = `X104974.001.001_S1_L001`:`X104974.001.016_S1_L001`,
#                names_to = "sample", 
#                values_to = "counts") %>%
#   group_by(gene_id) %>%
#   mutate(total = sum(counts)) %>%
#   ungroup() %>% 
#   arrange(desc(total)) %>%
#   mutate(gene_name = factor(gene_name, levels = unique(gene_name))) %>%
#   ggplot(aes(x = gene_name, y = counts, fill = `sample`)) +
#     geom_bar(position = "stack", stat = "identity") +
#     scale_fill_viridis(option = "viridis", discrete = T) +
#     theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
#     coord_flip() +
#     scale_x_discrete(limits = rev) +
#     labs(fill = "Sample") +
#     ylab(label = "TPM") + xlab(label = "Gene name")

gene_types_df %>%
  left_join(gene_tpm, by = c("gene_id", "gene_name")) %>%
  filter(rowSums(across(.cols = `104974-001-001_S1_L001`:`104974-001-016_S1_L001`)) != 0) %>%
  pivot_longer(cols = `104974-001-001_S1_L001`:`104974-001-016_S1_L001`,
               names_to = "sample",
               values_to = "counts") %>%
  left_join(samples, by = "sample") %>%
  group_by(gene_id) %>%
  mutate(total = sum(counts)) %>%
  ungroup() %>%
  arrange(gene_type, desc(total)) %>% 
  mutate(gene_name = factor(gene_name, levels = unique(gene_name))) %>%
  ggplot(aes(x = gene_name, y = counts, fill = sample_short)) +
    geom_bar(position = "stack", stat = "identity") +
    scale_fill_viridis(option = "viridis", discrete = T) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    coord_flip() +
    scale_x_discrete(limits = rev) +
    labs(fill = "Sample", title = "RefSeq annotated rRNA - TPM") +
    ylab(label = "TPM") + xlab(label = "Gene name") +
    facet_wrap(~kit) +
    scale_y_continuous(labels = scales::label_number(scale_cut = cut_short_scale()))
```

```{r fig.height=18, fig.width=10}
gene_types_gencode_df %>%
  left_join(gene_tpm_gencode, by = c("gene_id", "gene_name")) %>%
  filter(rowSums(across(starts_with("X"))) != 0) %>%
  pivot_longer(cols = `X104974.001.001_S1_L001`:`X104974.001.016_S1_L001`,
               names_to = "sample",
               values_to = "counts") %>%
  mutate(sample = str_replace_all(sample, "\\.", replacement = "-")) %>%
  mutate(sample = str_sub(sample, 2, -1)) %>%
  left_join(samples, by = "sample") %>%
  group_by(gene_id) %>%
  mutate(total = sum(counts)) %>%
  ungroup() %>%
  arrange(gene_type, desc(total)) %>% 
  mutate(gene_name = factor(gene_name, levels = unique(gene_name))) %>%
  ggplot(aes(x = gene_name, y = counts, fill = sample_short)) +
    geom_bar(position = "stack", stat = "identity") +
    scale_fill_viridis(option = "viridis", discrete = T) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    coord_flip() +
    scale_x_discrete(limits = rev) +
    labs(fill = "Sample", title = "GENCODE annotated rRNA - TPM") +
    ylab(label = "TPM") + xlab(label = "Gene name") +
    facet_wrap(~kit) +
    scale_y_continuous(labels = scales::label_number(scale_cut = cut_short_scale()))
```

##### Globins

For human globins, the following genes were included: "HBB","HBD","HBBP1","HBG1","HBG2","HBE1","HBZ","HBM","HBA2","HBA1","HBQ1".

```{r globins, message=FALSE}
globin_names <- c("HBB", "HBD", "HBBP1", "HBG1", "HBG2", "HBE1", "HBZ", "HBM", "HBA2", "HBA1", "HBQ1")
globin_df <- tibble(gene_name = globin_names, gene_type = "human_globin") %>% left_join(transcript_info$gene, by = "gene_name")

gene_types_df <- gene_types_df %>% bind_rows(globin_df)

# export full gene type file
write_csv2(gene_types_df, here("data/ref-refseq-refseq/gene_types_R.csv"))

# some gene id share the same gene name
# gene_types_df %>%
#     group_by(gene_name) %>%
#     filter(n()>1)
```

Merge globin and rRNA annotations with gene annotation dataframe:

```{r combine-annotations-1}
gene_annotation <- gene_annotation %>% 
  left_join(gene_types_df, by = c("gene_id", "gene_name")) %>%
  mutate("gene_type" = coalesce(gene_type, species))
```

### Add annotations to counts

```{r combine-annotations-2}
# merge gene_type annotation and fill in missing types using species name
annotate_gene_type <- function(data, gene_annotation) {
  data %>%
    left_join(gene_annotation, by = c("gene_id", "gene_name"))
  }

gene_counts_txi <- annotate_gene_type(gene_counts_txi, gene_annotation)
gene_tpm <- annotate_gene_type(gene_tpm, gene_annotation)

# for SummarizedExperiment DESeq2 object, add annotations to rowData
rowData(gse) <- gene_annotation

# ensure genes are in the same order in annotation data and count files
# stopifnot(all(rownames(txi$abundance) == gene_annotation$gene_id))
stopifnot(all(rownames(gse) == gene_annotation$gene_id))
```

### Create gene subset filters for species and gene types

TODO: after count filtering

```{r}
pk_genes <- gene_annotation %>% filter(species == "pk") %>% select(gene_id)

```

### Subset DESeq object on Pk genes

```{r}
gse_pk <- gse[rowData(gse)$species == "pk",]

gene_annotation %>% filter(species == "pk") %>% select(gene_id)
```

## Filtering out low count genes

The raw counts can optionally be pre-filtered to remove rows that carry little to no information on gene expression levels. The [RNA-seq tutorial by Love et. al 2019](https://bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html#pre-filtering-the-dataset) and the [DESeq2 vignette](https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#pre-filtering) recommended a minimum count of 10 reads for a minimal number of samples (e.g., the size of the smallest group). Since we do not have replicates for the different treatment groups, we instead opt for a minimum of 10 reads for at least a single sample, as a conservative filtering approach (e.g., only genes for which none of the samples have a higher read count than 10 are removed).

```{r message=FALSE}
# create filter for all genes on the level of the DESeq/tximport gene counts
filter_all_genes <- rowSums(counts(gse) >= 10) >= 1
# should be the same as filtering on tximport raw gene counts from nf-core pipeline
stopifnot(all( filter_all_genes == (rowSums(gene_counts_txi[,3:18] >= 10) >= 1)))

# create filtered versions of full deseq2 objects
gse_filtered <- gse[filter_all_genes, ]
gse_pk_filtered <- gse_filtered[rowData(gse_filtered)$species == "pk",]

# create filtered versions of other count matrices
gene_tpm_filtered <- gene_tpm[filter_all_genes,]
gene_counts_txi_filtered <- gene_counts_txi[filter_all_genes,]

# check if filter was applied correctly, i.e. order of genes is the same in gse and tpm matrix
stopifnot(all.equal(gene_tpm_filtered, gene_tpm[(rowSums(gene_counts_txi[,3:18] >= 10) >= 1),]))
stopifnot(gene_tpm$gene_id == rownames(gse))
stopifnot(gene_tpm_filtered$gene_id == rownames(gse_filtered))
stopifnot(all.equal(gene_counts_txi_filtered, gene_counts_txi[(rowSums(gene_counts_txi[,3:18] >= 10) >= 1),]))
stopifnot(gene_counts_txi$gene_id == rownames(gse))
stopifnot(gene_counts_txi$gene_id == gene_tpm$gene_id)
stopifnot(gene_counts_txi_filtered$gene_id == rownames(gse_filtered))
```

The number of genes were filtered down from `r dim(gse)[1]`  to `r dim(gse_filtered)[1]` and from `r dim(gse_pk)[1]` to `r dim(gse_pk_filtered)[1]`, for all genes and *P. knowlesi* genes respectively. So in conclusion, the effect is relatively minor for *P. knowlesi*, but a substantial number of human genes were removed.

It is difficult to determine an appropriate threshold for filtering on the TPM level, so instead we filter the same genes as on the raw count level (>=10 for at least 1 sample).

```{r}
dim(gene_tpm)
mean.tpm <- apply(gene_tpm[, 3:18], 1, mean)
hist(log10(mean.tpm))
# hist(log10(mean.tpm+1e-5))

dim(gene_tpm_filtered)
mean.tpm <- apply(gene_tpm_filtered[, 3:18], 1, mean)
hist(log10(mean.tpm))
```

Lastly, we also perform filtering on protein coding Pk genes only:

```{r}
gse_pk_filtered_protein <- gse_pk_filtered[rowData(gse_pk_filtered)$gene_biotype == "protein_coding",]
vsd_pk_filtered_protein <- vst(gse_pk_filtered_protein, blind = TRUE)
```

## Transform dataframes to long format

```{r}
# cast to long format
to_long <- function(data) {
  data %>%
    pivot_longer(cols = `104974-001-001_S1_L001`:`104974-001-016_S1_L001`, names_to = "sample", values_to = "counts")
}

gene_counts_txi_long <- to_long(gene_counts_txi)
gene_counts_txi_filtered_long <- to_long(gene_counts_txi_filtered)

gene_tpm_long <- to_long(gene_tpm)
gene_tpm_filtered_long <- to_long(gene_tpm_filtered)
```

## Transform counts for visualisation

For visualisation purposes (e.g., PCA or heatmaps), gene counts should be transformed so that they all have the same range of variances, i.e. homoskedastic, instead of the variance depending on the mean. DESeq2 offers two transform options to stabilize the variance, `vst` and `rlog`. According to [Love et al., 2019](https://bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html#exploratory-analysis-and-visualization), rlog, while slower, works better for small datasets (n < 30) and when there are large differences in sequencing depth across samples. 

We use `blind=TRUE` to obtain a fully unsupervised transformation (independent of the design, although we are using an intercept-only model anyway), generally intended to observe batch effects. However, since we can have large differences in counts between samples due to our experimental design (e.g., zero counts for rRNA or globins in some samples), the transformation might not work as well as intended and shrink values too much ([DESeq2 vignette](https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#blind-dispersion-estimation)).

Also note that attempting to transform unfiltered counts can also lead to strange outcomes.

Lastly, the transformation should ideally be performed once on the entire set of gene counts (after filtering out low counts), and only then subset on different gene groups of interest (e.g., species, protein-coding genes only, etc.), rather than re-applying the transformation to a subset of counts.

```{r}
# variance stabilizing transformation without filtering
vsd <- vst(gse, blind = TRUE)
p <- meanSdPlot(assay(vsd))
p$gg + labs(title = "VST on unfiltered counts")

# extract transformed gene counts
# gene_counts_vst <- data.frame(assays(vsd)[[1]])
# colnames(gene_counts_vst) <- colData(gse)$sample
# gene_counts_vst <- gene_counts_vst %>%
#   mutate(gene_id = rownames(gene_counts_vst)) %>%
#   left_join(transcript_info$gene, by = "gene_id")

# repeat for filtered object
vsd_filtered <- vst(gse_filtered, blind = TRUE)
p <- meanSdPlot(assay(vsd_filtered))
p$gg + labs(title = "VST on filtered counts")

# compare with rlog transform
rld_filtered <- rlog(gse_filtered, blind=TRUE)
p <- meanSdPlot(assay(rld_filtered))
p$gg + labs(title = "Rlog on filtered counts")

# transform after subsetting - not recommended
vsd_pk <- vst(gse_pk, blind = TRUE)
p <- meanSdPlot(assay(vsd_pk))
p$gg + labs(title = "VST on unfiltered counts after subsetting Pk genes")

vsd_pk_filtered <- vst(gse_pk_filtered, blind = TRUE)
p <- meanSdPlot(assay(vsd_pk_filtered))
p$gg + labs(title = "VST on filtered counts after subsetting Pk genes")

rld_pk <- rlog(gse_pk, blind = TRUE)
p <- meanSdPlot(assay(rld_pk))
p$gg + labs(title = "Rlog on filtered counts after subsetting Pk genes")

# extract transformed gene counts
# gene_counts_vst_pk <- data.frame(assays(vsd_pk)[[1]])
# colnames(gene_counts_vst_pk) <- colData(vsd_pk)$sample
# gene_counts_vst_pk <- gene_counts_vst_pk %>%
#   mutate(gene_id = rownames(gene_counts_vst_pk)) %>%
#   left_join(transcript_info$gene, by = "gene_id")
```

## Export RDS objects {#rds-objects}

Save all output to rds object for easier loading.

```{r message = FALSE}
# dir.create(here("results-refseq/r_objects/"), showWarnings = FALSE)
# 
# purrr::iwalk(
#   tibble:::lst(
#     fastq_screen,
#     fastq_screen_mapping_stats_df,
#     gene_annotation,
#     gene_counts_salmon,
#     gene_counts_salmon_filtered,
#     gene_counts_txi,
#     gene_counts_txi_long,
#     gene_counts_txi_filtered,
#     gene_counts_txi_filtered_long,
#     gene_tpm,
#     gene_tpm_long,
#     gene_tpm_filtered,
#     gene_tpm_filtered_long,
#     gse,
#     gse_filtered,
#     gse_pk,
#     gse_pk_filtered,
#     gse_pk_filtered_protein,
#     mapping_stats_df,
#     samples,
#     vsd,
#     vsd_filtered,
#     vsd_pk,
#     vsd_pk_filtered,
#     vsd_pk_filtered_protein
#   ),
#   ~ saveRDS(.x, file = here(paste0("results-refseq/r_objects/", .y, ".RDS")))
# )
```

Data clean up

```{r}
# clean up
rm(gene_types_df)
rm(globin_df)
rm(gene_tpm_gencode)
rm(gene_types_gencode_df)
rm(transcript_info_gencode)
rm(transcript_info)
rm(gff)
```

# Analyses and figures

```{r show=FALSE, message = FALSE}
dir.create(here("results-refseq/figures/"), showWarnings = FALSE)
```

## Summary of annotated genes 

Number of genes per species:

```{r}
gene_counts_txi %>% group_by(species) %>% summarise(n=n()) %>% ungroup %>% mutate(total = sum(n), rel.freq = n / total)
```

Number of genes per species after filtering low counts (n < 10 for all samples):

```{r}
gene_counts_txi_filtered %>% group_by(species) %>% summarise(n=n()) %>% ungroup %>% mutate(total = sum(n), rel.freq = n / total)
```

Number of gene types per species:

```{r}
gene_counts_txi %>% group_by(species, gene_type) %>% summarise(n=n()) %>% ungroup %>% mutate(total = sum(n), rel.freq = round(n / total, 5))
```

Number of gene types per species after filtering out low counts (n < 10 for all samples):

```{r}
gene_counts_txi_filtered %>% group_by(species, gene_type) %>% summarise(n=n()) %>% ungroup %>% mutate(total = sum(n), rel.freq = round(n / total, 5))
```

```{r}
n_genes <- dim(gene_counts_txi)[1]
stopifnot(dim(gene_counts_txi)[1] == (gene_counts_salmon %>% group_by(sample) %>% summarise(n = n()) %>% select(n) %>% unique()))
n_genes_filtered <- dim(gse_filtered)[1]

n_human_genes <- gene_counts_txi %>% filter(species == "human") %>% count() %>% pull()
n_human_genes_filtered <- gene_counts_txi_filtered %>% filter(species == "human") %>% count() %>% pull()

n_pk_genes <- dim(gse_pk)[1]
stopifnot( dim(gse_pk)[1] == (gene_counts_txi %>% filter(species == "pk") %>% count()))
n_pk_genes_filtered <- dim(gse_pk_filtered)[1]
stopifnot( dim(gse_pk_filtered)[1] == (gene_counts_txi_filtered %>% filter(species == "pk") %>% count()))

# intersect(rownames(counts(gse_pk)), gff %>%
#   filter(gene_biotype == "protein_coding") %>%
#   pull(gene_id))

# n_pk_protein_coding <- dim(gse_pk)[1]

```

## Comparison of total RNA levels

> The direct comparison of RPKM and TPM across samples is meaningful only when there are equal total RNAs between compared samples and the distribution of RNA populations are close to each other. 

> Make sure both samples use the same RNA isolation approach [poly(A)+ selection versus ribosomal RNA depletion]. If not, they should not be compared.

> Check the fraction of the ribosomal, mitochondrial and globin RNAs, and the top highly expressed transcripts and see whether such RNAs constitute a very large part of the sequenced reads in a sample, and thus decrease the sequencing “real estate” available for the remaining genes in that sample. If the calculated fractions in two samples differ significantly, do not compare RPKM or TPM values directly.

> TPM should never be used for quantitative comparisons across samples when the total RNA contents and its distributions are very different. However, under appropriate circumstances, TPM can still be useful for qualitative comparison such as PCA and clustering analysis. 

(Zhao et al., 2020 - http://dx.doi.org/10.1261/rna.074922.120)

Looking at our samples, especially the ones prepared using the Illumina kit exhibit much higher total RNA content. Based on the output of Picard MarkDuplicates, the differences can at least partially be attributed to the number of PCR duplicates. The proportion of duplicates is comparable across kits, but they do appear to be consistently higher for some depletion methods (no filter > centpip > plasmo > pmacs > cellulose).

```{r}
gene_counts_txi_filtered %>% summarise(across(`104974-001-001_S1_L001`:`104974-001-016_S1_L001`, sum)) %>%
  to_long() %>% 
  left_join(samples, by = "sample") %>% 
  ggplot(aes(x = sample_short, y = counts, fill = kit_name)) + 
  geom_bar(stat = "identity") + 
  scale_fill_viridis(option = "viridis", discrete = T) +
                     # labels = c("mRNA Illumina", "mRNA Qiagen globin depletion", "mRNA Qiagen globin/rRNA depletion", "tRNA Qiagen globin/rRNA depletion")) + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_y_continuous(labels = scales::label_number(scale_cut = cut_short_scale())) +
  labs(x = "Sample", y = "Number of reads", fill = "Kit")
```

```{r}
pcr_dups %>% 
  left_join(samples, by = join_by(Sample == sample)) %>% 
  mutate(READS_IN_DUPLICATE_PAIRS = 2*READ_PAIR_DUPLICATES) %>% 
  mutate(READS_IN_UNIQUE_PAIRS = 2 * (READ_PAIRS_EXAMINED - READ_PAIR_DUPLICATES)) %>% 
  mutate(READS_IN_UNIQUE_UNPAIRED = UNPAIRED_READS_EXAMINED - UNPAIRED_READ_DUPLICATES) %>% 
  mutate(READS_IN_DUPLICATE_PAIRS_OPTICAL = 2 * READ_PAIR_OPTICAL_DUPLICATES) %>% 
  mutate(READS_IN_DUPLICATE_PAIRS_NONOPTICAL = READS_IN_DUPLICATE_PAIRS - READS_IN_DUPLICATE_PAIRS_OPTICAL) %>% 
  mutate(READS_IN_DUPLICATE_UNPAIRED = UNPAIRED_READ_DUPLICATES) %>% 
  mutate(READS_UNMAPPED = UNMAPPED_READS) %>% 
  select(c(sample_name, condition, kit, READS_IN_UNIQUE_PAIRS, READS_IN_UNIQUE_UNPAIRED, READS_IN_DUPLICATE_PAIRS_OPTICAL, READS_IN_DUPLICATE_PAIRS_NONOPTICAL, READS_IN_DUPLICATE_UNPAIRED)) %>% 
  pivot_longer(cols = 4:8, names_to = "type", values_to = "counts") %>% 
  ggplot(aes(x = sample_name, y = counts, fill = type)) +
      geom_bar(position = "stack", stat = "identity") +
      scale_fill_viridis(option = "viridis", discrete = T) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
      coord_flip() +
      scale_x_discrete(limits = rev) +
      scale_y_continuous(labels = scales::label_number(scale_cut = cut_short_scale())) +
      labs(x = "Sample", y = "Number of reads", fill = "Picard deduplication stats")
```

## Alignment statistics

The output reported by `samtool flagstats` and STAR's `Log.final.out` files cannot be compared directly: the latter reports counts of read-pairs as opposed to individual reads, and the former does not contain unmapped reads (these are not added to the `.bam` outputs by default) while also counting multi-mapped reads.

```{r}
star_stats
```

```{r}
star_stats  %>% 
  pivot_longer(cols = 2:6, names_to = "type", values_to = "counts") %>% 
  left_join(samples, by = join_by(Sample == sample)) %>% 
  ggplot(aes(x = sample_name, y = counts, fill = type)) +
      geom_bar(position = "stack", stat = "identity") +
      scale_fill_viridis(option = "viridis", discrete = T) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
      coord_flip() +
      scale_x_discrete(limits = rev) +
      scale_y_continuous(labels = scales::label_number(scale_cut = cut_short_scale())) +
      labs(x = "Sample", y = "Number of read pairs", fill = "Mapping stats")
```

```{r}
mapping_stats_df
```

```{r}
mapping_stats_df %>% 
  left_join(samples, by = "sample") %>% 
  pivot_longer(cols = primary_mapped_pk:primary_mapped_human, names_to = "type", values_to = "counts") %>% 
  ggplot(aes(x = sample_name, y = counts, fill = type)) +
      geom_bar(position = "stack", stat = "identity") +
      scale_fill_viridis(option = "viridis", discrete = T) + 
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
      coord_flip() +
      scale_x_discrete(limits = rev) +
      scale_y_continuous(labels = scales::label_number(scale_cut = cut_short_scale())) +
      labs(x = "Sample", y = "Number of primary reads", fill = "Samtools flagstat")
```

## Top expressed reads in TPM {.tabset}

TODO: for filtered data

```{r}
gene_tpm_long %>%
  group_by(sample) %>%
  top_n(25, counts) %>%
  arrange(sample, desc(counts), .by_group = T)
```

```{r,results='asis', echo=FALSE}
for (i in unique(gene_tpm_long$sample)) {
  cat("### Sample ", i, " \n")
  print(knitr::kable(
    gene_tpm_long %>%
      filter(sample == i) %>%
      top_n(25, counts) %>%
      arrange(sample, desc(counts), .by_group = T)
  ))
  cat("\n")
}
```

## FastQ Screen {.tabset}

Also see figure in [results-refseq/fastq-screen/](results-refseq/fastq-screen/) and [results-refseq/multiqc-fastq-screen-samtools-pk/multiqc_report.html](results-refseq/multiqc-fastq-screen-samtools-pk/multiqc_report.html).

### All reads for both species

```{r}
fastq_screen_transformed <- bind_rows(
  fastq_screen %>%
    mutate(`%_single_genome` = `%_One_hit_one_genome` + `%_Multiple_hits_one_genome`) %>%
    mutate(`%_multiple_genomes` = `%_One_hit_multiple_genomes` + `%_Multiple_hits_multiple_genomes`) %>%
    mutate(`No_hits` = `%_One_hit_multiple_genomes` + `%_Multiple_hits_multiple_genomes`) %>%
    select(c(sample, Genome, `#Reads_processed`, `%_single_genome`, `%_multiple_genomes`)) %>%
    pivot_wider(names_from = Genome, values_from = c(`%_single_genome`, `%_multiple_genomes`)) %>%
    select(!`%_multiple_genomes_PknowlesiH`) %>%
    rename(`%_multiple_genomes` = `%_multiple_genomes_Human`) %>%
    mutate(pct = 1 - `%_single_genome_PknowlesiH` - `%_single_genome_Human` - `%_multiple_genomes`) %>%
    select(sample, pct) %>%
    mutate("Mapped reads" = "No hits"),
  fastq_screen %>%
    mutate(pct = `%_One_hit_one_genome` + `%_Multiple_hits_one_genome`) %>%
    select(c(sample, Genome, pct)) %>%
    rename("Mapped reads" = Genome),
  fastq_screen %>%
    mutate(pct = `%_One_hit_multiple_genomes` + `%_Multiple_hits_multiple_genomes`) %>%
    select(c(sample, pct)) %>%
    distinct(sample, .keep_all = TRUE) %>%
    mutate("Mapped reads" = "Multiple genomes")
) %>%
  mutate(`Mapped reads` = factor(`Mapped reads`, levels = c("Human", "PknowlesiH", "Multiple genomes", "No hits"))) %>% 
  left_join(samples, by = "sample")

ggplot(fastq_screen_transformed, aes(x = sample_short, y = pct, fill = `Mapped reads`)) +
  geom_bar(position = "stack", stat = "identity") +
  scale_fill_viridis(option = "viridis", discrete = T) +
  coord_flip() +
  scale_x_discrete(limits = rev) +
  labs(fill = expression("FastQ Screen mapped reads"), x = "Sample") +
  ylab(label = "% of mapped reads")
```

### Human reads only

```{r}
# Human reads
fastq_screen %>% pivot_longer(
  cols = starts_with("%"),
  names_to = "fastq_screen_pct_type",
  values_to = "fastq_screen_pct_value") %>%
  filter(Genome == "Human") %>% 
  left_join(samples) %>% 
  ggplot(aes(x = sample_short, y = fastq_screen_pct_value, fill = fastq_screen_pct_type)) +
    geom_bar(position = "stack", stat = "identity") +
    scale_fill_viridis(option = "viridis", discrete = T) +
    coord_flip() +
    scale_x_discrete(limits = rev) +
    labs(fill = "FastQ screen statistics for human reads", x = "Sample", y = "% of mapped reads")
```

### Pk reads

```{r}
# Pk reads
fastq_screen %>% pivot_longer(
  cols = starts_with("%"),
  names_to = "fastq_screen_pct_type",
  values_to = "fastq_screen_pct_value"
) %>%
  filter(Genome == "PknowlesiH") %>% 
  left_join(samples) %>% 
    ggplot(aes(x = sample_short, y = fastq_screen_pct_value, fill = fastq_screen_pct_type)) +
    geom_bar(position = "stack", stat = "identity") +
    scale_fill_viridis(option = "viridis", discrete = T) +
    coord_flip() +
    scale_x_discrete(limits = rev) +
    labs(fill = expression("FastQ screen statistics for" ~ italic("P. knowlesi") ~ "reads"),  x = "Sample", y = "% of mapped reads")
```

## rRNA and globin levels

```{r}
# switched to gene_annotation df

gene_annotation %>%
  # select rRNA and globins
  filter(! gene_type %in% c("human","pk")) %>% 
  left_join(gene_tpm[1:18], by = c("gene_id", "gene_name")) %>%
  filter(rowSums(across(.cols = `104974-001-001_S1_L001`:`104974-001-016_S1_L001`)) != 0) %>%
  pivot_longer(cols = `104974-001-001_S1_L001`:`104974-001-016_S1_L001`,
               names_to = "sample",
               values_to = "counts") %>%
  left_join(samples, by = "sample") %>%
  group_by(gene_id) %>%
  mutate(total = sum(counts)) %>% 
  ungroup() %>% 
  arrange(species, gene_type, desc(total)) %>% 
  mutate(gene_name = factor(gene_name, levels = unique(gene_name))) %>%
  ggplot(aes(x = gene_name, y = counts, fill = sample_short)) +
    geom_bar(position = "stack", stat = "identity") +
    scale_fill_viridis(option = "viridis", discrete = T) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    coord_flip() +
    scale_x_discrete(limits = rev) +
    labs(fill = "Sample") +
    ylab(label = "TPM") + xlab(label = "Gene name") +
    facet_wrap(~kit) +
    scale_y_continuous(labels = scales::label_number(scale_cut = cut_short_scale()))
```

TODO: log TPM is often plotted instead of raw values, but can be misleading on barplot...

```{r}
gene_annotation %>%
  filter(! gene_type %in% c("human","pk")) %>% 
  left_join(gene_tpm[1:18], by = c("gene_id", "gene_name")) %>%
  filter(rowSums(across(.cols = `104974-001-001_S1_L001`:`104974-001-016_S1_L001`)) != 0) %>%
  pivot_longer(cols = `104974-001-001_S1_L001`:`104974-001-016_S1_L001`,
               names_to = "sample",
               values_to = "counts") %>%
  left_join(samples, by = "sample") %>%
  group_by(gene_id) %>%
  mutate(total = sum(counts)) %>% 
  ungroup() %>% 
  arrange(species, gene_type, desc(total)) %>% 
  mutate(gene_name = factor(gene_name, levels = unique(gene_name))) %>%
  ggplot(aes(x = gene_name, y = counts)) +
    geom_point(aes(colour = sample_name)) +
    scale_colour_viridis(option = "viridis", discrete = T) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    coord_flip() +
    scale_x_discrete(limits = rev) +
    # labs(colour = "Sample") +
    labs(x = "Gene name", y = "log(TPM+1)", colour = "Sample") +
    facet_wrap(~sample_name) +
    scale_y_continuous(trans = 'log1p')
```

Globin-only:

```{r}
gene_annotation %>%
  filter(gene_type == "human_globin") %>% 
  left_join(gene_tpm[1:18], by = c("gene_id", "gene_name")) %>%
  filter(rowSums(across(.cols = `104974-001-001_S1_L001`:`104974-001-016_S1_L001`)) != 0) %>%
  pivot_longer(cols = `104974-001-001_S1_L001`:`104974-001-016_S1_L001`,
               names_to = "sample",
               values_to = "counts") %>%
  left_join(samples, by = "sample") %>%
  group_by(gene_id) %>%
  mutate(total = sum(counts)) %>% 
  ungroup() %>% 
  arrange(species, gene_type, desc(total)) %>% 
  mutate(gene_name = factor(gene_name, levels = unique(gene_name))) %>%
  ggplot(aes(x = gene_name, y = counts)) +
    geom_point(aes(colour = sample)) +
    scale_colour_viridis(option = "viridis", discrete = T) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    coord_flip() +
    scale_x_discrete(limits = rev) +
    labs(x = "Gene name", y = "TPM", colour = "Sample") +
    facet_wrap(~sample_name) +
    scale_y_continuous(labels = scales::label_number(scale_cut = cut_short_scale()))
```

Human rRNA only:

```{r}
gene_annotation %>%
  filter(gene_type == "human_rRNA") %>% 
  left_join(gene_tpm[1:18], by = c("gene_id", "gene_name")) %>%
  filter(rowSums(across(.cols = `104974-001-001_S1_L001`:`104974-001-016_S1_L001`)) != 0) %>%
  pivot_longer(cols = `104974-001-001_S1_L001`:`104974-001-016_S1_L001`,
               names_to = "sample",
               values_to = "counts") %>%
  left_join(samples, by = "sample") %>%
  group_by(gene_id) %>%
  mutate(total = sum(counts)) %>% 
  ungroup() %>% 
  arrange(species, gene_type, desc(total)) %>% 
  mutate(gene_name = factor(gene_name, levels = unique(gene_name))) %>%
  ggplot(aes(x = gene_name, y = counts)) +
    geom_point(aes(colour = sample)) +
    scale_colour_viridis(option = "viridis", discrete = T) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    coord_flip() +
    scale_x_discrete(limits = rev) +
    labs(x = "Gene name", y = "TPM", colour = "Sample") +
    facet_wrap(~sample_name) +
    scale_y_continuous(trans = 'log1p')

```

## Plot percentage of reads for each gene type (rRNA/globin) and species {.tabset}

**TPM values appear to be the most appropriate metric to use.**

### Gene-level counts tximport

These gene counts were aggregated across all samples (tximport of transcript counts).

```{r}
gene_counts_txi_by_sample_gene_type <- gene_counts_txi_long %>%
  left_join(mapping_stats_df, by = "sample") %>%
  group_by(condition, gene_type) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads))

p <- ggplot(gene_counts_txi_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(position = "fill", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_txi_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(position = "dodge", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_txi_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()
```

### Gene-level counts tximport - filtered

These gene counts were aggregated across all samples (tximport of transcript counts) and filtered to only genes with a minimum count of 10 reads for at least 1 sample.

```{r}
gene_counts_txi_filtered_by_sample_gene_type <- gene_counts_txi_filtered_long %>%
  left_join(mapping_stats_df, by = "sample") %>%
  group_by(condition, gene_type) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads))

p <- ggplot(gene_counts_txi_filtered_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(position = "fill", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_txi_filtered_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(position = "dodge", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_txi_filtered_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()
```

### Gene-level TPM

```{r}
gene_tpm_by_sample_gene_type <- gene_tpm_long %>%
  left_join(mapping_stats_df, by = "sample") %>%
  group_by(condition, gene_type) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads))

p <- ggplot(gene_tpm_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(position = "fill", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_tpm_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(position = "dodge", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_tpm_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()
```

### Gene-level TPM - filtered

Filtered to only genes with a minimum count of 10 reads for at least 1 sample.

```{r}
gene_tpm_filtered_by_sample_gene_type <- gene_tpm_filtered_long %>%
  left_join(mapping_stats_df, by = "sample") %>%
  group_by(condition, gene_type) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads))

p <- ggplot(gene_tpm_filtered_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(position = "fill", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_tpm_filtered_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(position = "dodge", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_tpm_filtered_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()
```

## Plot percentage of reads for each species {.tabset}

### Gene-level counts tximport {.tabset}

These gene counts were aggregated across all samples (tximport of transcript counts).

#### Samples

```{r}
gene_counts_txi_by_samples_species <- gene_counts_txi_long %>%
  left_join(mapping_stats_df, by = "sample") %>%
  left_join(samples) %>%
  group_by(sample, species) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads))

p <- ggplot(gene_counts_txi_by_samples_species, aes(y = sum_reads, x = sample, fill = species))
p + geom_bar(position = "fill", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_txi_by_samples_species, aes(y = sum_reads, x = sample, fill = species))
p + geom_bar(position = "dodge", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_txi_by_samples_species, aes(y = sum_reads, x = sample, fill = species))
p + geom_bar(stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()
```

#### Conditions

```{r}
gene_counts_txi_by_samples_species <- gene_counts_txi_long %>%
  left_join(mapping_stats_df, by = "sample") %>%
  left_join(samples) %>%
  group_by(condition, species) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads))

p <- ggplot(gene_counts_txi_by_samples_species, aes(y = sum_reads, x = condition, fill = species))
p + geom_bar(position = "fill", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_txi_by_samples_species, aes(y = sum_reads, x = condition, fill = species))
p + geom_bar(position = "dodge", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_txi_by_samples_species, aes(y = sum_reads, x = condition, fill = species))
p + geom_bar(stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()
```

### Gene-level counts tximport - filtered {.tabset}

These gene counts were aggregated across all samples (tximport of transcript counts) and filtered to only genes with a minimum count of 10 reads for at least 1 sample.

#### Samples

```{r}
gene_counts_txi_filtered_by_samples_species <- gene_counts_txi_filtered_long %>%
  left_join(mapping_stats_df, by = "sample") %>%
  left_join(samples) %>%
  group_by(sample, species) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads))

p <- ggplot(gene_counts_txi_filtered_by_samples_species, aes(y = sum_reads, x = sample, fill = species))
p + geom_bar(position = "fill", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_txi_filtered_by_samples_species, aes(y = sum_reads, x = sample, fill = species))
p + geom_bar(position = "dodge", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_txi_filtered_by_samples_species, aes(y = sum_reads, x = sample, fill = species))
p + geom_bar(stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()
```

#### Conditions

```{r}
gene_counts_txi_filtered_by_samples_species <- gene_counts_txi_filtered_long %>%
  left_join(mapping_stats_df, by = "sample") %>%
  left_join(samples) %>%
  group_by(condition, species) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads))

p <- ggplot(gene_counts_txi_filtered_by_samples_species, aes(y = sum_reads, x = condition, fill = species))
p + geom_bar(position = "fill", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_txi_filtered_by_samples_species, aes(y = sum_reads, x = condition, fill = species))
p + geom_bar(position = "dodge", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_txi_filtered_by_samples_species, aes(y = sum_reads, x = condition, fill = species))
p + geom_bar(stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()
```

### Gene-level TPM {.tabset}

#### Samples

```{r}
gene_tpm_by_samples_species <- gene_tpm_long %>%
  left_join(mapping_stats_df, by = "sample") %>%
  left_join(samples) %>%
  group_by(sample, species) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads))

p <- ggplot(gene_tpm_by_samples_species, aes(y = sum_reads, x = sample, fill = species))
p + geom_bar(position = "fill", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_tpm_by_samples_species, aes(y = sum_reads, x = sample, fill = species))
p + geom_bar(position = "dodge", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_tpm_by_samples_species, aes(y = sum_reads, x = sample, fill = species))
p + geom_bar(stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()
```

#### Conditions

```{r}
gene_tpm_by_samples_species <- gene_tpm_long %>%
  left_join(mapping_stats_df, by = "sample") %>%
  left_join(samples) %>%
  group_by(condition, species) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads))

p <- ggplot(gene_tpm_by_samples_species, aes(y = sum_reads, x = condition, fill = species))
p + geom_bar(position = "fill", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_tpm_by_samples_species, aes(y = sum_reads, x = condition, fill = species))
p + geom_bar(position = "dodge", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_tpm_by_samples_species, aes(y = sum_reads, x = condition, fill = species))
p + geom_bar(stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()
```

## Pk specific gene type plots

Try to use https://nf-co.re/rnaseq/1.1/docs/output#featurecounts?

```
└─▶ featureCounts -t "transcript" -g "gene_biotype" -B -C -p -s 1 -T 6 -a data/ref-refseq-refseq/GCF_000006355.2_GCA_000006355.2_genomic.gtf -o results-refseq/featureCounts-pk.txt results-refseq/star_salmon/104974-001-001_S1_L001.markdup.sorted.pk.bam
```

## Gene recovery {.tabset}

<!-- ```{r} -->
<!-- gene_counts_vst_pk_filtered %>%  -->
<!--   mutate(across(.cols = starts_with("104974"), .fns = ~ as.integer(.x > 0))) -->

<!-- ``` -->

```{r}
# create binary matrices with presence/absence of genes
gene_tpm_filtered_binary <- gene_tpm_filtered %>%
  mutate(across(.cols =`104974-001-001_S1_L001`:`104974-001-016_S1_L001`, .fns = ~ .x > 0)) %>%
  select(!c(gene_name, gene_type, species))

gene_tpm_pk_filtered_binary <- gene_tpm_filtered %>%
  mutate(across(.cols = `104974-001-001_S1_L001`:`104974-001-016_S1_L001`, .fns = ~ .x > 0)) %>%
  filter(species == "pk") %>%
  select(!c(gene_name, gene_type, species))

gene_tpm_filtered_binary_list <- purrr::set_names(colnames(gene_tpm_filtered_binary %>% select(`104974-001-001_S1_L001`:`104974-001-016_S1_L001`))) %>%
  purrr::imap(~ gene_tpm_filtered_binary$gene_id[gene_tpm_filtered_binary[, .x, drop = T]])

gene_tpm_pk_filtered_binary_list <- purrr::set_names(colnames(gene_tpm_pk_filtered_binary %>% select(`104974-001-001_S1_L001`:`104974-001-016_S1_L001`))) %>%
  purrr::imap(~ gene_tpm_pk_filtered_binary$gene_id[gene_tpm_pk_filtered_binary[, .x, drop = T]])

comparison_centpip <- samples %>%
  filter(WBC_depletion == "centpip") %>%
  select(sample)

comparison_plasmo <- samples %>%
  filter(WBC_depletion == "plasmo") %>%
  select(sample)

comparison_pmacs <- samples %>%
  filter(WBC_depletion == "pmacs") %>%
  select(sample)

comparison_cellulose <- samples %>%
  filter(WBC_depletion == "cellulose") %>%
  select(sample)

comparison_nofilter <- samples %>%
  filter(WBC_depletion == "nofilter") %>%
  filter(!sample == "104974-001-006_S1_L001")

comparison_mrna_qiagen <- samples %>%
  filter(kit == "mRNA_qiagen_FSglob")

comparison_trna_qiagen <- samples %>%
  filter(kit == "tRNA_qiagen_FSglobH")

comparison_mrna_illumina <- samples %>%
  filter(kit == "mRNA_illumina")

draw_venn <- function(set_list, names, title, output) {
  VennDiagram::venn.diagram(
    x = set_list,
    category.names = names,
    # Circles
    lwd = 2,
    # lty = "blank",
    # col = palette.colors(3),
    # fill = palette.colors(3),
    # col = hcl.colors(3, palette = "viridis", alpha = 0.9),
    # fill = hcl.colors(3, palette = "viridis", alpha = 0.4),
    col = hcl.colors(3, palette = "Set 2", alpha = 0.9),
    fill = hcl.colors(3, palette = "Set 2", alpha = 0.4),

    # Numbers
    cex = .9,
    fontface = "bold",
    fontfamily = "sans",

    # Set names
    cat.cex = 0.9,
    cat.fontface = "bold",
    cat.default.pos = "outer",
    cat.pos = c(-20, 20, 135),
    cat.dist = c(0.055, 0.055, 0.11),
    cat.fontfamily = "sans",
    rotation = 1,

    # title
    main = title,
    main.cex = 1.1,
    main.fontface = "bold",
    main.fontfamily = "sans",

    # output features
    height = 1200,
    width = 1200,
    resolution = 300,
    compression = "lzw",
    filename = output,
    output = TRUE,
    
    disable.logging = TRUE
  )
}
```

### All genes - kits per filter

Gene counts were filtered on a minimum (raw) count of 10 for at least 1 sample before checking for presence/absence. 

```{r}
v <- draw_venn(
  gene_tpm_filtered_binary_list[comparison_centpip$sample],
  c("mRNA Qiagen FS globin", "tRNA Qiagen FS globin/rRNA", "mRNA Illumina"),
  "Centpip",
  NULL
  # here("results-refseq/figures/venn-centpip.png")
)
grid.newpage()
grid.draw(v)

v <- draw_venn(
  gene_tpm_filtered_binary_list[comparison_plasmo$sample],
    c("mRNA Qiagen FS globin", "tRNA Qiagen FS globin/rRNA", "mRNA Illumina"),
  "Plasmo",
  NULL
  # here("results-refseq/figures/venn-plasmo.png")
)
grid.newpage()
grid.draw(v)

v <- draw_venn(
  gene_tpm_filtered_binary_list[comparison_pmacs$sample],
    c("mRNA Qiagen FS globin", "tRNA Qiagen FS globin/rRNA", "mRNA Illumina"),
  "PMACS",
  NULL
  # here("results-refseq/figures/venn-pmacs.png")
)
grid.newpage()
grid.draw(v)

v <- draw_venn(
  gene_tpm_filtered_binary_list[comparison_cellulose$sample],
    c("mRNA Qiagen FS globin", "tRNA Qiagen FS globin/rRNA", "mRNA Illumina"),
  "Cellulose",
  NULL
  # here("results-refseq/figures/venn-cellulose.png")
)
grid.newpage()
grid.draw(v)

v <- draw_venn(
  gene_tpm_filtered_binary_list[comparison_nofilter$sample],
    c("mRNA Qiagen FS globin", "tRNA Qiagen FS globin/rRNA", "mRNA Illumina"),
  "Cellulose",
  NULL
  # here("results-refseq/figures/venn-nofilter.png")
)
grid.newpage()
grid.draw(v)
```

### *P. knowlesi* genes - kits per filter

Gene counts were filtered on a minimum (raw) count of 10 for at least 1 sample before checking for presence/absence. 

```{r}
v <- draw_venn(
  gene_tpm_pk_filtered_binary_list[comparison_centpip$sample],
    c("mRNA Qiagen FS globin", "tRNA Qiagen FS globin/rRNA", "mRNA Illumina"),
  "Centpip",
  NULL
  # here("results-refseq/figures/venn-centpip_pk.png")
)
grid.newpage()
grid.draw(v)

v <- draw_venn(
  gene_tpm_pk_filtered_binary_list[comparison_plasmo$sample],
    c("mRNA Qiagen FS globin", "tRNA Qiagen FS globin/rRNA", "mRNA Illumina"),
  "Plasmo",
  NULL
  # here("results-refseq/figures/venn-plasmo_pk.png")
)
grid.newpage()
grid.draw(v)

v <- draw_venn(
  gene_tpm_pk_filtered_binary_list[comparison_pmacs$sample],
    c("mRNA Qiagen FS globin", "tRNA Qiagen FS globin/rRNA", "mRNA Illumina"),
  "PMACS",
  NULL
  # here("results-refseq/figures/venn-pmacs_pk.png")
)
grid.newpage()
grid.draw(v)

v <- draw_venn(
  gene_tpm_pk_filtered_binary_list[comparison_cellulose$sample],
    c("mRNA Qiagen FS globin", "tRNA Qiagen FS globin/rRNA", "mRNA Illumina"),
  "Cellulose",
  NULL
  # here("results-refseq/figures/venn-cellulose_pk.png")
)
grid.newpage()
grid.draw(v)

v <- draw_venn(
  gene_tpm_pk_filtered_binary_list[comparison_nofilter$sample],
    c("mRNA Qiagen FS globin", "tRNA Qiagen FS globin/rRNA", "mRNA Illumina"),
  "No filter",
  NULL
  # here("results-refseq/figures/venn-nofilter_pk.png")
)
grid.newpage()
grid.draw(v)
```

### All genes - filters per kit

```{r}
# Use upset plots instead of five-way venn diagram (alternatively check out proportional area venn)

# mRNA Qiagen
UpSetR::upset(as.data.frame(gene_tpm_filtered_binary %>% 
                              select((comparison_mrna_qiagen %>% select(sample) %>% pull())) %>%
                              rename(all_of(c(
                                "no filter" = "104974-001-001_S1_L001",
                                "cent/pip" = "104974-001-002_S1_L001", 
                                "plasmodipur" = "104974-001-003_S1_L001", 
                                "PMACS" = "104974-001-004_S1_L001", 
                                "cellulose" = "104974-001-005_S1_L001"))) %>% 
                              mutate_all(as.integer)),
              keep.order = TRUE,
              sets.x.label = "Genes per filter",
              mainbar.y.label = "Intersections for mRNA Qiagen",
              order.by = "freq")

# tRNA Qiagen
UpSetR::upset(as.data.frame(gene_tpm_filtered_binary %>% 
                              select((comparison_trna_qiagen %>% select(sample) %>% pull())) %>%
                              rename(all_of(c(
                                "no filter" = "104974-001-007_S1_L001",
                                "cent/pip" = "104974-001-008_S1_L001", 
                                "plasmodipur" = "104974-001-009_S1_L001", 
                                "PMACS" = "104974-001-010_S1_L001", 
                                "cellulose" = "104974-001-011_S1_L001"))) %>% 
                              mutate_all(as.integer)),
              keep.order = TRUE,
              sets.x.label = "Genes per filter",
              mainbar.y.label = "Intersections for tRNA Qiagen",
              order.by = "freq")

# mRNA Illumina
UpSetR::upset(as.data.frame(gene_tpm_filtered_binary %>% 
                              select((comparison_mrna_illumina %>% select(sample) %>% pull())) %>%
                              rename(all_of(c(
                                "no filter" = "104974-001-012_S1_L001",
                                "cent/pip" = "104974-001-013_S1_L001", 
                                "plasmodipur" = "104974-001-014_S1_L001", 
                                "PMACS" = "104974-001-015_S1_L001", 
                                "cellulose" = "104974-001-016_S1_L001"))) %>% 
                              mutate_all(as.integer)),
              keep.order = TRUE,
              sets.x.label = "Genes per filter",
              mainbar.y.label = "Intersections for mRNA Illumina",
              order.by = "freq")
```

### Pk genes - filters per kit

```{r}
# mRNA Qiagen
UpSetR::upset(as.data.frame(gene_tpm_pk_filtered_binary %>% 
                              select((comparison_mrna_qiagen %>% select(sample) %>% pull())) %>%
                              rename(all_of(c(
                                "no filter" = "104974-001-001_S1_L001",
                                "cent/pip" = "104974-001-002_S1_L001", 
                                "plasmodipur" = "104974-001-003_S1_L001", 
                                "PMACS" = "104974-001-004_S1_L001", 
                                "cellulose" = "104974-001-005_S1_L001"))) %>% 
                              mutate_all(as.integer)),
              keep.order = TRUE,
              sets.x.label = "Genes per filter",
              mainbar.y.label = "Intersections for mRNA Qiagen",
              order.by = "freq")

# tRNA Qiagen
UpSetR::upset(as.data.frame(gene_tpm_pk_filtered_binary %>% 
                              select((comparison_trna_qiagen %>% select(sample) %>% pull())) %>%
                              rename(all_of(c(
                                "no filter" = "104974-001-007_S1_L001",
                                "cent/pip" = "104974-001-008_S1_L001", 
                                "plasmodipur" = "104974-001-009_S1_L001", 
                                "PMACS" = "104974-001-010_S1_L001", 
                                "cellulose" = "104974-001-011_S1_L001"))) %>% 
                              mutate_all(as.integer)),
              keep.order = TRUE,
              sets.x.label = "Genes per filter",
              mainbar.y.label = "Intersections for tRNA Qiagen",
              order.by = "freq")

# mRNA Illumina
UpSetR::upset(as.data.frame(gene_tpm_pk_filtered_binary %>% 
                              select((comparison_mrna_illumina %>% select(sample) %>% pull())) %>%
                              rename(all_of(c(
                                "no filter" = "104974-001-012_S1_L001",
                                "cent/pip" = "104974-001-013_S1_L001", 
                                "plasmodipur" = "104974-001-014_S1_L001", 
                                "PMACS" = "104974-001-015_S1_L001", 
                                "cellulose" = "104974-001-016_S1_L001"))) %>% 
                              mutate_all(as.integer)),
              keep.order = TRUE,
              sets.x.label = "Genes per filter",
              mainbar.y.label = "Intersections for mRNA Illumina",
              order.by = "freq")
```

## Sample-to-sample distances {.tabset}

### All genes

```{r}
sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- vsd$condition
colnames(sampleDistMatrix) <- NULL
pheatmap(sampleDistMatrix,
  clustering_distance_rows = sampleDists,
  clustering_distance_cols = sampleDists,
  color = viridis(
    n = 256, alpha = 1,
    begin = 0, end = 1, option = "viridis"
  )
)
```

### *P. knowlesi* only

```{r}
sampleDists <- dist(t(assay(vsd_pk)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- vsd$condition
colnames(sampleDistMatrix) <- NULL
pheatmap(sampleDistMatrix,
  clustering_distance_rows = sampleDists,
  clustering_distance_cols = sampleDists,
  color = viridis(
    n = 256, alpha = 1,
    begin = 0, end = 1, option = "viridis"
  )
)
```

## Coverage statistics

```{r}
mapping_stats_df
```

## Heatmap of count matrix {.tabset}

**TODO: try quantile breaks for better separation between values instead, see https://slowkow.com/notes/pheatmap-tutorial/** 

<!-- TODO: check if heatmaps should be made using transformed or raw counts -->

### All genes

```{r}
select <- order(rowMeans(counts(gse, normalized = FALSE)),
  decreasing = TRUE
)[1:20]
df <- as.data.frame(colData(gse)[, c("RNA_type", "removal", "WBC_depletion")])
pheatmap(assay(vsd)[select, ],
  cluster_rows = FALSE, show_rownames = FALSE,
  cluster_cols = FALSE, annotation_col = df
)
```

```{r}
df <- as.data.frame(colData(gse)[, c("kit", "WBC_depletion")])
pheatmap(assay(vsd)[select, ],
  cluster_rows = FALSE, show_rownames = FALSE,
  cluster_cols = FALSE, annotation_col = df
)
```

### *P. knowlesi* only

Without clustering:

```{r}
select <- order(rowMeans(counts(gse_pk, normalized = FALSE)),
  decreasing = TRUE
)[1:20]
df <- as.data.frame(colData(gse_pk)[, c("RNA_type", "removal", "WBC_depletion")])
pheatmap(assay(vsd_pk)[select, ],
  cluster_rows = FALSE, show_rownames = FALSE,
  cluster_cols = FALSE, annotation_col = df,
  color = viridis(10)
)
```

With clustering:

```{r}
select <- order(rowMeans(counts(gse_pk, normalized = FALSE)),
  decreasing = TRUE
)[1:20]
df <- as.data.frame(colData(gse_pk)[, c("RNA_type", "removal", "WBC_depletion")])
pheatmap(assay(vsd_pk)[select, ],
  cluster_rows = FALSE, show_rownames = FALSE,
  cluster_cols = TRUE, annotation_col = df,
  color = viridis(10)
)
#
#  ,
# annotation_colors = list(
#   RNA_type = palette.colors(n = length(unique(df$RNA_type)), palette = "Okabe-Ito"),
#   removal = palette.colors(n = length(unique(df$removal)), palette = "Okabe-Ito"),
#   WBC_depletion = palette.colors(n = length(unique(df$WBC_depletion)), palette = "Okabe-Ito")
# )
# )
```

```{r}
df <- as.data.frame(colData(gse_pk)[, c("kit", "WBC_depletion")])
pheatmap(assay(vsd_pk)[select, ],
  cluster_rows = FALSE, show_rownames = FALSE,
  cluster_cols = FALSE, annotation_col = df
)
```

<!-- TODO - subsetting should happen after transform (except low counts), but check for effects since differences in non-protein genes might be stark -->

Protein-coding only (not normalized):

```{r}
select <- order(rowMeans(counts(gse_pk[rowData(gse_pk)$gene_biotype == "protein_coding",], normalized = FALSE)),
                decreasing = TRUE
                )[1:20]
df <- as.data.frame(colData(gse_pk)[, c("RNA_type", "removal", "WBC_depletion")])

pheatmap(assay(vsd_pk)[select, ],
  cluster_rows = FALSE, show_rownames = FALSE,
  cluster_cols = FALSE, annotation_col = df,
  color = viridis(10)
)
```

```{r}
select <- order(rowMeans(counts(gse_pk_filtered_protein, normalized = FALSE)),
                decreasing = TRUE
                )[1:20]
df <- as.data.frame(colData(gse_pk_filtered_protein)[, c("RNA_type", "removal", "WBC_depletion")])

pheatmap(assay(vsd_pk_filtered_protein)[select, ],
  cluster_rows = FALSE, show_rownames = FALSE,
  cluster_cols = FALSE, annotation_col = df,
  color = viridis(10)
)
```

Protein-coding only + filtered on low counts:

```{r}
select <- order(rowMeans(counts(gse_pk_filtered[rowData(gse_pk_filtered)$gene_biotype == "protein_coding",], normalized = FALSE)),
  decreasing = TRUE
)[1:20]
df <- as.data.frame(colData(gse_pk)[, c("RNA_type", "removal", "WBC_depletion")])

pheatmap(assay(vsd_pk_filtered)[select, ],
  cluster_rows = FALSE, show_rownames = FALSE,
  cluster_cols = FALSE, annotation_col = df,
  color = viridis(10)
)
```

## PCA plots {.tabset}

TODO check impact of filtering transformed data vs filtering prior to transformation

All PCA plots use the top 500 most variable genes, unless otherwise stated.

### VSD - all genes

```{r}
pcaData <- plotPCA(vsd, intgroup = c("kit", "WBC_depletion"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit, shape = WBC_depletion)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(title = expression(paste("PCA of VST counts of human and ", italic("P. knowlesi"), " genes")),
       color = "Library preparation kit", shape = "Depletion method") +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

### VSD - all genes - filtered

Filtered on minimum count of 10 reads for at least 1 sample.

```{r}
pcaData <- plotPCA(vsd_filtered, intgroup = c("kit", "WBC_depletion"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit, shape = WBC_depletion)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(title = expression(paste("PCA of VST counts of human and ", italic("P. knowlesi"), " genes")),
       color = "Library preparation kit", shape = "Depletion method") +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

### VSD - Pk genes

```{r}
pcaData <- plotPCA(vsd_pk, intgroup = c("kit", "WBC_depletion"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit, shape = WBC_depletion)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(title = expression(paste("PCA of VST counts of ", italic("P. knowlesi"), " genes")),
       color = "Library preparation kit", shape = "Depletion method") +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

### VSD - Pk genes - filtered

Filtered on minimum count of 10 reads for at least 1 sample.

```{r}
pcaData <- plotPCA(vsd_pk_filtered, intgroup = c("kit", "WBC_depletion"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit, shape = WBC_depletion)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(title = expression(paste("PCA of VST counts of ", italic("P. knowlesi"), " genes")),
       color = "Library preparation kit", shape = "Depletion method") +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

### VSD - Pk genes - filtered - protein coding only

```{r}
protein_gene_filter <- intersect(rownames(rowData(vsd_pk_filtered)), gene_annotation %>% filter(gene_biotype == "protein_coding" & species == "pk") %>% select(gene_id) %>% pull())

# protein_gene_filter <- gene_annotation %>% filter(gene_biotype == "protein_coding" & species == "pk") %>% select(gene_id) %>% pull()
# gene annotation would need to be filtered on low counts too...

# this works too:
# vsd_pk_filtered[rowData(vsd_pk_filtered)$gene_biotype == "protein_coding"]

pcaData <- plotPCA(vsd_pk_filtered[protein_gene_filter,], intgroup = c("kit", "WBC_depletion"), returnData = TRUE, ntop=500)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit, shape = WBC_depletion)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(title = expression(paste("PCA of VST counts of ", italic("P. knowlesi"), " protein coding genes")),
       color = "Library preparation kit", shape = "Depletion method") +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

### VSD - Pk genes - filtered - protein coding only - top 1000 genes

```{r}
protein_gene_filter <- intersect(rownames(rowData(vsd_pk_filtered)), gene_annotation %>% filter(gene_biotype == "protein_coding" & species == "pk") %>% select(gene_id) %>% pull())

pcaData <- plotPCA(vsd_pk_filtered[protein_gene_filter,], intgroup = c("kit", "WBC_depletion"), returnData = TRUE, ntop=1000)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit, shape = WBC_depletion)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(title = expression(paste("PCA of VST counts of ", italic("P. knowlesi"), " protein coding genes")),
       color = "Library preparation kit", shape = "Depletion method") +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

### VSD - Pk genes - filtered - protein coding only - top all genes

```{r}
protein_gene_filter <- intersect(rownames(rowData(vsd_pk_filtered)), gene_annotation %>% filter(gene_biotype == "protein_coding" & species == "pk") %>% select(gene_id) %>% pull())

pcaData <- plotPCA(vsd_pk_filtered[protein_gene_filter,], intgroup = c("kit", "WBC_depletion"), returnData = TRUE, ntop=5098)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit, shape = WBC_depletion)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(title = expression(paste("PCA of VST counts of ", italic("P. knowlesi"), " protein coding genes")),
       color = "Library preparation kit", shape = "Depletion method") +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

<!-- TODO: try vst on pre-filtered (pk genes + low counts + protein coding) directly -->




```{r}


meanSdPlot(assay(vsd_prefiltered))


pcaData <- plotPCA(vsd_prefiltered, intgroup = c("kit", "WBC_depletion"), returnData = TRUE, ntop=5098)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit, shape = WBC_depletion)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(title = expression(paste("PCA of VST counts of ", italic("P. knowlesi"), " protein coding genes")),
       color = "Library preparation kit", shape = "Depletion method") +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

### Transformations and filtering considerations 

Quick check to see if applying VST to pre-filtered set of counts differs from filtering a pre-transformed list of counts. It appears not to for the Pk genes and subsetting on protein coding genes only, but it does for bigger filters. 

```{r}
vsd_prefiltered <- vst(gse_pk_filtered[protein_gene_filter])
vsd_prefiltered
vsd_pk_filtered[protein_gene_filter]
vsd[protein_gene_filter]


rowData(vsd_prefiltered)
rowData(vsd_pk_filtered[protein_gene_filter])
rowData(vsd[protein_gene_filter])

all.equal(vsd_prefiltered,vsd_pk_filtered[protein_gene_filter])

all.equal(vsd_prefiltered,rowData(vsd[protein_gene_filter]))
```

---

Transform after filtering on protein coding


```{r}
vsd_test <- varianceStabilizingTransformation(gse_pk_filtered[(rowData(gse_pk_filtered)$gene_biotype == "protein_coding")], blind = TRUE)

pcaData <- plotPCA(vsd_test, intgroup = c("kit", "WBC_depletion"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit, shape = WBC_depletion)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(title = expression(paste("PCA of VST counts of human and ", italic("P. knowlesi"), " genes")),
       color = "Library preparation kit", shape = "Depletion method") +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

Transform before filtering on protein coding

```{r}
pcaData <- plotPCA(vsd_pk_filtered[rowData(vsd_pk_filtered)$gene_biotype == "protein_coding"], intgroup = c("kit", "WBC_depletion"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit, shape = WBC_depletion)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(title = expression(paste("PCA of VST counts of human and ", italic("P. knowlesi"), " genes")),
       color = "Library preparation kit", shape = "Depletion method") +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

Transform after filtering on non-protein coding

```{r}
# vsd_test <- vst(gse_pk_filtered[(rowData(gse_pk_filtered)$gene_biotype != "protein_coding")], blind = TRUE)

vsd_test <- varianceStabilizingTransformation(gse_pk_filtered[(rowData(gse_pk_filtered)$gene_biotype != "protein_coding")], blind = TRUE)

pcaData <- plotPCA(vsd_test, intgroup = c("kit", "WBC_depletion"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit, shape = WBC_depletion)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(title = expression(paste("PCA of VST counts of human and ", italic("P. knowlesi"), " genes")),
       color = "Library preparation kit", shape = "Depletion method") +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

Transform before filtering on non-protein coding

```{r}
pcaData <- plotPCA(vsd_pk_filtered[rowData(vsd_pk_filtered)$gene_biotype != "protein_coding"], intgroup = c("kit", "WBC_depletion"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit, shape = WBC_depletion)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(title = expression(paste("PCA of VST counts of human and ", italic("P. knowlesi"), " genes")),
       color = "Library preparation kit", shape = "Depletion method") +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

## Correlation plots

### TPM {.tabset}

#### All genes

```{r message=FALSE, cache=TRUE}
# subset of samples for quick visualisation tests
# co_genes <- ggpairs(gene_tpm %>% filter(species == "pk"), columns = 3:5, lower = list(
#   continuous = "smooth"
# ), upper = list(continuous = wrap("cor", size = 2)))
# co_genes + theme_bw() + scale_x_continuous(trans = "log1p") + scale_y_continuous(trans = "log1p")


# all samples
co_genes <- ggpairs(gene_tpm %>% filter(species == "pk"), columns = 3:18, lower = list(
  continuous = "smooth"
), upper = list(continuous = wrap("cor", size = 2)))
co_genes + theme_bw() + scale_x_continuous(trans = "log1p") + scale_y_continuous(trans = "log1p")

# ggsave(here("results-refseq/figures/correlation_all_genes.png"))
```

#### Protein-coding genes

```{r message=FALSE}
co_genes <- ggpairs(
  gene_tpm %>%
    filter(species == "pk") %>%
    left_join(gff, by = "gene_id") %>%
    filter(gene_biotype == "protein_coding"),
  columns = 3:18, lower = list(continuous = "smooth"),
  upper = list(continuous = wrap("cor", size = 2))
)
co_genes + theme_bw() + scale_x_continuous(trans = "log1p") + scale_y_continuous(trans = "log1p")

# ggsave(here("results-refseq/figures/correlation_protein_coding_genes.png"))
```

<!-- #### Pk VST genes -->

<!-- ```{r message=FALSE} -->
<!-- co_genes <- ggpairs( -->
<!--   gene_counts_vst_pk %>% -->
<!--     filter(species == "pk") %>% -->
<!--     left_join(gff, by = "gene_id") %>% -->
<!--     filter(gene_biotype == "protein_coding_gene"), -->
<!--   columns = 1:16, lower = list(continuous = "smooth"), -->
<!--   upper = list(continuous = wrap("cor", size = 2)) -->
<!-- ) -->
<!-- co_genes + theme_bw() + scale_x_continuous(trans = "log1p") + scale_y_continuous(trans = "log1p") -->

<!-- ggsave(here("results-refseq/figures/correlation_protein_coding_genes_pk.png")) -->
<!-- ``` -->


----

Number of Pk genes in transcript_info file:

```{r}
dim(transcript_info$gene %>% filter(grepl("PKNH", gene_id)))
```

Number of Pk genes in count matrices:

```{r}
gene_tpm %>% filter(species == "pk")
```

Number of Pk genes in gff:

```{r}
read_delim(here("data/ref-refseq-refseq/GCF_000006355.2_GCA_000006355.2_genomic.gff"), comment = "#", col_names = c("seqname", "source", "feature", "start", "end", "score", "strand", "frame", "attribute")) %>%
  mutate(id = row_number()) %>%
  separate_rows(attribute, sep = ";") %>%
  separate(attribute, c("attribute", "attribute_value"), sep = "=") %>%
  pivot_wider(names_from = attribute, values_from = attribute_value) %>%
  filter(!is.na(gene_biotype))
```

```{r}
read_delim(here("data/ref-refseq-refseq/GCF_000006355.2_GCA_000006355.2_genomic.gff"), comment = "#", col_names = c("seqname", "source", "feature", "start", "end", "score", "strand", "frame", "attribute")) %>%
    mutate(id = row_number()) %>%
    separate_rows(attribute, sep = ";") %>%
    separate(attribute, c("attribute", "attribute_value"), sep = "=") %>%
    pivot_wider(names_from = attribute, values_from = attribute_value) %>% filter(feature == "gene")
```


```{r}
read_delim(here("data/ref-refseq-refseq/GCF_000006355.2_GCA_000006355.2_genomic.gff"), comment = "#", col_names = c("seqname", "source", "feature", "start", "end", "score", "strand", "frame", "attribute")) %>%
    mutate(id = row_number()) %>%
    separate_rows(attribute, sep = ";") %>%
    separate(attribute, c("attribute", "attribute_value"), sep = "=") %>%
    pivot_wider(names_from = attribute, values_from = attribute_value) %>% filter(gene_biotype == "protein_coding")
```

Non-protein coding genes:

```{r}
read_delim(here("data/ref-refseq-refseq/GCF_000006355.2_GCA_000006355.2_genomic.gff"), comment = "#", col_names = c("seqname", "source", "feature", "start", "end", "score", "strand", "frame", "attribute")) %>%
    mutate(id = row_number()) %>%
    separate_rows(attribute, sep = ";") %>%
    separate(attribute, c("attribute", "attribute_value"), sep = "=") %>%
    pivot_wider(names_from = attribute, values_from = attribute_value) %>% filter(!gene_biotype == "protein_coding")
```

```{r}
read_delim(here("data/ref-refseq-refseq/GCF_000006355.2_GCA_000006355.2_genomic.gff"), comment = "#", col_names = c("seqname", "source", "feature", "start", "end", "score", "strand", "frame", "attribute")) %>%
    mutate(id = row_number()) %>%
    separate_rows(attribute, sep = ";") %>%
    separate(attribute, c("attribute", "attribute_value"), sep = "=") %>%
    pivot_wider(names_from = attribute, values_from = attribute_value) %>% filter(gene_biotype == "rRNA")
```
