---
title: "WBC depletion methods - analysis"
output:
  html_document:
    df_print: paged
    toc: true
    toc_float: true
    code_folding: hide
---

```{r setup}
knitr::opts_knit$set(root.dir = here::here())
```

```{r include=FALSE}
# Since some of the required packages are only available in Bioconductor,
# renv needs to be initialised as follows:
# renv::init(bioconductor = "3.19")

library(BiocManager)
library(here)
library(ggplot2)
library(viridis)
library(DESeq2)
library(tximport)
library(txdbmaker)
library(vsn)
library(hexbin)
library(pheatmap)
library(dplyr)
library(tidyr)
library(readr)
library(purrr)
library(stringr)
library(GGally)
library(VennDiagram)
library(UpSetR)
# library(GenomicFeatures)

ggplot2::theme_set(theme_bw())
```

# Data import

**If the RDS objects have already been created in a previous run, the following cell can be uncommented and run, and the remaining cells in the data import section can be skipped. See [RDS object section](#rds-objects)).**

```{r message = FALSE}
# data_list <- list.files(path = here("results-refseq/r_objects"),
#                         pattern = ".RDS",
#                         recursive = TRUE, full.names = TRUE) %>%
#   set_names(., str_remove(basename(.), ".RDS")) %>%
#   purrr::imap(~ readRDS(file = .x ))
# 
# list2env(data_list, .GlobalEnv)
# rm(data_list)
```


## Read in fastq-screen and mapping statistics

```{r}
# read in fastq screen files
fastq_screen <- list.files(
  path = here("results-refseq/fastq-screen"), pattern = "_screen.txt",
  recursive = TRUE, full.names = TRUE
) %>%
  read_delim(skip = 1, delim = "\t", n_max = 2, id = "read", show_col_types = FALSE) %>%
  rename("#Multiple_hits_multiple_genomes" = Multiple_hits_multiple_genomes) %>%
  mutate(read = str_remove(basename(read), "_screen.txt")) %>%
  mutate(sample = as.factor(str_remove(read, "_R[1,2].*"))) %>%
  mutate(Genome = as.factor(Genome))

# combine R1 and R2 per species and recalculate percentages
fastq_screen <- fastq_screen %>%
  select(!starts_with("%")) %>%
  group_by(sample, Genome) %>%
  summarise(across(.cols = `#Reads_processed`:`#Multiple_hits_multiple_genomes`, .fns = ~ sum(.x)), .groups = "drop_last") %>%
  mutate(across(.cols = `#Unmapped`:`#Multiple_hits_multiple_genomes`, .fns = ~ .x / `#Reads_processed`, .names = "%_{.col}")) %>%
  rename_with(~ str_replace(.x, "#", ""), .cols = starts_with("%"))

# read in flagstat and metadata file and join with fastq screen df
mapping_stats_df <- read_delim((here("results-refseq/mapping_stats.csv")), col_types = list(sample = "factor")) %>%
  mutate(primary_mapped_human = primary_mapped_all - primary_mapped_pk) %>%
  mutate(primary_mapped_human = primary_mapped_all - primary_mapped_pk)

# join the two dataframes
fastq_screen_mapping_stats_df <- mapping_stats_df %>%
  left_join(fastq_screen, by = "sample")
```

## Read in `salmon` transcript counts/abundance files

`salmon` outputs two types of counts: transcript-level (`quant.sf`) and gene-level (`quant.genes.sf`). Both offer feature length, effective length, TPM, and number of reads. However, when importing counts with `tximport` for downstream differential expression analysis, transcript-level is preferred, because `tximport` will perform gene-level aggregation across all samples, instead of the per-sample aggregation that is done for the `quant.genes.sf` files. See: https://github.com/COMBINE-lab/salmon/issues/437.

For `tximport`, there are still several different ways of creating count matrices. See the description of different options provided by the nf-core/rnaseq pipeline: https://nf-co.re/rnaseq/3.14.0/docs/output/#pseudoalignment and the `tximport` docs here: https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html.

```{r include=FALSE}
# read in sample metadata
samples <- read_csv(here("./data/samplesheet.csv"), col_select = c("sample", "condition", "RNA_type", "library_kit", "removal", "WBC_depletion", "kit"), col_types = "c")

# read in transcript info - transcript id, gene id and genes names
transcript_info <- read.csv(here("results-refseq/star_salmon/tx2gene.tsv"),
  sep = "\t", header = FALSE,
  col.names = c("tx", "gene_id", "gene_name")
)
transcript_info <- list(
  transcript = transcript_info,
  gene = unique(transcript_info[, 2:3]),
  tx2gene = transcript_info[, 1:2]
)

# salmon quant.genes.sf gene-level counts
# per-sample aggregation
gene_counts_salmon <- list.files(
  path = here("results-refseq/star_salmon"), pattern = "quant.genes.sf",
  recursive = TRUE, full.names = TRUE
) %>%
  read_delim(id = "sample") %>%
  mutate(sample = basename(dirname(sample))) %>%
  rename(gene_id = Name) %>%
  left_join(transcript_info$gene, by = "gene_id")

# txi gene-level counts - same as using tximport and DESeqDataSetFromTximport except rounding needs to be done manually
# gene-level aggregation across all samples
gene_counts_txi <- read_delim(here("results-refseq/star_salmon/salmon.merged.gene_counts.tsv")) %>%
  mutate(across(starts_with("X"), \(x) as.integer(round(x))))

# txi gene-level TPM abundances
gene_tpm <- read_delim(here("results-refseq/star_salmon/salmon.merged.gene_tpm.tsv"))

# # txi gene-level library-scaled counts
# gene_counts_scaled <- read_delim(here("results-refseq/star_salmon/salmon.merged.gene_counts_scaled.tsv"))

# # txi gene-level length + library scaled counts
# gene_counts_length_scaled <- read_delim(here("results-refseq/star_salmon/salmon.merged.gene_counts_length_scaled.tsv"))
```

## DESeq tximport objects

The final variant is the DESeq2 approach using `vst` (variance stabilizing transformation) on the raw counts (see [DESeq2 vignette](https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#data-transformations-and-visualization). This is the preferred approach for visualisations. The main goal of this transformation is to remove the dependence of the variance on the mean. However, since we have many zero counts (e.g., rRNA being depleted in some samples), the transformation might not work as well as intended.

```{r message=FALSE}
# txi original counts + offset - to be used with DESeq2's vst/rlog transform for visualisations
salmon_quant_files <- paste0("./results-refseq/star_salmon/", list.files(path = "./results-refseq/star_salmon/", pattern = "quant.sf", recursive = TRUE))

# all genes
txi <- tximport(salmon_quant_files, type = "salmon", tx2gene = transcript_info$tx2gene, txOut = FALSE)
gse <- DESeqDataSetFromTximport(txi, colData = samples, design = ~1)

# variance stabilizing transformation
vsd <- vst(gse, blind = TRUE)
meanSdPlot(assay(vsd))

# extract transformed gene counts
# gene_counts_vst <- data.frame(assays(vsd)[[1]])
# colnames(gene_counts_vst) <- colData(gse)$sample
# gene_counts_vst <- gene_counts_vst %>%
#   mutate(gene_id = rownames(gene_counts_vst)) %>%
#   left_join(transcript_info$gene, by = "gene_id")

# Pk genes
txi_pk <- tximport(salmon_quant_files,
  type = "salmon",
  tx2gene = transcript_info$tx2gene %>% filter(str_detect(gene_id, "PKNH")),
  txOut = FALSE
)
gse_pk <- DESeqDataSetFromTximport(txi_pk, colData = samples, design = ~1)

# variance stabilizing transformation
vsd_pk <- vst(gse_pk, blind = TRUE)
meanSdPlot(assay(vsd_pk))

# extract transformed gene counts
# gene_counts_vst_pk <- data.frame(assays(vsd_pk)[[1]])
# colnames(gene_counts_vst_pk) <- colData(vsd_pk)$sample
# gene_counts_vst_pk <- gene_counts_vst_pk %>%
#   mutate(gene_id = rownames(gene_counts_vst_pk)) %>%
#   left_join(transcript_info$gene, by = "gene_id")

# clean up
rm(txi)
rm(txi_pk)
```

## Filtering raw counts

The raw counts can optionally be pre-filtered, e.g. to a minimum of 10 reads for at least a single sample, as suggested by the DESeq2 vignette.

```{r message=FALSE}
# create filter for all genes on the level of the DESeq/tximport gene counts
filter_all_genes <- rowSums(counts(gse) >= 10) >= 1
# should be the same as filtering on tximport raw gene counts from nf-core pipeline
stopifnot(all( filter_all_genes == (rowSums(gene_counts_txi[,3:18] >= 10) >= 1)))

# create filtered versions of full deseq2 objects
gse_filtered <- gse[filter_all_genes, ]
vsd_filtered <- vst(gse_filtered, blind = TRUE)
meanSdPlot(assay(vsd_filtered))

# create filter for pk genes
filter_pk_genes <- rowSums(counts(gse_pk) >= 10) >= 1

# create filtered versions of Pk deseq2 objects
gse_pk_filtered <- gse_pk[filter_pk_genes, ]
vsd_pk_filtered <- vst(gse_pk_filtered, blind = TRUE)
meanSdPlot(assay(vsd_pk_filtered))

# create filtered versions of other count matrices
gene_tpm_filtered <- gene_tpm[filter_all_genes,]
gene_counts_txi_filtered <- gene_tpm[filter_all_genes,]
# gene counts salmon are in long format, so use join instead of filter mask
gene_counts_salmon_filtered <- gene_counts_salmon %>% inner_join(gene_tpm_filtered %>% select("gene_id"), by = "gene_id")

# # filtering on pk tximport list should be the same as extracting pk genes from all gene filter
# gse_pk_filtered2 <- gene_counts_gse_filtered[gene_counts_gse_filtered$species == "pk",1:18]
# 
# gene_counts_pk_gse_filtered <- data.frame(assays(gse_pk_filtered)[[1]])
# colnames(gene_counts_pk_gse_filtered) <- colData(gse)$sample
# gene_counts_pk_gse_filtered <- gene_counts_pk_gse_filtered %>%
#   mutate(gene_id = rownames(gene_counts_pk_gse_filtered)) %>%
#   left_join(transcript_info$gene, by = "gene_id")
# 
# stopifnot(all(gse_pk_filtered2 == gene_counts_pk_gse_filtered))

# for the other matrices, filtering using deseq-filter list should be the same as filtering on tximport csv
# gene_counts_gse_filtered <- data.frame(assays(gse_filtered)[[1]])
# colnames(gene_counts_gse_filtered) <- colData(gse)$sample
# gene_counts_gse_filtered <- gene_counts_gse_filtered %>%
#   mutate(gene_id = rownames(gene_counts_gse_filtered)) %>%
#   left_join(transcript_info$gene, by = "gene_id")
# gene_tpm_filtered <- gene_tpm %>% inner_join(gene_counts_gse_filtered %>% select("gene_id"), by = "gene_id")
# 
# gene_tpm_filtered2 <- gene_tpm[(rowSums(gene_counts_txi[,3:18] >= 10) >= 1),]
# 
# stopifnot(all(gene_tpm_filtered == gene_tpm_filtered2))
# stopifnot(all(gene_tpm[filter_all_genes,] == gene_tpm_filtered))
```

The number of genes were filtered down from `r dim(gse)[1]`  to `r dim(gse_filtered)[1]` and from `r dim(gse_pk)[1]` to `r dim(gse_pk_filtered)[1]`, for all genes and *P. knowlesi* genes respectively (only for VST transformed counts, not for the TPM or other count matrices). So in conclusion, the effect is relatively minor for *P. knowlesi*, but for all reads a substantial number of low counts are removed.

It is difficult to determine an appropriate threshold for filtering on the TPM level, so instead we filter the same genes as before on the raw count level (>=10 for at least 1 sample).

```{r}
dim(gene_tpm)
mean.tpm <- apply(gene_tpm[, 3:18], 1, mean)
hist(log10(mean.tpm))
# hist(log10(mean.tpm+1e-5))

dim(gene_tpm_filtered)
mean.tpm <- apply(gene_tpm_filtered[, 3:18], 1, mean)
hist(log10(mean.tpm))
```

## Add species information

Split datasets into *P. knowlesi* and human subsets:

```{r}
add_species <- function(data) {
  data %>%
    mutate(species = case_when(
      str_detect(gene_id, "PKNH") ~ "pk",
      .default = "human"
    ))
}

gene_counts_salmon <- add_species(gene_counts_salmon)
gene_counts_salmon_filtered <- add_species(gene_counts_salmon_filtered)

gene_counts_txi <- add_species(gene_counts_txi)
gene_counts_txi_filtered <- add_species(gene_counts_txi_filtered)

gene_tpm <- add_species(gene_tpm)
gene_tpm_filtered <- add_species(gene_tpm_filtered)
```

## Annotating rRNA and globin genes

### rRNA {.tabset}

The RefSeq `gff` files contains info on `rRNA` in the `gene_biotype` attribute of gene entries (including mitochondrial rRNA) for both species (i.e., we collected featured labelled as `gene` in the 3rd column of the gff, not rRNA directly, since counts were aggregated on the gene level).

```
NC_011902.2	RefSeq	gene	825471	825594	.	+	.	ID=gene-PKNH_0117700;Dbxref=GeneID:7318405;Name=PKNH_0117700;end_range=825594,.;gbkey=Gene;gene_biotype=rRNA;locus_tag=PKNH_0117700;old_locus_tag=PKH_011712;partial=true;start_range=.,825471

NC_011902.2	RefSeq	rRNA	825471	825594	.	+	.	ID=rna-XR_002461722.1;Parent=gene-PKNH_0117700;Dbxref=GeneID:7318405,GenBank:XR_002461722.1;Name=XR_002461722.1;end_range=825594,.;gbkey=rRNA;locus_tag=PKNH_0117700;orig_transcript_id=gnl|WGS:CADCXE|mrna.PKNH_0117700-RA;partial=true;product=5.8S ribosomal RNA;start_range=.,825471;transcript_id=XR_002461722.1

NC_011902.2	RefSeq	exon	825471	825594	.	+	.	ID=exon-XR_002461722.1-1;Parent=rna-XR_002461722.1;Dbxref=GeneID:7318405,GenBank:XR_002461722.1;end_range=825594,.;gbkey=rRNA;locus_tag=PKNH_0117700;orig_transcript_id=gnl|WGS:CADCXE|mrna.PKNH_0117700-RA;partial=true;product=5.8S ribosomal RNA;start_range=.,825471;transcript_id=XR_002461722.1
```

The bash script in [`scripts/2_expression_analysis/0_gene_annotations-refseq.sh`](./scripts/2_expression_analysis/0_gene_annotations-refseq.sh) was used to retrieve the rRNA from both annotation files.

```{bash}
#| echo: fenced
# zcat "${ref_human}" | awk '$3 ~ /gene/' | cut -f9 | grep -E "gene_biotype=rRNA" | grep -v "rRNA_pseudo" | grep -o "ID=[^;]*" | sed 's/ID=/human_rRNA;/' >> "${output_file}"
# zcat "${ref_pk}" | cut -f9 | grep -E "gene_biotype=rRNA" | grep -o "ID=[^;]*" | sed 's/ID=/pk_rRNA;/' >> "${output_file}"
```

```{r message=FALSE}
# read in gene type table
gene_types_df <- read_csv2(here("data/ref-refseq-refseq/gene_types.csv")) %>% left_join(transcript_info$gene, by = "gene_id")
```

**Note that the GENCODE and RefSeq annotation file contain different subsets of annotated rRNA genes!** See semi-comprehensive list of human rRNA genes here: https://www.genenames.org/data/genegroup/#!/group/848. Some are not included in any reference assembly, but some are only present in either GENCODE or RefSeq. Some examples include RNA18S and 28S being present in RefSeq only. Moreover, for GENCODE, the majority of rRNA counts are zero across the board, with the exception of:

See tables below for a full breakdown:

<!-- (Checked using `awk -F';' '$1 ~ /human_rRNA/ {print $2}' data/ref-refseq-refseq/gene_types.csv | while read -r line; do echo "$line"; grep "${line}" results-refseq/star_salmon/tx2gene.tsv; grep "$line" results-refseq/star_salmon/**/quant.gene*; done`) -->

```{r message=FALSE}
knitr::kable(
  gene_types_df %>%
    filter((gene_type == "human_rRNA") | (gene_type == "Mt_rRNA")) %>%
    left_join(gene_tpm, by = c("gene_id", "gene_name")) %>%
    filter(rowSums(across(starts_with("X"))) != 0),
  caption = "Non-zero counts of human rRNA genes annotated in RefSeq"
)

# check if GENCODE results are also present in directory
if (file.exists(here("data/ref/gene_types.csv"))) {
  transcript_info_gencode <- read.csv(here("results/star_salmon/tx2gene.tsv"),
    sep = "\t", header = FALSE,
    col.names = c("tx", "gene_id", "gene_name")
  )
  transcript_info_gencode <- list(
    transcript = transcript_info_gencode,
    gene = unique(transcript_info_gencode[, 2:3]),
    tx2gene = transcript_info_gencode[, 1:2]
  )
  gene_types_gencode_df <- read_csv2(here("data/ref/gene_types.csv")) %>%
    left_join(transcript_info_gencode$gene, by = "gene_id")

  gene_tpm_gencode <- read_delim(here("results/star_salmon/salmon.merged.gene_tpm.tsv"))

  knitr::kable(
    gene_types_gencode_df %>%
      filter((gene_type == "human_rRNA") | (gene_type == "Mt_rRNA")) %>%
      left_join(gene_tpm_gencode, by = c("gene_id", "gene_name")) %>%
      filter(rowSums(across(starts_with("X"))) != 0),
    caption = "Non-zero counts of human rRNA genes annotated in GENCODE"
  )
}
```

### Globins

For human globins, the following genes were included: "HBB","HBD","HBBP1","HBG1","HBG2","HBE1","HBZ","HBM","HBA2","HBA1","HBQ1".

```{r}
globin_names <- c("HBB", "HBD", "HBBP1", "HBG1", "HBG2", "HBE1", "HBZ", "HBM", "HBA2", "HBA1", "HBQ1")
globin_df <- tibble(gene_name = globin_names, gene_type = "human_globin") %>% left_join(transcript_info$gene, by = "gene_name")

gene_types_df <- gene_types_df %>% bind_rows(globin_df)

# export full gene type file
write_csv2(gene_types_df, here("data/ref-refseq-refseq/gene_types_R.csv"))

# some gene id share the same gene name
# gene_types_df %>%
#     group_by(gene_name) %>%
#     filter(n()>1)
```

Merge annotations with counts and fill in missing gene_types (= non-rRNA/globin) with species identifier:

```{r}
# merge gene_type annotation and fill in missing types using species name
annotate_gene_type <- function(data, gene_types) {
  data %>%
    left_join(gene_types, by = c("gene_id", "gene_name")) %>%
    mutate("gene_type" = coalesce(gene_type, species))
}

gene_counts_salmon <- annotate_gene_type(gene_counts_salmon, gene_types_df)
gene_counts_salmon_filtered <- annotate_gene_type(gene_counts_salmon_filtered, gene_types_df)

gene_counts_txi <- annotate_gene_type(gene_counts_txi, gene_types_df)
gene_counts_txi_filtered <- annotate_gene_type(gene_counts_txi_filtered, gene_types_df)

gene_tpm <- annotate_gene_type(gene_tpm, gene_types_df)
gene_tpm_filtered <- annotate_gene_type(gene_tpm_filtered, gene_types_df)

# cast to long format
to_long <- function(data) {
  data %>%
    pivot_longer(cols = `X104974.001.001_S1_L001`:`X104974.001.016_S1_L001`, names_to = "sample", values_to = "counts") %>%
    mutate(sample = str_replace_all(sample, "\\.", replacement = "-")) %>%
    mutate(sample = str_sub(sample, 2, -1))
}

gene_counts_txi_long <- to_long(gene_counts_txi)
gene_counts_txi_filtered_long <- to_long(gene_counts_txi_filtered)

gene_tpm_long <- to_long(gene_tpm)
gene_tpm_filtered_long <- to_long(gene_tpm_filtered)
```

### Other gene types

Extract general gene types from GFF file:

```{r message=FALSE, cache=TRUE}
# https://stackoverflow.com/questions/63969719/tidyr-separate-a-column-into-a-variable-number-of-columns
gff <- read_delim(here("data/ref-refseq-refseq/GCF_000006355.2_GCA_000006355.2_genomic.gff"), comment = "#", col_names = c("seqname", "source", "feature", "start", "end", "score", "strand", "frame", "attribute")) %>%
  mutate(id = row_number()) %>%
  separate_rows(attribute, sep = ";") %>%
  separate(attribute, c("attribute", "attribute_value"), sep = "=") %>%
  pivot_wider(names_from = attribute, values_from = attribute_value) %>%
  filter(!is.na(gene_biotype)) %>%
  select(c(gene_biotype, ID)) %>%
  rename(gene_id = ID) %>%
  left_join(transcript_info$gene, by = "gene_id")
```


## Export RDS objects {#rds-objects}

Save all output to rds object for easier loading.

```{r message = FALSE}
# dir.create(here("results-refseq/r_objects/"), showWarnings = FALSE)
# 
# purrr::iwalk(
#   tibble:::lst(
#     fastq_screen,
#     fastq_screen_mapping_stats_df,
#     gene_counts_salmon,
#     gene_counts_salmon_filtered,
#     gene_counts_txi,
#     gene_counts_txi_long,
#     gene_counts_txi_filtered,
#     gene_counts_txi_filtered_long,
#     gene_tpm,
#     gene_tpm_long,
#     gene_tpm_filtered,
#     gene_tpm_filtered_long,
#     gene_types_df,
#     gene_types_gencode_df,
#     gff,
#     globin_df,
#     gse,
#     gse_filtered,
#     gse_pk,
#     gse_pk_filtered,
#     mapping_stats_df,
#     samples,
#     transcript_info,
#     transcript_info_gencode,
#     vsd,
#     vsd_filtered,
#     vsd_pk,
#     vsd_pk_filtered
#   ),
#   ~ saveRDS(.x, file = here(paste0("results-refseq/r_objects/", .y, ".RDS")))
# )
# 
# # clean up
# rm(gene_types_df)
# rm(globin_df)
# rm(gene_tpm_gencode)
# rm(gene_types_gencode_df)
# rm(transcript_info_gencode)
```

# Analyses and figures

```{r show=FALSE, message = FALSE}
dir.create(here("results-refseq/figures/"), showWarnings = FALSE)
```


## Flagstat mapping statistics

```{r}
mapping_stats_df
```

## Top expressed reads in TPM {.tabset}

```{r}
gene_tpm_long %>%
  group_by(sample) %>%
  top_n(25, counts) %>%
  arrange(sample, desc(counts), .by_group = T)
```

```{r,results='asis', echo=FALSE}
for (i in unique(gene_tpm_long$sample)) {
  cat("### Sample ", i, " \n")
  print(knitr::kable(
    gene_tpm_long %>%
      filter(sample == i) %>%
      top_n(25, counts) %>%
      arrange(sample, desc(counts), .by_group = T)
  ))
  cat("\n")
}
```

## FastQ Screen {.tabset}

Also see figure in [results-refseq/fastq-screen/](results-refseq/fastq-screen/) and [results-refseq/multiqc-fastq-screen-samtools-pk/multiqc_report.html](results-refseq/multiqc-fastq-screen-samtools-pk/multiqc_report.html).

### Combined - sample names

```{r}
fastq_screen_transformed <- bind_rows(
  fastq_screen %>%
    mutate(`%_single_genome` = `%_One_hit_one_genome` + `%_Multiple_hits_one_genome`) %>%
    mutate(`%_multiple_genomes` = `%_One_hit_multiple_genomes` + `%_Multiple_hits_multiple_genomes`) %>%
    mutate(`No_hits` = `%_One_hit_multiple_genomes` + `%_Multiple_hits_multiple_genomes`) %>%
    select(c(sample, Genome, `#Reads_processed`, `%_single_genome`, `%_multiple_genomes`)) %>%
    pivot_wider(names_from = Genome, values_from = c(`%_single_genome`, `%_multiple_genomes`)) %>%
    select(!`%_multiple_genomes_PknowlesiH`) %>%
    rename(`%_multiple_genomes` = `%_multiple_genomes_Human`) %>%
    mutate(pct = 1 - `%_single_genome_PknowlesiH` - `%_single_genome_Human` - `%_multiple_genomes`) %>%
    select(sample, pct) %>%
    mutate("Mapped reads" = "No hits"),
  fastq_screen %>%
    mutate(pct = `%_One_hit_one_genome` + `%_Multiple_hits_one_genome`) %>%
    select(c(sample, Genome, pct)) %>%
    rename("Mapped reads" = Genome),
  fastq_screen %>%
    mutate(pct = `%_One_hit_multiple_genomes` + `%_Multiple_hits_multiple_genomes`) %>%
    select(c(sample, pct)) %>%
    distinct(sample, .keep_all = TRUE) %>%
    mutate("Mapped reads" = "Multiple genomes")
) %>%
  mutate(`Mapped reads` = factor(`Mapped reads`, levels = c("Human", "PknowlesiH", "Multiple genomes", "No hits")))

ggplot(fastq_screen_transformed, aes(x = sample, y = pct, fill = `Mapped reads`)) +
  geom_bar(position = "stack", stat = "identity") +
  scale_fill_viridis(option = "viridis", discrete = T) +
  coord_flip() +
  scale_x_discrete(limits = rev) +
  labs(fill = expression("FastQ Screen mapped reads")) +
  ylab(label = "% of mapped reads")
```

### Combined - conditions

```{r}
ggplot(fastq_screen_transformed %>% left_join(samples, by = "sample"), aes(x = condition, y = pct, fill = `Mapped reads`)) +
  geom_bar(position = "stack", stat = "identity") +
  scale_fill_viridis(option = "viridis", discrete = T) +
  coord_flip() +
  scale_x_discrete(limits = rev) +
  labs(fill = expression("FastQ Screen mapped reads")) +
  ylab(label = "% of mapped reads")


# fastq_screen %>%
#   filter(Genome == "Human") %>%
#   mutate(`No hits` = )
#
#   pivot_longer(cols = starts_with("%"),
#                               names_to = "fastq_screen_pct_type",
#                               values_to = "fastq_screen_pct_value")
#
# fastq_screen %>%
#   group_by(sample, Genome) %>%
#   summarise(`%_total_hits` = sum(c(`%_One_hit_one_genome`, `%_Multiple_hits_one_genome`))) %>%
#   ungroup()
```

### Human reads

```{r}
# Human reads
ggplot(
  fastq_screen %>% pivot_longer(
    cols = starts_with("%"),
    names_to = "fastq_screen_pct_type",
    values_to = "fastq_screen_pct_value"
  ) %>%
    filter(Genome == "Human"),
  aes(x = sample, y = fastq_screen_pct_value, fill = fastq_screen_pct_type)
) +
  geom_bar(position = "stack", stat = "identity") +
  scale_fill_viridis(option = "viridis", discrete = T) +
  coord_flip() +
  scale_x_discrete(limits = rev) +
  labs(fill = "FastQ screen statistics for human reads")
```

### Pk reads

```{r}
# Pk reads
ggplot(
  fastq_screen %>% pivot_longer(
    cols = starts_with("%"),
    names_to = "fastq_screen_pct_type",
    values_to = "fastq_screen_pct_value"
  ) %>%
    filter(Genome == "PknowlesiH"),
  aes(x = sample, y = fastq_screen_pct_value, fill = fastq_screen_pct_type)
) +
  geom_bar(position = "stack", stat = "identity") +
  scale_fill_viridis(option = "viridis", discrete = T) +
  coord_flip() +
  scale_x_discrete(limits = rev) +
  labs(fill = expression("FastQ screen statistics for" ~ italic("P. knowlesi") ~ "reads"))
```

## Plot percentage of reads for each gene type (rRNA/globin) and species {.tabset}

**TPM values appear to be the most appropriate metric to use.**

### Salmon gene-level counts

These gene counts were aggregated at the sample level (salmon gene output).

```{r}
gene_counts_salmon_by_sample_gene_type <- gene_counts_salmon %>%
  left_join(mapping_stats_df, by = "sample") %>%
  group_by(condition, gene_type) %>%
  summarise(sum_reads = sum(NumReads, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads))

p <- ggplot(gene_counts_salmon_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(position = "fill", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_salmon_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(position = "dodge", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_salmon_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()
```

### Salmon gene-level counts - filtered

These gene counts were aggregated at the sample level (salmon gene output) and filtered to only genes with a minimum count of 10 reads for at least 1 sample.

*Filtering seems to work ok on dataframe level, but the plots do not appear to change because the differences in total count are so minor.**

```{r}
gene_counts_salmon_filtered_by_sample_gene_type <- gene_counts_salmon_filtered %>%
  left_join(mapping_stats_df, by = "sample") %>%
  group_by(condition, gene_type) %>%
  summarise(sum_reads = sum(NumReads, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads))

p <- ggplot(gene_counts_salmon_filtered_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(position = "fill", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_salmon_filtered_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(position = "dodge", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_salmon_filtered_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()
```

### Gene-level counts tximport

These gene counts were aggregated across all samples (tximport of transcript counts).

```{r}
gene_counts_txi_by_sample_gene_type <- gene_counts_txi_long %>%
  left_join(mapping_stats_df, by = "sample") %>%
  group_by(condition, gene_type) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads))

p <- ggplot(gene_counts_txi_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(position = "fill", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_txi_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(position = "dodge", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_txi_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()
```

### Gene-level counts tximport - filtered

These gene counts were aggregated across all samples (tximport of transcript counts) and filtered to only genes with a minimum count of 10 reads for at least 1 sample.

```{r}
gene_counts_txi_filtered_by_sample_gene_type <- gene_counts_txi_filtered_long %>%
  left_join(mapping_stats_df, by = "sample") %>%
  group_by(condition, gene_type) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads))

p <- ggplot(gene_counts_txi_filtered_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(position = "fill", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_txi_filtered_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(position = "dodge", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_txi_filtered_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()
```

### Gene-level TPM

```{r}
gene_tpm_by_sample_gene_type <- gene_tpm_long %>%
  left_join(mapping_stats_df, by = "sample") %>%
  group_by(condition, gene_type) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads))

p <- ggplot(gene_tpm_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(position = "fill", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_tpm_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(position = "dodge", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_tpm_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()
```

### Gene-level TPM - filtered

Filtered to only genes with a minimum count of 10 reads for at least 1 sample.

```{r}
gene_tpm_filtered_by_sample_gene_type <- gene_tpm_filtered_long %>%
  left_join(mapping_stats_df, by = "sample") %>%
  group_by(condition, gene_type) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads))

p <- ggplot(gene_tpm_filtered_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(position = "fill", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_tpm_filtered_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(position = "dodge", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_tpm_filtered_by_sample_gene_type, aes(y = sum_reads, x = condition, fill = gene_type))
p + geom_bar(stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()
```

## Plot percentage of reads for each species {.tabset}

### Salmon gene-level counts {.tabset}

These gene counts were aggregated at the sample level (salmon gene output).

#### Samples

```{r}
gene_counts_salmon_by_sample_species <- gene_counts_salmon %>%
  left_join(mapping_stats_df, by = "sample") %>%
  left_join(samples) %>%
  group_by(sample, species) %>%
  summarise(sum_reads = sum(NumReads, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads))

p <- ggplot(gene_counts_salmon_by_sample_species, aes(y = sum_reads, x = sample, fill = species))
p + geom_bar(position = "fill", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_salmon_by_sample_species, aes(y = sum_reads, x = sample, fill = species))
p + geom_bar(position = "dodge", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_salmon_by_sample_species, aes(y = sum_reads, x = sample, fill = species))
p + geom_bar(stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()
```

#### Conditions

```{r}
gene_counts_salmon_by_sample_species <- gene_counts_salmon %>%
  left_join(mapping_stats_df, by = "sample") %>%
  left_join(samples) %>%
  group_by(condition, species) %>%
  summarise(sum_reads = sum(NumReads, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads))

p <- ggplot(gene_counts_salmon_by_sample_species, aes(y = sum_reads, x = condition, fill = species))
p + geom_bar(position = "fill", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_salmon_by_sample_species, aes(y = sum_reads, x = condition, fill = species))
p + geom_bar(position = "dodge", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_salmon_by_sample_species, aes(y = sum_reads, x = condition, fill = species))
p + geom_bar(stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()
```

### Salmon gene-level counts - filtered {.tabset}

These gene counts were aggregated at the sample level (salmon gene output) and filtered to only genes with a minimum count of 10 reads for at least 1 sample.

#### Samples

```{r}
gene_counts_salmon_filtered_by_sample_species <- gene_counts_salmon_filtered %>%
  left_join(mapping_stats_df, by = "sample") %>%
  left_join(samples) %>%
  group_by(sample, species) %>%
  summarise(sum_reads = sum(NumReads, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads))

p <- ggplot(gene_counts_salmon_filtered_by_sample_species, aes(y = sum_reads, x = sample, fill = species))
p + geom_bar(position = "fill", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_salmon_filtered_by_sample_species, aes(y = sum_reads, x = sample, fill = species))
p + geom_bar(position = "dodge", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_salmon_filtered_by_sample_species, aes(y = sum_reads, x = sample, fill = species))
p + geom_bar(stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()
```

#### Conditions

```{r}
gene_counts_salmon_filtered_by_sample_species <- gene_counts_salmon_filtered %>%
  left_join(mapping_stats_df, by = "sample") %>%
  left_join(samples) %>%
  group_by(condition, species) %>%
  summarise(sum_reads = sum(NumReads, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads))

p <- ggplot(gene_counts_salmon_filtered_by_sample_species, aes(y = sum_reads, x = condition, fill = species))
p + geom_bar(position = "fill", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_salmon_filtered_by_sample_species, aes(y = sum_reads, x = condition, fill = species))
p + geom_bar(position = "dodge", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_salmon_filtered_by_sample_species, aes(y = sum_reads, x = condition, fill = species))
p + geom_bar(stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()
```

### Gene-level counts tximport {.tabset}

These gene counts were aggregated across all samples (tximport of transcript counts).

#### Samples

```{r}
gene_counts_txi_by_samples_species <- gene_counts_txi_long %>%
  left_join(mapping_stats_df, by = "sample") %>%
  left_join(samples) %>%
  group_by(sample, species) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads))

p <- ggplot(gene_counts_txi_by_samples_species, aes(y = sum_reads, x = sample, fill = species))
p + geom_bar(position = "fill", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_txi_by_samples_species, aes(y = sum_reads, x = sample, fill = species))
p + geom_bar(position = "dodge", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_txi_by_samples_species, aes(y = sum_reads, x = sample, fill = species))
p + geom_bar(stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()
```

#### Conditions

```{r}
gene_counts_txi_by_samples_species <- gene_counts_txi_long %>%
  left_join(mapping_stats_df, by = "sample") %>%
  left_join(samples) %>%
  group_by(condition, species) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads))

p <- ggplot(gene_counts_txi_by_samples_species, aes(y = sum_reads, x = condition, fill = species))
p + geom_bar(position = "fill", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_txi_by_samples_species, aes(y = sum_reads, x = condition, fill = species))
p + geom_bar(position = "dodge", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_txi_by_samples_species, aes(y = sum_reads, x = condition, fill = species))
p + geom_bar(stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()
```

### Gene-level counts tximport - filtered {.tabset}

These gene counts were aggregated across all samples (tximport of transcript counts) and filtered to only genes with a minimum count of 10 reads for at least 1 sample.

#### Samples

```{r}
gene_counts_txi_filtered_by_samples_species <- gene_counts_txi_filtered_long %>%
  left_join(mapping_stats_df, by = "sample") %>%
  left_join(samples) %>%
  group_by(sample, species) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads))

p <- ggplot(gene_counts_txi_filtered_by_samples_species, aes(y = sum_reads, x = sample, fill = species))
p + geom_bar(position = "fill", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_txi_filtered_by_samples_species, aes(y = sum_reads, x = sample, fill = species))
p + geom_bar(position = "dodge", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_txi_filtered_by_samples_species, aes(y = sum_reads, x = sample, fill = species))
p + geom_bar(stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()
```

#### Conditions

```{r}
gene_counts_txi_filtered_by_samples_species <- gene_counts_txi_filtered_long %>%
  left_join(mapping_stats_df, by = "sample") %>%
  left_join(samples) %>%
  group_by(condition, species) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads))

p <- ggplot(gene_counts_txi_filtered_by_samples_species, aes(y = sum_reads, x = condition, fill = species))
p + geom_bar(position = "fill", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_txi_filtered_by_samples_species, aes(y = sum_reads, x = condition, fill = species))
p + geom_bar(position = "dodge", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_counts_txi_filtered_by_samples_species, aes(y = sum_reads, x = condition, fill = species))
p + geom_bar(stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()
```

### Gene-level TPM {.tabset}

#### Samples

```{r}
gene_tpm_by_samples_species <- gene_tpm_long %>%
  left_join(mapping_stats_df, by = "sample") %>%
  left_join(samples) %>%
  group_by(sample, species) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads))

p <- ggplot(gene_tpm_by_samples_species, aes(y = sum_reads, x = sample, fill = species))
p + geom_bar(position = "fill", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_tpm_by_samples_species, aes(y = sum_reads, x = sample, fill = species))
p + geom_bar(position = "dodge", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_tpm_by_samples_species, aes(y = sum_reads, x = sample, fill = species))
p + geom_bar(stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()
```

#### Conditions

```{r}
gene_tpm_by_samples_species <- gene_tpm_long %>%
  left_join(mapping_stats_df, by = "sample") %>%
  left_join(samples) %>%
  group_by(condition, species) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads))

p <- ggplot(gene_tpm_by_samples_species, aes(y = sum_reads, x = condition, fill = species))
p + geom_bar(position = "fill", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_tpm_by_samples_species, aes(y = sum_reads, x = condition, fill = species))
p + geom_bar(position = "dodge", stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()

p <- ggplot(gene_tpm_by_samples_species, aes(y = sum_reads, x = condition, fill = species))
p + geom_bar(stat = "identity") + scale_fill_viridis(option = "viridis", discrete = T) + coord_flip()
```

## Pk specific gene type plots

Try to use https://nf-co.re/rnaseq/1.1/docs/output#featurecounts?

```
└─▶ featureCounts -t "transcript" -g "gene_biotype" -B -C -p -s 1 -T 6 -a data/ref-refseq-refseq/GCF_000006355.2_GCA_000006355.2_genomic.gtf -o results-refseq/featureCounts-pk.txt results-refseq/star_salmon/104974-001-001_S1_L001.markdup.sorted.pk.bam
```

## Gene recovery {.tabset}

### All genes

<!-- ```{r} -->
<!-- gene_counts_vst_pk_filtered %>%  -->
<!--   mutate(across(.cols = starts_with("104974"), .fns = ~ as.integer(.x > 0))) -->

<!-- ``` -->

```{r}
# create binary matrices with presence/absence of genes
# rename sample names to format in sample info df
gene_tpm_binary <- gene_tpm %>%
  mutate(across(.cols = starts_with("X"), .fns = ~ .x > 0)) %>%
  select(!c(gene_name, gene_type, species)) %>%
  rename_with(~ str_replace_all(string = ., pattern = "\\.", replacement = "-"), starts_with("X")) %>%
  rename_with(~ str_sub(., 2, -1), starts_with("X"))

gene_tpm_pk_binary <- gene_tpm %>%
  mutate(across(.cols = starts_with("X"), .fns = ~ .x > 0)) %>%
  filter(species == "pk") %>%
  select(!c(gene_name, gene_type, species)) %>%
  rename_with(~ str_replace_all(string = ., pattern = "\\.", replacement = "-"), starts_with("X")) %>%
  rename_with(~ str_sub(., 2, -1), starts_with("X"))

gene_tpm_pk_filtered_binary <- gene_tpm_pk_binary %>% inner_join(gene_tpm_filtered %>% select(gene_id), by = "gene_id")

gene_tpm_binary_list <- purrr::set_names(colnames(gene_tpm_binary %>% select(starts_with("104974")))) %>%
  purrr::imap(~ gene_tpm_binary$gene_id[gene_tpm_binary[, .x, drop = T]])

gene_tpm_pk_binary_list <- purrr::set_names(colnames(gene_tpm_pk_binary %>% select(starts_with("104974")))) %>%
  purrr::imap(~ gene_tpm_pk_binary$gene_id[gene_tpm_pk_binary[, .x, drop = T]])

gene_tpm_pk_filtered_binary_list <- purrr::set_names(colnames(gene_tpm_pk_filtered_binary %>% select(starts_with("104974")))) %>%
  purrr::imap(~ gene_tpm_pk_filtered_binary$gene_id[gene_tpm_pk_filtered_binary[, .x, drop = T]])

# gene_counts_vst_pk_filtered_binary_list <- gene_counts_vst_pk_filtered %>%
  # mutate(across(.cols = starts_with("104974"), .fns = ~ .x > 0))

# gene_tpm_binary_list <- lapply(colnames(gene_tpm_binary %>%
#   select(starts_with("x"))), function(x) {
#   gene_tpm_binary$gene_id[gene_tpm_binary[, x, drop = T]]
# })
#
# gene_tpm_pk_binary_list <- lapply(colnames(gene_tpm_pk_binary %>%
#   select(starts_with("x"))), function(x) {
#   gene_tpm_pk_binary$gene_id[gene_tpm_pk_binary[, x, drop = T]]
# })

comparison_centpip <- samples %>%
  filter(WBC_depletion == "centpip") %>%
  select(sample)

comparison_plasmo <- samples %>%
  filter(WBC_depletion == "plasmo") %>%
  select(sample)

comparison_pmacs <- samples %>%
  filter(WBC_depletion == "pmacs") %>%
  select(sample)

comparison_cellulose <- samples %>%
  filter(WBC_depletion == "cellulose") %>%
  select(sample)

comparison_nofilter <- samples %>%
  filter(WBC_depletion == "nofilter") %>%
  filter(!sample == "104974-001-006_S1_L001")

draw_venn <- function(set_list, names, title, output) {
  VennDiagram::venn.diagram(
    x = set_list,
    category.names = names,
    # Circles
    lwd = 2,
    # lty = "blank",
    # col = palette.colors(3),
    # fill = palette.colors(3),
    # col = hcl.colors(3, palette = "viridis", alpha = 0.9),
    # fill = hcl.colors(3, palette = "viridis", alpha = 0.4),
    col = hcl.colors(3, palette = "Set 2", alpha = 0.9),
    fill = hcl.colors(3, palette = "Set 2", alpha = 0.4),

    # Numbers
    cex = .9,
    fontface = "bold",
    fontfamily = "sans",

    # Set names
    cat.cex = 0.9,
    cat.fontface = "bold",
    cat.default.pos = "outer",
    cat.pos = c(-20, 20, 135),
    cat.dist = c(0.055, 0.055, 0.11),
    cat.fontfamily = "sans",
    rotation = 1,

    # title
    main = title,
    main.cex = 1.1,
    main.fontface = "bold",
    main.fontfamily = "sans",

    # output features
    height = 1200,
    width = 1200,
    resolution = 300,
    compression = "lzw",
    filename = output,
    output = TRUE
  )
}

v <- draw_venn(
  gene_tpm_binary_list[comparison_centpip$sample],
  c("mRNA Qiagen - FSglob", "tRNA Qiagen - FSglobH", "mRNA Illumina"),
  "Centpip",
  here("results-refseq/figures/venn-centpip.png")
)
v

v <- draw_venn(
  gene_tpm_binary_list[comparison_plasmo$sample],
  c("mRNA Qiagen - Fsglob", "tRNA Qiagen - FSglobH", "mRNA Illumina"),
  "Plasmo",
  here("results-refseq/figures/venn-plasmo.png")
)
v

v <- draw_venn(
  gene_tpm_binary_list[comparison_pmacs$sample],
  c("mRNA Qiagen - FSglob", "tRNA Qiagen - FSglobH", "mRNA Illumina"),
  "Pmacs",
  here("results-refseq/figures/venn-pmacs.png")
)
v

v <- draw_venn(
  gene_tpm_binary_list[comparison_cellulose$sample],
  c("mRNA Qiagen - FSglob", "tRNA Qiagen - FSglobH", "mRNA Illumina"),
  "Cellulose",
  here("results-refseq/figures/venn-cellulose.png")
)
v

v <- draw_venn(
  gene_tpm_binary_list[comparison_nofilter$sample],
  c("mRNA Qiagen - FSglob", "tRNA Qiagen - FSglobH", "mRNA Illumina"),
  "Cellulose",
  here("results-refseq/figures/venn-nofilter.png")
)
v
```

### *P. knowlesi* genes only

```{r}
v <- draw_venn(
  gene_tpm_pk_binary_list[comparison_centpip$sample],
  c("mRNA Qiagen - FSglob", "tRNA Qiagen - FSglobH", "mRNA Illumina"),
  "Centpip",
  here("results-refseq/figures/venn-centpip_pk.png")
)
v

v <- draw_venn(
  gene_tpm_pk_binary_list[comparison_plasmo$sample],
  c("mRNA Qiagen - Fsglob", "tRNA Qiagen - FSglobH", "mRNA Illumina"),
  "Plasmo",
  here("results-refseq/figures/venn-plasmo_pk.png")
)
v

v <- draw_venn(
  gene_tpm_pk_binary_list[comparison_pmacs$sample],
  c("mRNA Qiagen - FSglob", "tRNA Qiagen - FSglobH", "mRNA Illumina"),
  "Pmacs",
  here("results-refseq/figures/venn-pmacs_pk.png")
)
v

v <- draw_venn(
  gene_tpm_pk_binary_list[comparison_cellulose$sample],
  c("mRNA Qiagen - FSglob", "tRNA Qiagen - FSglobH", "mRNA Illumina"),
  "Cellulose",
  here("results-refseq/figures/venn-cellulose_pk.png")
)
v

v <- draw_venn(
  gene_tpm_pk_binary_list[comparison_nofilter$sample],
  c("mRNA Qiagen - FSglob", "tRNA Qiagen - FSglobH", "mRNA Illumina"),
  "No filter",
  here("results-refseq/figures/venn-nofilter_pk.png")
)
v
```

### *P. knowlesi* genes only - filtered

Gene counts were filtered on a minimum (raw) count of 10 for at least 1 sample before checking for presence/absence. 

```{r}
v <- draw_venn(
  gene_tpm_pk_filtered_binary_list[comparison_centpip$sample],
  c("mRNA Qiagen - FSglob", "tRNA Qiagen - FSglobH", "mRNA Illumina"),
  "Centpip",
  here("results-refseq/figures/venn-centpip_pk_filtered.png")
)
v

v <- draw_venn(
  gene_tpm_pk_filtered_binary_list[comparison_plasmo$sample],
  c("mRNA Qiagen - Fsglob", "tRNA Qiagen - FSglobH", "mRNA Illumina"),
  "Plasmo",
  here("results-refseq/figures/venn-plasmo_pk_filtered.png")
)
v

v <- draw_venn(
  gene_tpm_pk_filtered_binary_list[comparison_pmacs$sample],
  c("mRNA Qiagen - FSglob", "tRNA Qiagen - FSglobH", "mRNA Illumina"),
  "Pmacs",
  here("results-refseq/figures/venn-pmacs_pk_filtered.png")
)
v

v <- draw_venn(
  gene_tpm_pk_filtered_binary_list[comparison_cellulose$sample],
  c("mRNA Qiagen - FSglob", "tRNA Qiagen - FSglobH", "mRNA Illumina"),
  "Cellulose",
  here("results-refseq/figures/venn-cellulose_pk_filtered.png")
)
v

v <- draw_venn(
  gene_tpm_pk_filtered_binary_list[comparison_nofilter$sample],
  c("mRNA Qiagen - FSglob", "tRNA Qiagen - FSglobH", "mRNA Illumina"),
  "No filter",
  here("results-refseq/figures/venn-nofilter_pk_filtered.png")
)
v
```

### Upset plot - Pk genes filtered on low counts

**TODO: change labels, order them logically and add additional metadata using bars or rectangles in inkscape. Also choose how many sets/intersets to plot.**

```{r}
UpSetR::upset(as.data.frame(gene_tpm_pk_filtered_binary %>% select(!gene_id) %>% mutate_all(as.integer)))
UpSetR::upset(as.data.frame(gene_tpm_pk_filtered_binary %>% select(!gene_id) %>% mutate_all(as.integer)), nsets = 16, nintersects = NA)
```

```{r}
#
# gene_tpm_overlap <- gene_tpm_overlap %>% mutate(across(.cols = starts_with("X"), ~ +as.logical(.x)))
#
# # UpSetR::upset(gene_tpm_overlap[,2:4])
#
#
#
#
#
# purrr::map(
#   gene_tpm_overlap %>% select(starts_with("x")),
#   ~ gene_tpm_overlap$gene_id[.x, ]
# )
#
# l <- lapply(gene_tpm_overlap %>% select(starts_with("x")), set)
#
#
# length(gene_tpm_overlap$gene_id[gene_tpm_overlap$X104974.001.001_S1_L001])
#
#
#
#
# ggplot(gene_tpm_long %>% filter(sample == "104974-001-001_S1_L001"), aes(x = sample, y = counts)) +
#   geom_boxplot()
#
# ggplot(gene_tpm_long %>% filter(sample == "104974-001-001_S1_L001"), aes(x = log(counts))) +
#   geom_histogram()
#
#
# # gene_tpm_long %>% filter(sample == "104974-001-001_S1_L001") %>% filter(counts > 0)
# gene_tpm_long %>% filter(counts > 0)
```

## Sample-to-sample distances {.tabset}

### All genes

```{r}
sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- vsd$condition
colnames(sampleDistMatrix) <- NULL
pheatmap(sampleDistMatrix,
  clustering_distance_rows = sampleDists,
  clustering_distance_cols = sampleDists,
  color = viridis(
    n = 256, alpha = 1,
    begin = 0, end = 1, option = "viridis"
  )
)
```

### *P. knowlesi* only

```{r}
sampleDists <- dist(t(assay(vsd_pk)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- vsd$condition
colnames(sampleDistMatrix) <- NULL
pheatmap(sampleDistMatrix,
  clustering_distance_rows = sampleDists,
  clustering_distance_cols = sampleDists,
  color = viridis(
    n = 256, alpha = 1,
    begin = 0, end = 1, option = "viridis"
  )
)
```

## Coverage statistics

```{r}
mapping_stats_df
```

## Correlation plots

### TPM {.tabset}

#### All genes

```{r message=FALSE, cache=TRUE}
# subset of samples for quick visualisation tests
# co_genes <- ggpairs(gene_tpm %>% filter(species == "pk"), columns = 3:5, lower = list(
#   continuous = "smooth"
# ), upper = list(continuous = wrap("cor", size = 2)))
# co_genes + theme_bw() + scale_x_continuous(trans = "log1p") + scale_y_continuous(trans = "log1p")


# all samples
co_genes <- ggpairs(gene_tpm %>% filter(species == "pk"), columns = 3:18, lower = list(
  continuous = "smooth"
), upper = list(continuous = wrap("cor", size = 2)))
co_genes + theme_bw() + scale_x_continuous(trans = "log1p") + scale_y_continuous(trans = "log1p")

# ggsave(here("results-refseq/figures/correlation_all_genes.png"))
```

#### Protein-coding genes

```{r message=FALSE}
co_genes <- ggpairs(
  gene_tpm %>%
    filter(species == "pk") %>%
    left_join(gff, by = "gene_id") %>%
    filter(gene_biotype == "protein_coding"),
  columns = 3:18, lower = list(continuous = "smooth"),
  upper = list(continuous = wrap("cor", size = 2))
)
co_genes + theme_bw() + scale_x_continuous(trans = "log1p") + scale_y_continuous(trans = "log1p")

# ggsave(here("results-refseq/figures/correlation_protein_coding_genes.png"))
```

<!-- #### Pk VST genes -->

<!-- ```{r message=FALSE} -->
<!-- co_genes <- ggpairs( -->
<!--   gene_counts_vst_pk %>% -->
<!--     filter(species == "pk") %>% -->
<!--     left_join(gff, by = "gene_id") %>% -->
<!--     filter(gene_biotype == "protein_coding_gene"), -->
<!--   columns = 1:16, lower = list(continuous = "smooth"), -->
<!--   upper = list(continuous = wrap("cor", size = 2)) -->
<!-- ) -->
<!-- co_genes + theme_bw() + scale_x_continuous(trans = "log1p") + scale_y_continuous(trans = "log1p") -->

<!-- ggsave(here("results-refseq/figures/correlation_protein_coding_genes_pk.png")) -->
<!-- ``` -->

## Heatmap of count matrix {.tabset}

**TODO: try quantile breaks for better separation between values instead, see https://slowkow.com/notes/pheatmap-tutorial/** 

### All genes

```{r}
select <- order(rowMeans(counts(gse, normalized = FALSE)),
  decreasing = TRUE
)[1:20]
df <- as.data.frame(colData(gse)[, c("RNA_type", "removal", "WBC_depletion")])
pheatmap(assay(vsd)[select, ],
  cluster_rows = FALSE, show_rownames = FALSE,
  cluster_cols = FALSE, annotation_col = df
)
```

```{r}
df <- as.data.frame(colData(gse)[, c("kit", "WBC_depletion")])
pheatmap(assay(vsd)[select, ],
  cluster_rows = FALSE, show_rownames = FALSE,
  cluster_cols = FALSE, annotation_col = df
)
```

### *P. knowlesi* only

Without clustering:

```{r}
select <- order(rowMeans(counts(gse_pk, normalized = FALSE)),
  decreasing = TRUE
)[1:20]
df <- as.data.frame(colData(gse_pk)[, c("RNA_type", "removal", "WBC_depletion")])
pheatmap(assay(vsd_pk)[select, ],
  cluster_rows = FALSE, show_rownames = FALSE,
  cluster_cols = FALSE, annotation_col = df,
  color = viridis(10)
)
```

With clustering:

```{r}
select <- order(rowMeans(counts(gse_pk, normalized = FALSE)),
  decreasing = TRUE
)[1:20]
df <- as.data.frame(colData(gse_pk)[, c("RNA_type", "removal", "WBC_depletion")])
pheatmap(assay(vsd_pk)[select, ],
  cluster_rows = FALSE, show_rownames = FALSE,
  cluster_cols = TRUE, annotation_col = df,
  color = viridis(10)
)
#
#  ,
# annotation_colors = list(
#   RNA_type = palette.colors(n = length(unique(df$RNA_type)), palette = "Okabe-Ito"),
#   removal = palette.colors(n = length(unique(df$removal)), palette = "Okabe-Ito"),
#   WBC_depletion = palette.colors(n = length(unique(df$WBC_depletion)), palette = "Okabe-Ito")
# )
# )
```

```{r}
df <- as.data.frame(colData(gse_pk)[, c("kit", "WBC_depletion")])
pheatmap(assay(vsd_pk)[select, ],
  cluster_rows = FALSE, show_rownames = FALSE,
  cluster_cols = FALSE, annotation_col = df
)
```

Protein-coding only (not normalized):

```{r}
gene_list <- intersect(rownames(counts(gse_pk)), gff %>%
  filter(gene_biotype == "protein_coding") %>%
  pull(gene_id))

select <- order(rowMeans(counts(gse_pk, normalized = FALSE)[gene_list, ]),
  decreasing = TRUE
)[1:20]
df <- as.data.frame(colData(gse_pk)[, c("RNA_type", "removal", "WBC_depletion")])

pheatmap(assay(vsd_pk)[select, ],
  cluster_rows = FALSE, show_rownames = FALSE,
  cluster_cols = FALSE, annotation_col = df,
  color = viridis(10)
)
```

Protein-coding only + filtered on low counts:

```{r}
# select <- order(rowMeans(counts(gse_pk, normalized = FALSE)[gene_counts_vst_pk_filtered$gene_id,]),
#   decreasing = TRUE
# )[1:20]

gene_list <- intersect(rownames(counts(gse_pk_filtered)), gff %>%
  filter(gene_biotype == "protein_coding") %>%
  pull(gene_id))

select <- order(rowMeans(counts(gse_pk_filtered, normalized = FALSE)[gene_list, ]),
  decreasing = TRUE
)[1:20]
df <- as.data.frame(colData(gse_pk)[, c("RNA_type", "removal", "WBC_depletion")])

pheatmap(assay(vsd_pk_filtered)[select, ],
  cluster_rows = FALSE, show_rownames = FALSE,
  cluster_cols = FALSE, annotation_col = df,
  color = viridis(10)
)
```

## PCA plots {.tabset}

### VSD - all genes

```{r}
pcaData <- plotPCA(vsd, intgroup = c("kit", "WBC_depletion"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit, shape = WBC_depletion)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(title = expression(paste("PCA of VST counts of human and ", italic("P. knowlesi"), " genes")),
       color = "Library preparation kit", shape = "Depletion method") +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

### VSD - all genes - filtered

Filtered on minimum count of 10 reads for at least 1 sample.

```{r}
pcaData <- plotPCA(vsd_filtered, intgroup = c("kit", "WBC_depletion"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit, shape = WBC_depletion)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(title = expression(paste("PCA of VST counts of human and ", italic("P. knowlesi"), " genes")),
       color = "Library preparation kit", shape = "Depletion method") +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

### VSD - Pk genes

```{r}
pcaData <- plotPCA(vsd_pk, intgroup = c("kit", "WBC_depletion"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit, shape = WBC_depletion)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(title = expression(paste("PCA of VST counts of ", italic("P. knowlesi"), " genes")),
       color = "Library preparation kit", shape = "Depletion method") +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

### VSD - Pk genes - filtered

Filtered on minimum count of 10 reads for at least 1 sample.

```{r}
pcaData <- plotPCA(vsd_pk_filtered, intgroup = c("kit", "WBC_depletion"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit, shape = WBC_depletion)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(title = expression(paste("PCA of VST counts of ", italic("P. knowlesi"), " genes")),
       color = "Library preparation kit", shape = "Depletion method") +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

### VSD - Pk genes - filtered - protein coding only

```{r}
protein_gene_filter <- intersect(rownames(rowData(vsd_pk_filtered)), gff %>% filter(gene_biotype == "protein_coding") %>% select(gene_id) %>% pull())

pcaData <- plotPCA(vsd_pk_filtered[protein_gene_filter,], intgroup = c("kit", "WBC_depletion"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit, shape = WBC_depletion)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(title = expression(paste("PCA of VST counts of ", italic("P. knowlesi"), " protein coding genes")),
       color = "Library preparation kit", shape = "Depletion method") +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

----

Number of Pk genes in transcript_info file:

```{r}
dim(transcript_info$gene %>% filter(grepl("PKNH", gene_id)))
```

Number of Pk genes in count matrices:

```{r}
gene_tpm %>% filter(species == "pk")
```

Number of Pk genes in gff:

```{r}
read_delim(here("data/ref-refseq-refseq/GCF_000006355.2_GCA_000006355.2_genomic.gff"), comment = "#", col_names = c("seqname", "source", "feature", "start", "end", "score", "strand", "frame", "attribute")) %>%
  mutate(id = row_number()) %>%
  separate_rows(attribute, sep = ";") %>%
  separate(attribute, c("attribute", "attribute_value"), sep = "=") %>%
  pivot_wider(names_from = attribute, values_from = attribute_value) %>%
  filter(!is.na(gene_biotype))
```

```{r}
read_delim(here("data/ref-refseq-refseq/GCF_000006355.2_GCA_000006355.2_genomic.gff"), comment = "#", col_names = c("seqname", "source", "feature", "start", "end", "score", "strand", "frame", "attribute")) %>%
    mutate(id = row_number()) %>%
    separate_rows(attribute, sep = ";") %>%
    separate(attribute, c("attribute", "attribute_value"), sep = "=") %>%
    pivot_wider(names_from = attribute, values_from = attribute_value) %>% filter(feature == "gene")
```


```{r}
read_delim(here("data/ref-refseq-refseq/GCF_000006355.2_GCA_000006355.2_genomic.gff"), comment = "#", col_names = c("seqname", "source", "feature", "start", "end", "score", "strand", "frame", "attribute")) %>%
    mutate(id = row_number()) %>%
    separate_rows(attribute, sep = ";") %>%
    separate(attribute, c("attribute", "attribute_value"), sep = "=") %>%
    pivot_wider(names_from = attribute, values_from = attribute_value) %>% filter(gene_biotype == "protein_coding")
```

Non-protein coding genes:

```{r}
read_delim(here("data/ref-refseq-refseq/GCF_000006355.2_GCA_000006355.2_genomic.gff"), comment = "#", col_names = c("seqname", "source", "feature", "start", "end", "score", "strand", "frame", "attribute")) %>%
    mutate(id = row_number()) %>%
    separate_rows(attribute, sep = ";") %>%
    separate(attribute, c("attribute", "attribute_value"), sep = "=") %>%
    pivot_wider(names_from = attribute, values_from = attribute_value) %>% filter(!gene_biotype == "protein_coding")
```

```{r}
read_delim(here("data/ref-refseq-refseq/GCF_000006355.2_GCA_000006355.2_genomic.gff"), comment = "#", col_names = c("seqname", "source", "feature", "start", "end", "score", "strand", "frame", "attribute")) %>%
    mutate(id = row_number()) %>%
    separate_rows(attribute, sep = ";") %>%
    separate(attribute, c("attribute", "attribute_value"), sep = "=") %>%
    pivot_wider(names_from = attribute, values_from = attribute_value) %>% filter(gene_biotype == "rRNA")
```
