---
title: "WBC depletion methods - analysis"
output:
  html_document:
    df_print: paged
    toc: true
    toc_float: true
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = here::here())
```

```{r load-libraries, include=FALSE}
# Since some of the required packages are only available in Bioconductor,
# renv needs to be initialised as follows:
# renv::init(bioconductor = "3.19")

library(BiocManager)
library(here)
library(ggplot2)
library(viridis)
library(DESeq2)
library(tximport)
library(txdbmaker)
library(vsn)
library(hexbin)
library(pheatmap)
library(dplyr)
library(tidyr)
library(readr)
library(purrr)
library(stringr)
library(GGally)
library(VennDiagram)
library(UpSetR)
library(scales)
library(ggpubr)
library(ggcorrplot)
library(ggtext)
library(svglite)

ggplot2::theme_set(theme_bw())
```

# Software citations

```{r, include=FALSE}
grateful::cite_packages(output="paragraph", cite.tidyverse = TRUE, out.dir=".")
```

We used R version 4.4.2 [@base] and the following R packages: AnnotationDbi v. 1.66.0 [@AnnotationDbi], BiocManager v. 1.30.23 [@BiocManager], DESeq2 v. 1.44.0 [@DESeq2], GGally v. 2.2.1 [@GGally], ggcorrplot v. 0.1.4.1 [@ggcorrplot], ggpubr v. 0.6.0 [@ggpubr], ggtext v. 0.1.2 [@ggtext], here v. 1.0.1 [@here], hexbin v. 1.28.3 [@hexbin], knitr v. 1.48 [@knitr2014; @knitr2015; @knitr2024], pheatmap v. 1.0.12 [@pheatmap], renv v. 1.0.7 [@renv], rmarkdown v. 2.27 [@rmarkdown2018; @rmarkdown2020; @rmarkdown2024], scales v. 1.3.0 [@scales], svglite v. 2.1.3 [@svglite], tidyverse v. 2.0.0 [@tidyverse], txdbmaker v. 1.0.1 [@txdbmaker], tximport v. 1.32.0 [@tximport], UpSetR v. 1.4.0 [@UpSetR], VennDiagram v. 1.7.3 [@VennDiagram], viridis v. 0.6.5 [@viridis], vsn v. 3.72.0 [@vsn].

# Data import

The RDS objects required for this report can be generated using the data import script.

If the RDS objects have already been created in a previous run, the following cell can be uncommented and run, and the remaining cells in the data import section can be skipped. See [RDS object section](#rds-objects)).

```{r data-import, include=FALSE, message = FALSE}
data_list <- list.files(path = here("results/r_objects"),
                        pattern = ".RDS",
                        recursive = TRUE, full.names = TRUE) %>%
  purrr::set_names(., stringr::str_remove(basename(.), ".RDS")) %>%
  purrr::imap(~ readRDS(file = .x ))

list2env(data_list, .GlobalEnv)
rm(data_list)
```

# Analyses and figures

```{r show=FALSE, message = FALSE}
dir.create(here("results/figures/"), showWarnings = FALSE)
```

## Summary of annotated genes 

Number of genes per species:

```{r}
gene_counts_txi %>%
  group_by(species) %>%
  summarise(n = n()) %>%
  ungroup() %>%
  mutate(total = sum(n), rel.freq = n / total)
```

Number of genes per species after filtering low counts (n < 10 for all samples):

```{r}
gene_counts_txi_filtered %>%
  group_by(species) %>%
  summarise(n = n()) %>%
  ungroup() %>%
  mutate(total = sum(n), rel.freq = n / total)
```

Number of gene types per species:

```{r}
gene_counts_txi %>%
  group_by(species, gene_type) %>%
  summarise(n = n()) %>%
  ungroup() %>%
  mutate(total = sum(n), rel.freq = round(n / total, 5))
```

Number of gene types per species after filtering out low counts (n < 10 for all samples):

```{r}
gene_counts_txi_filtered %>%
  group_by(species, gene_type) %>%
  summarise(n = n()) %>%
  ungroup() %>%
  mutate(total = sum(n), rel.freq = round(n / total, 5))
```

```{r}
n_genes <- dim(gene_counts_txi)[1]
# stopifnot(dim(gene_counts_txi)[1] == (gene_counts_salmon %>% group_by(sample) %>% summarise(n = n()) %>% select(n) %>% unique()))
n_genes_filtered <- dim(gse_filtered)[1]

n_human_genes <- gene_counts_txi %>%
  filter(species == "human") %>%
  count() %>%
  pull()
n_human_genes_filtered <- gene_counts_txi_filtered %>%
  filter(species == "human") %>%
  count() %>%
  pull()

n_pk_genes <- dim(gse_pk)[1]
stopifnot(dim(gse_pk)[1] == (gene_counts_txi %>% filter(species == "pk") %>% count()))
n_pk_genes_filtered <- dim(gse_pk_filtered)[1]
stopifnot(dim(gse_pk_filtered)[1] == (gene_counts_txi_filtered %>% filter(species == "pk") %>% count()))
```

Number of unique genes per sample:

```{r}
gene_counts_txi_filtered %>%
  summarise(across(starts_with("104974"), ~sum(!is.na(.x))))

gene_counts_txi_filtered %>% 
  to_long() %>% 
  group_by(sample) %>% 
  summarize(n = n(),
            n_10 = sum(counts >= 10)
  )
```

Number of unique Pk genes per sample (after filtering):

```{r}
gene_counts_txi_filtered %>%
  filter(species=="pk") %>% 
  summarise(across(starts_with("104974"), ~sum(!is.na(.x))))

gene_counts_txi_filtered %>% 
  filter(species=="pk") %>% 
  to_long() %>% 
  group_by(sample) %>% 
  summarize(n = n(),
            n_10 = sum(counts >= 10)
  )
```

### Gene counts

Number of gene counts per sample:

```{r}
colSums(gene_counts_txi %>% select(starts_with("104974")))
```

```{r, fig.height=6}
gene_counts_txi %>% 
  summarise(across(`104974-001-001_S1_L001`:`104974-001-016_S1_L001`, sum)) %>% 
  to_long() %>% 
  left_join(samples, by = "sample") %>%
  ggplot(aes(x = sample_name, y = counts)) +
    geom_bar(stat = "identity") +
    scale_fill_viridis(option = "viridis", discrete = T) +
    scale_y_continuous(labels = scales::label_number(scale_cut = cut_short_scale())) +
    labs(x = "Sample", y = "Total gene counts") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
```

Number of Pk gene counts per sample:

```{r}
colSums(gene_counts_txi %>% filter(species == "pk") %>% select(starts_with("104974")))
```

```{r, fig.height=6}
gene_counts_txi %>% filter(species == "pk") %>% 
  summarise(across(`104974-001-001_S1_L001`:`104974-001-016_S1_L001`, sum)) %>% 
  to_long() %>% 
  left_join(samples, by = "sample") %>%
  ggplot(aes(x = sample_name, y = counts)) +
    geom_bar(stat = "identity") +
    scale_fill_viridis(option = "viridis", discrete = T) +
    scale_y_continuous(labels = scales::label_number(scale_cut = cut_short_scale())) +
    labs(x = "Sample", y = "Pk gene counts") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
```

## Comparison of total RNA levels

> The direct comparison of RPKM and TPM across samples is meaningful only when there are equal total RNAs between compared samples and the distribution of RNA populations are close to each other. 

> Make sure both samples use the same RNA isolation approach [poly(A)+ selection versus ribosomal RNA depletion]. If not, they should not be compared.

> Check the fraction of the ribosomal, mitochondrial and globin RNAs, and the top highly expressed transcripts and see whether such RNAs constitute a very large part of the sequenced reads in a sample, and thus decrease the sequencing “real estate” available for the remaining genes in that sample. If the calculated fractions in two samples differ significantly, do not compare RPKM or TPM values directly.

> TPM should never be used for quantitative comparisons across samples when the total RNA contents and its distributions are very different. However, under appropriate circumstances, TPM can still be useful for qualitative comparison such as PCA and clustering analysis. 

(Zhao et al., 2020 - http://dx.doi.org/10.1261/rna.074922.120)

Looking at our samples, especially the ones prepared using the Illumina kit exhibit much higher total RNA content. Based on the output of Picard MarkDuplicates, the differences can at least partially be attributed to the number of PCR duplicates. The proportion of duplicates is comparable across kits, but they do appear to be consistently higher for some depletion methods (no filter > centpip > plasmo > pmacs > cellulose).

```{r}
gene_counts_txi_filtered %>%
  summarise(across(`104974-001-001_S1_L001`:`104974-001-016_S1_L001`, sum)) %>%
  to_long() %>%
  left_join(samples, by = "sample") %>%
  ggplot(aes(x = sample_short, y = counts, fill = kit_name)) +
  geom_bar(stat = "identity") +
  scale_fill_viridis(option = "viridis", discrete = T) +
  # labels = c("mRNA Illumina", "mRNA Qiagen globin depletion", "mRNA Qiagen globin/rRNA depletion", "tRNA Qiagen globin/rRNA depletion")) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  scale_y_continuous(labels = scales::label_number(scale_cut = cut_short_scale())) +
  labs(x = "Sample", y = "Number of gene-level counts", fill = "Kit")
```

```{r, fig.width=10, out.width="100%"}
pcr_dups %>%
  left_join(samples, by = join_by(Sample == sample)) %>%
  mutate(READS_IN_DUPLICATE_PAIRS = 2 * READ_PAIR_DUPLICATES) %>%
  mutate(READS_IN_UNIQUE_PAIRS = 2 * (READ_PAIRS_EXAMINED - READ_PAIR_DUPLICATES)) %>%
  mutate(READS_IN_UNIQUE_UNPAIRED = UNPAIRED_READS_EXAMINED - UNPAIRED_READ_DUPLICATES) %>%
  mutate(READS_IN_DUPLICATE_PAIRS_OPTICAL = 2 * READ_PAIR_OPTICAL_DUPLICATES) %>%
  mutate(READS_IN_DUPLICATE_PAIRS_NONOPTICAL = READS_IN_DUPLICATE_PAIRS - READS_IN_DUPLICATE_PAIRS_OPTICAL) %>%
  mutate(READS_IN_DUPLICATE_UNPAIRED = UNPAIRED_READ_DUPLICATES) %>%
  mutate(READS_UNMAPPED = UNMAPPED_READS) %>%
  select(c(sample_name, condition, kit, READS_IN_UNIQUE_PAIRS, READS_IN_UNIQUE_UNPAIRED, READS_IN_DUPLICATE_PAIRS_OPTICAL, READS_IN_DUPLICATE_PAIRS_NONOPTICAL, READS_IN_DUPLICATE_UNPAIRED)) %>%
  pivot_longer(cols = 4:8, names_to = "type", values_to = "counts") %>%
  ggplot(aes(x = sample_name, y = counts, fill = type)) +
  geom_bar(position = "stack", stat = "identity") +
  scale_fill_viridis(option = "viridis", discrete = T) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  coord_flip() +
  scale_x_discrete(limits = rev) +
  scale_y_continuous(labels = scales::label_number(scale_cut = cut_short_scale())) +
  labs(x = "Sample", y = "Number of reads", fill = "Picard deduplication stats")
```

## Alignment statistics {.tabset}

The output reported by `samtool flagstats` and STAR's `Log.final.out` files cannot be compared directly: the latter reports counts of read-pairs as opposed to individual reads, and the former does not contain unmapped reads (these are not added to the `.bam` outputs by default) while also counting multi-mapped reads. We opted to report the STAR statistics, but the flagstat output is included for completion's sake. 

### STAR alignment statistics

```{r}
star_stats
```

```{r, fig.height=6, out.width="100%"}
star_stats %>%
  pivot_longer(cols = 2:6, names_to = "type", values_to = "counts") %>%
  left_join(samples, by = join_by(Sample == sample)) %>%
  ggplot(aes(x = sample_name, y = counts, fill = type)) +
  geom_bar(position = "stack", stat = "identity") +
  scale_fill_viridis(option = "viridis", discrete = T) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  coord_flip() +
  scale_x_discrete(limits = rev) +
  scale_y_continuous(labels = scales::label_number(scale_cut = cut_short_scale())) +
  labs(x = "Sample", y = "Number of read pairs", fill = "Mapping stats")
```

### samtools flagstat

```{r}
mapping_stats_df
```

```{r, fig.height=6, out.width="100%"}
mapping_stats_df %>%
  left_join(samples, by = "sample") %>%
  pivot_longer(cols = flagstat_primary_pk:flagstat_primary_human, names_to = "type", values_to = "counts") %>%
  ggplot(aes(x = sample_name, y = counts, fill = type)) +
  geom_bar(position = "stack", stat = "identity") +
  scale_fill_viridis(option = "viridis", discrete = T) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  coord_flip() +
  scale_x_discrete(limits = rev) +
  scale_y_continuous(labels = scales::label_number(scale_cut = cut_short_scale())) +
  labs(x = "Sample", y = "Number of primary reads", fill = "Samtools flagstat")
```

## Top expressed reads in TPM {.tabset}

```{r,results='asis', echo=FALSE}
for (i in unique(gene_tpm_filtered_long$sample)) {
  cat("### Sample ", i, " \n")
  print(knitr::kable(
    gene_tpm_long %>%
      filter(sample == i) %>%
      top_n(25, counts) %>%
      arrange(sample, desc(counts), .by_group = T)
  ))
  cat("\n")
}
```

## FastQ Screen statistics {.tabset}

Also see figure in [results/fastq-screen/](results/fastq-screen/) and [results/multiqc-fastq-screen-samtools-pk/multiqc_report.html](results/multiqc-fastq-screen-samtools-pk/multiqc_report.html).

### All reads for both species

```{r}
fastq_screen_transformed <- bind_rows(
  fastq_screen %>%
    mutate(`%_single_genome` = `%_One_hit_one_genome` + `%_Multiple_hits_one_genome`) %>%
    mutate(`%_multiple_genomes` = `%_One_hit_multiple_genomes` + `%_Multiple_hits_multiple_genomes`) %>%
    mutate(`No_hits` = `%_One_hit_multiple_genomes` + `%_Multiple_hits_multiple_genomes`) %>%
    select(c(sample, Genome, `#Reads_processed`, `%_single_genome`, `%_multiple_genomes`)) %>%
    pivot_wider(names_from = Genome, values_from = c(`%_single_genome`, `%_multiple_genomes`)) %>%
    select(!`%_multiple_genomes_PknowlesiH`) %>%
    rename(`%_multiple_genomes` = `%_multiple_genomes_Human`) %>%
    mutate(pct = 1 - `%_single_genome_PknowlesiH` - `%_single_genome_Human` - `%_multiple_genomes`) %>%
    select(sample, pct) %>%
    mutate("Mapped reads" = "No hits"),
  fastq_screen %>%
    mutate(pct = `%_One_hit_one_genome` + `%_Multiple_hits_one_genome`) %>%
    select(c(sample, Genome, pct)) %>%
    rename("Mapped reads" = Genome),
  fastq_screen %>%
    mutate(pct = `%_One_hit_multiple_genomes` + `%_Multiple_hits_multiple_genomes`) %>%
    select(c(sample, pct)) %>%
    distinct(sample, .keep_all = TRUE) %>%
    mutate("Mapped reads" = "Multiple genomes")
) %>%
  # change order of levels here to adjust colours
  mutate(`Mapped reads` = factor(`Mapped reads`,
    levels = c("PknowlesiH", "Human", "Multiple genomes", "No hits"),
    labels = c("P. knowlesi", "Human", "Multiple hits", "No hits")
  )) %>%
  left_join(samples, by = "sample")

ggplot(fastq_screen_transformed, aes(x = sample_short, y = pct, fill = `Mapped reads`)) +
  geom_bar(position = "stack", stat = "identity") +
  scale_fill_viridis(option = "viridis", discrete = T, limits = rev) +
  coord_flip() +
  scale_x_discrete(limits = rev) +
  labs(fill = expression("FastQ Screen mapped reads"), x = "Sample") +
  ylab(label = "% of mapped reads")
```

### Human reads only

```{r}
# Human reads
fastq_screen %>%
  pivot_longer(
    cols = starts_with("%"),
    names_to = "fastq_screen_pct_type",
    values_to = "fastq_screen_pct_value"
  ) %>%
  filter(Genome == "Human") %>%
  left_join(samples) %>%
  ggplot(aes(x = sample_short, y = fastq_screen_pct_value, fill = fastq_screen_pct_type)) +
  geom_bar(position = "stack", stat = "identity") +
  scale_fill_viridis(option = "viridis", discrete = T) +
  coord_flip() +
  scale_x_discrete(limits = rev) +
  labs(fill = "FastQ screen statistics for human reads", x = "Sample", y = "% of mapped reads")
```

### Pk reads

```{r}
# Pk reads
fastq_screen %>%
  pivot_longer(
    cols = starts_with("%"),
    names_to = "fastq_screen_pct_type",
    values_to = "fastq_screen_pct_value"
  ) %>%
  filter(Genome == "PknowlesiH") %>%
  left_join(samples) %>%
  ggplot(aes(x = sample_short, y = fastq_screen_pct_value, fill = fastq_screen_pct_type)) +
  geom_bar(position = "stack", stat = "identity") +
  scale_fill_viridis(option = "viridis", discrete = T) +
  coord_flip() +
  scale_x_discrete(limits = rev) +
  labs(fill = expression("FastQ screen statistics for" ~ italic("P. knowlesi") ~ "reads"), x = "Sample", y = "% of mapped reads")
```

## rRNA and globin levels {.tabset}

### Barplot of TPM for rRNA genes per condition

```{r, fig.height=14, out.width="100%"}
gene_annotation %>%
  # select rRNA and globins
  filter(!gene_type %in% c("human", "pk")) %>%
  left_join(gene_tpm_filtered[1:18], by = c("gene_id", "gene_name")) %>%
  filter(rowSums(across(.cols = `104974-001-001_S1_L001`:`104974-001-016_S1_L001`)) != 0) %>%
  pivot_longer(
    cols = `104974-001-001_S1_L001`:`104974-001-016_S1_L001`,
    names_to = "sample",
    values_to = "counts"
  ) %>%
  left_join(samples, by = "sample") %>%
  group_by(gene_id) %>%
  mutate(total = sum(counts)) %>%
  ungroup() %>%
  arrange(species, gene_type, desc(total)) %>%
  mutate(gene_name = factor(gene_name, levels = unique(gene_name))) %>%
  ggplot(aes(x = gene_name, y = counts, fill = sample_short)) +
  geom_bar(position = "stack", stat = "identity") +
  scale_fill_viridis(option = "viridis", discrete = T) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  coord_flip() +
  scale_x_discrete(limits = rev) +
  labs(fill = "Sample") +
  ylab(label = "TPM") +
  xlab(label = "Gene name") +
  facet_wrap(~kit_name) +
  scale_y_continuous(labels = scales::label_number(scale_cut = cut_short_scale()))
```

### Dot plot of log(TPM+1) for rRNA genes per condition

Log TPM is often plotted instead of raw values, but can be misleading on barplot...

```{r, fig.height=12, out.width="100%"}
gene_annotation %>%
  filter(!gene_type %in% c("human", "pk")) %>%
  left_join(gene_tpm_filtered[1:18], by = c("gene_id", "gene_name")) %>%
  filter(rowSums(across(.cols = `104974-001-001_S1_L001`:`104974-001-016_S1_L001`)) != 0) %>%
  pivot_longer(
    cols = `104974-001-001_S1_L001`:`104974-001-016_S1_L001`,
    names_to = "sample",
    values_to = "counts"
  ) %>%
  left_join(samples, by = "sample") %>%
  group_by(gene_id) %>%
  mutate(total = sum(counts)) %>%
  ungroup() %>%
  arrange(species, gene_type, desc(total)) %>%
  mutate(gene_name = factor(gene_name, levels = unique(gene_name))) %>%
  ggplot(aes(x = gene_name, y = counts)) +
  geom_point(aes(colour = sample_name)) +
  scale_colour_viridis(option = "viridis", discrete = T) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  coord_flip() +
  scale_x_discrete(limits = rev) +
  # labs(colour = "Sample") +
  labs(x = "Gene name", y = "log(TPM+1)", colour = "Sample") +
  facet_wrap(~sample_name) +
  scale_y_continuous(trans = "log1p")
```

### Globin only

```{r, out.height="130%"}
gene_annotation %>%
  filter(gene_type == "human_globin") %>%
  left_join(gene_tpm_filtered[1:18], by = c("gene_id", "gene_name")) %>%
  filter(rowSums(across(.cols = `104974-001-001_S1_L001`:`104974-001-016_S1_L001`)) != 0) %>%
  pivot_longer(
    cols = `104974-001-001_S1_L001`:`104974-001-016_S1_L001`,
    names_to = "sample",
    values_to = "counts"
  ) %>%
  left_join(samples, by = "sample") %>%
  group_by(gene_id) %>%
  mutate(total = sum(counts)) %>%
  ungroup() %>%
  arrange(species, gene_type, desc(total)) %>%
  mutate(gene_name = factor(gene_name, levels = unique(gene_name))) %>%
  ggplot(aes(x = gene_name, y = counts)) +
  geom_point(aes(colour = sample)) +
  scale_colour_viridis(option = "viridis", discrete = T) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  coord_flip() +
  scale_x_discrete(limits = rev) +
  labs(x = "Gene name", y = "TPM", colour = "Sample") +
  facet_wrap(~sample_name) +
  scale_y_continuous(labels = scales::label_number(scale_cut = cut_short_scale()))
```

### Human rRNA only

```{r, fig.height=12, out.width="100%"}
gene_annotation %>%
  filter(gene_type == "human_rRNA") %>%
  left_join(gene_tpm_filtered[1:18], by = c("gene_id", "gene_name")) %>%
  filter(rowSums(across(.cols = `104974-001-001_S1_L001`:`104974-001-016_S1_L001`)) != 0) %>%
  pivot_longer(
    cols = `104974-001-001_S1_L001`:`104974-001-016_S1_L001`,
    names_to = "sample",
    values_to = "counts"
  ) %>%
  left_join(samples, by = "sample") %>%
  group_by(gene_id) %>%
  mutate(total = sum(counts)) %>%
  ungroup() %>%
  arrange(species, gene_type, desc(total)) %>%
  mutate(gene_name = factor(gene_name, levels = unique(gene_name))) %>%
  ggplot(aes(x = gene_name, y = counts)) +
  geom_point(aes(colour = sample)) +
  scale_colour_viridis(option = "viridis", discrete = T) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  coord_flip() +
  scale_x_discrete(limits = rev) +
  labs(x = "Gene name", y = "TPM", colour = "Sample") +
  facet_wrap(~sample_name) +
  scale_y_continuous(trans = "log1p")
```

## Plot percentage of reads for each gene type (rRNA/globin) and species {.tabset}

**TPM values appear to be the most appropriate metric to use.**

### Gene-level counts

These gene counts were aggregated across all samples (tximport of transcript counts) and filtered to only genes with a minimum count of 10 reads for at least 1 sample.

```{r, message=FALSE}
gene_counts_txi_filtered_long %>%
  left_join(samples, by = "sample") %>%
  group_by(sample_name, gene_type) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads)) %>%
  ggplot(aes(y = sum_reads, x = sample_name, fill = gene_type)) +
  geom_bar(position = "stack", stat = "identity") +
  scale_fill_viridis(
    option = "viridis", discrete = T,
    # limits = rev,
    labels = c("human", "human globin", "human rRNA", expression(italic("P. knowlesi")), expression(italic("P. knowlesi rRNA")))
  ) +
  coord_flip() +
  scale_x_discrete(limits = rev) +
  scale_y_continuous(labels = scales::label_number(scale_cut = cut_short_scale())) +
  labs(x = "Sample", y = "Sum of gene counts", fill = "RNA type")
```

### Gene-level TPM

Filtered to only genes with a minimum count of 10 reads for at least 1 sample.

```{r, message=FALSE}
gene_tpm_filtered_long %>%
  left_join(samples, by = "sample") %>%
  group_by(sample_name, gene_type) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads)) %>%
  ggplot(aes(y = sum_reads, x = sample_name, fill = gene_type)) +
  geom_bar(position = "stack", stat = "identity") +
  scale_fill_viridis(
    option = "viridis", discrete = T,
    # limits = rev,
    labels = c("human", "human globin", "human rRNA", expression(italic("P. knowlesi")), expression(italic("P. knowlesi rRNA")))
  ) +
  coord_flip() +
  scale_x_discrete(limits = rev) +
  scale_y_continuous(labels = scales::label_number(scale_cut = cut_short_scale())) +
  labs(x = "Sample", y = "Sum of TPM counts", fill = "RNA type")
```


All gene biotypes: 

```{r, message=FALSE}
gene_tpm_filtered_long %>%
  left_join(samples, by = "sample") %>%
  mutate(gene_type = case_when(species == "pk" ~ gene_biotype, TRUE ~ gene_type)) %>%
  group_by(sample_name, gene_type) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads)) %>%
  ggplot(aes(y = sum_reads, x = sample_name, fill = gene_type)) +
  geom_bar(position = "stack", stat = "identity") +
  # scale_fill_viridis(
    # option = "viridis", discrete = T,
    # limits = rev,
    # labels = c("human", "human globin", "human rRNA", expression(italic("P. knowlesi")), expression(italic("P. knowlesi rRNA")))
  # ) +
  coord_flip() +
  scale_x_discrete(limits = rev) +
  scale_y_continuous(labels = scales::label_number(scale_cut = cut_short_scale())) +
  labs(x = "Sample", y = "Sum of TPM counts", fill = "RNA type")
```

## Plot percentage of reads for each species {.tabset}

### Gene-level counts

These gene counts were aggregated across all samples (tximport of transcript counts) and filtered to only genes with a minimum count of 10 reads for at least 1 sample.

```{r, message=FALSE}
# absolute numbers
gene_counts_txi_filtered_long %>%
  left_join(samples, by = "sample") %>%
  group_by(sample_name, species) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads)) %>%
  ggplot(aes(y = sum_reads, x = sample_name, fill = species)) +
  geom_bar(position = "stack", stat = "identity") +
  scale_fill_viridis(
    option = "viridis", discrete = T,
    # limits = rev
  ) +
  coord_flip() +
  scale_x_discrete(limits = rev) +
  scale_y_continuous(labels = scales::label_number(scale_cut = cut_short_scale())) +
  labs(x = "Sample", y = "Sum of gene counts", fill = "Species")

# proportion
gene_counts_txi_filtered_long %>%
  left_join(samples, by = "sample") %>%
  group_by(sample_name, species) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads)) %>%
  ggplot(aes(y = sum_reads, x = sample_name, fill = species)) +
  geom_bar(position = "fill", stat = "identity") +
  scale_fill_viridis(
    option = "viridis", discrete = T,
    # limits = rev
  ) +
  coord_flip() +
  scale_x_discrete(limits = rev) +
  scale_y_continuous(labels = scales::label_number(scale_cut = cut_short_scale())) +
  labs(x = "Sample", y = "Proportion of gene counts", fill = "Species")
```

### Gene-level TPM

Note that since TPM is already scaled by library size, the absolute and proportion plots are identical.

```{r, message=FALSE}
# absolute numbers
gene_tpm_filtered_long %>%
  left_join(samples, by = "sample") %>%
  group_by(sample_name, species) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads)) %>%
  ggplot(aes(y = sum_reads, x = sample_name, fill = species)) +
  geom_bar(position = "stack", stat = "identity") +
  scale_fill_viridis(
    option = "viridis", discrete = T,
    # limits = rev
  ) +
  coord_flip() +
  scale_x_discrete(limits = rev) +
  scale_y_continuous(labels = scales::label_number(scale_cut = cut_short_scale())) +
  labs(x = "Sample", y = "Sum of TPM counts", fill = "Species")

# proportion
gene_tpm_long %>%
  left_join(samples, by = "sample") %>%
  group_by(sample_name, species) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads)) %>%
  ggplot(aes(y = sum_reads, x = sample_name, fill = species)) +
  geom_bar(position = "fill", stat = "identity") +
  scale_fill_viridis(
    option = "viridis", discrete = T,
    # limits = rev
  ) +
  coord_flip() +
  scale_x_discrete(limits = rev) +
  scale_y_continuous(labels = scales::label_number(scale_cut = cut_short_scale())) +
  labs(x = "Sample", y = "Proportion of TPM counts", fill = "Species")
```

<!-- ## Pk specific gene type plots -->

<!-- Try to use https://nf-co.re/rnaseq/1.1/docs/output#featurecounts? -->

<!-- ``` -->
<!-- └─▶ featureCounts -t "transcript" -g "gene_biotype" -B -C -p -s 1 -T 6 -a data/ref/GCF_000006355.2_GCA_000006355.2_genomic.gtf -o results/featureCounts-pk.txt results/star_salmon/104974-001-001_S1_L001.markdup.sorted.pk.bam -->
<!-- ``` -->

## Gene recovery {.tabset}

```{r venn-prep}
# create binary matrices with presence/absence of genes
gene_tpm_filtered_binary <- gene_tpm_filtered %>%
  mutate(across(.cols = `104974-001-001_S1_L001`:`104974-001-016_S1_L001`, .fns = ~ .x > 0)) %>%
  select(!c(gene_name, gene_type, species))

gene_tpm_pk_filtered_binary <- gene_tpm_filtered %>%
  mutate(across(.cols = `104974-001-001_S1_L001`:`104974-001-016_S1_L001`, .fns = ~ .x > 0)) %>%
  filter(species == "pk") %>%
  select(!c(gene_name, gene_type, species))

gene_tpm_filtered_binary_list <- purrr::set_names(colnames(gene_tpm_filtered_binary %>% select(`104974-001-001_S1_L001`:`104974-001-016_S1_L001`))) %>%
  purrr::imap(~ gene_tpm_filtered_binary$gene_id[gene_tpm_filtered_binary[, .x, drop = T]])

gene_tpm_pk_filtered_binary_list <- purrr::set_names(colnames(gene_tpm_pk_filtered_binary %>% select(`104974-001-001_S1_L001`:`104974-001-016_S1_L001`))) %>%
  purrr::imap(~ gene_tpm_pk_filtered_binary$gene_id[gene_tpm_pk_filtered_binary[, .x, drop = T]])

comparison_centpip <- samples %>%
  filter(WBC_depletion == "centpip") %>%
  select(sample)

comparison_plasmo <- samples %>%
  filter(WBC_depletion == "plasmo") %>%
  select(sample)

comparison_pmacs <- samples %>%
  filter(WBC_depletion == "pmacs") %>%
  select(sample)

comparison_cellulose <- samples %>%
  filter(WBC_depletion == "cellulose") %>%
  select(sample)

comparison_nofilter <- samples %>%
  filter(WBC_depletion == "nofilter") %>%
  filter(!sample == "104974-001-006_S1_L001")

comparison_mrna_qiagen <- samples %>%
  filter(kit == "mRNA_qiagen_FSglob")

comparison_trna_qiagen <- samples %>%
  filter(kit == "tRNA_qiagen_FSglobH")

comparison_mrna_illumina <- samples %>%
  filter(kit == "mRNA_illumina")

draw_venn <- function(set_list, names, title, output) {
  VennDiagram::venn.diagram(
    x = set_list,
    category.names = names,
    # Circles
    lwd = 2,
    # lty = "blank",
    # col = palette.colors(3),
    # fill = palette.colors(3),
    # col = hcl.colors(3, palette = "viridis", alpha = 0.9),
    # fill = hcl.colors(3, palette = "viridis", alpha = 0.4),
    # col = hcl.colors(3, palette = "Set 2", alpha = 0.9),
    # fill = hcl.colors(3, palette = "Set 2", alpha = 0.4),
    col = hcl.colors(3, palette = "viridis", alpha = 0.9),
    fill = hcl.colors(3, palette = "viridis", alpha = 0.4),
    
    # Numbers
    cex = .9,
    fontface = "bold",
    fontfamily = "sans",

    # Set names
    cat.cex = 0.9,
    cat.fontface = "bold",
    cat.default.pos = "outer",
    cat.pos = c(-20, 20, 135),
    cat.dist = c(0.055, 0.055, 0.11),
    cat.fontfamily = "sans",
    rotation = 1,

    # title
    main = title,
    main.cex = 1.1,
    main.fontface = "bold",
    main.fontfamily = "sans",

    # output features
    height = 1200,
    width = 1200,
    resolution = 300,
    compression = "lzw",
    filename = output,
    output = TRUE,
    disable.logging = TRUE
  )
}
```

### All genes - kits per filter

Gene counts were filtered on a minimum (raw) count of 10 for at least 1 sample before checking for presence/absence. 

```{r venn-kits-per-filter, results='hide'}
v <- draw_venn(
  gene_tpm_filtered_binary_list[comparison_centpip$sample],
  c("mRNA Qiagen FS globin", "tRNA Qiagen FS globin/rRNA", "mRNA Illumina"),
  "Centpip",
  NULL
  # here("results/figures/venn-centpip.png")
)
grid.newpage()
grid.draw(v)

v <- draw_venn(
  gene_tpm_filtered_binary_list[comparison_plasmo$sample],
  c("mRNA Qiagen FS globin", "tRNA Qiagen FS globin/rRNA", "mRNA Illumina"),
  "Plasmo",
  NULL
  # here("results/figures/venn-plasmo.png")
)
grid.newpage()
grid.draw(v)

v <- draw_venn(
  gene_tpm_filtered_binary_list[comparison_pmacs$sample],
  c("mRNA Qiagen FS globin", "tRNA Qiagen FS globin/rRNA", "mRNA Illumina"),
  "PMACS",
  NULL
  # here("results/figures/venn-pmacs.png")
)
grid.newpage()
grid.draw(v)

v <- draw_venn(
  gene_tpm_filtered_binary_list[comparison_cellulose$sample],
  c("mRNA Qiagen FS globin", "tRNA Qiagen FS globin/rRNA", "mRNA Illumina"),
  "Cellulose",
  NULL
  # here("results/figures/venn-cellulose.png")
)
grid.newpage()
grid.draw(v)

v <- draw_venn(
  gene_tpm_filtered_binary_list[comparison_nofilter$sample],
  c("mRNA Qiagen FS globin", "tRNA Qiagen FS globin/rRNA", "mRNA Illumina"),
  "Cellulose",
  NULL
  # here("results/figures/venn-nofilter.png")
)
grid.newpage()
grid.draw(v)
```

### *P. knowlesi* genes - kits per filter

Gene counts were filtered on a minimum (raw) count of 10 for at least 1 sample before checking for presence/absence. 

```{r venn-kits-per-filter-pk, results='hide'}
v <- draw_venn(
  gene_tpm_pk_filtered_binary_list[comparison_centpip$sample],
  c("mRNA Qiagen FS globin", "tRNA Qiagen FS globin/rRNA", "mRNA Illumina"),
  "Centpip",
  NULL
  # here("results/figures/venn-centpip_pk.png")
)
grid.newpage()
grid.draw(v)

v <- draw_venn(
  gene_tpm_pk_filtered_binary_list[comparison_plasmo$sample],
  c("mRNA Qiagen FS globin", "tRNA Qiagen FS globin/rRNA", "mRNA Illumina"),
  "Plasmo",
  NULL
  # here("results/figures/venn-plasmo_pk.png")
)
grid.newpage()
grid.draw(v)

v <- draw_venn(
  gene_tpm_pk_filtered_binary_list[comparison_pmacs$sample],
  c("mRNA Qiagen FS globin", "tRNA Qiagen FS globin/rRNA", "mRNA Illumina"),
  "PMACS",
  NULL
  # here("results/figures/venn-pmacs_pk.png")
)
grid.newpage()
grid.draw(v)

v <- draw_venn(
  gene_tpm_pk_filtered_binary_list[comparison_cellulose$sample],
  c("mRNA Qiagen FS globin", "tRNA Qiagen FS globin/rRNA", "mRNA Illumina"),
  "Cellulose",
  NULL
  # here("results/figures/venn-cellulose_pk.png")
)
grid.newpage()
grid.draw(v)

v <- draw_venn(
  gene_tpm_pk_filtered_binary_list[comparison_nofilter$sample],
  c("mRNA Qiagen FS globin", "tRNA Qiagen FS globin/rRNA", "mRNA Illumina"),
  "No filter",
  NULL
  # here("results/figures/venn-nofilter_pk.png")
)
grid.newpage()
grid.draw(v)
```

### All genes - filters per kit

```{r upset-all}
# Use upset plots instead of five-way venn diagram (alternatively check out proportional area venn)

# mRNA Qiagen
UpSetR::upset(
  as.data.frame(gene_tpm_filtered_binary %>%
    select((comparison_mrna_qiagen %>% select(sample) %>% pull())) %>%
    rename(all_of(c(
      "no filter" = "104974-001-001_S1_L001",
      "cent/pip" = "104974-001-002_S1_L001",
      "plasmodipur" = "104974-001-003_S1_L001",
      "PMACS" = "104974-001-004_S1_L001",
      "cellulose" = "104974-001-005_S1_L001"
    ))) %>%
    mutate_all(as.integer)),
  keep.order = TRUE,
  sets.x.label = "Genes per filter",
  mainbar.y.label = "Intersections for mRNA Qiagen",
  order.by = "freq"
)

# tRNA Qiagen
UpSetR::upset(
  as.data.frame(gene_tpm_filtered_binary %>%
    select((comparison_trna_qiagen %>% select(sample) %>% pull())) %>%
    rename(all_of(c(
      "no filter" = "104974-001-007_S1_L001",
      "cent/pip" = "104974-001-008_S1_L001",
      "plasmodipur" = "104974-001-009_S1_L001",
      "PMACS" = "104974-001-010_S1_L001",
      "cellulose" = "104974-001-011_S1_L001"
    ))) %>%
    mutate_all(as.integer)),
  keep.order = TRUE,
  sets.x.label = "Genes per filter",
  mainbar.y.label = "Intersections for tRNA Qiagen",
  order.by = "freq"
)

# mRNA Illumina
UpSetR::upset(
  as.data.frame(gene_tpm_filtered_binary %>%
    select((comparison_mrna_illumina %>% select(sample) %>% pull())) %>%
    rename(all_of(c(
      "no filter" = "104974-001-012_S1_L001",
      "cent/pip" = "104974-001-013_S1_L001",
      "plasmodipur" = "104974-001-014_S1_L001",
      "PMACS" = "104974-001-015_S1_L001",
      "cellulose" = "104974-001-016_S1_L001"
    ))) %>%
    mutate_all(as.integer)),
  keep.order = TRUE,
  sets.x.label = "Genes per filter",
  mainbar.y.label = "Intersections for mRNA Illumina",
  order.by = "freq"
)
```

### Pk genes - filters per kit

```{r upset-pk}
# mRNA Qiagen
UpSetR::upset(
  as.data.frame(gene_tpm_pk_filtered_binary %>%
    select((comparison_mrna_qiagen %>% select(sample) %>% pull())) %>%
    rename(all_of(c(
      "no filter" = "104974-001-001_S1_L001",
      "cent/pip" = "104974-001-002_S1_L001",
      "plasmodipur" = "104974-001-003_S1_L001",
      "PMACS" = "104974-001-004_S1_L001",
      "cellulose" = "104974-001-005_S1_L001"
    ))) %>%
    mutate_all(as.integer)),
  keep.order = TRUE,
  sets.x.label = "Genes per filter",
  mainbar.y.label = "Intersections for mRNA Qiagen",
  order.by = "freq"
)

# tRNA Qiagen
UpSetR::upset(
  as.data.frame(gene_tpm_pk_filtered_binary %>%
    select((comparison_trna_qiagen %>% select(sample) %>% pull())) %>%
    rename(all_of(c(
      "no filter" = "104974-001-007_S1_L001",
      "cent/pip" = "104974-001-008_S1_L001",
      "plasmodipur" = "104974-001-009_S1_L001",
      "PMACS" = "104974-001-010_S1_L001",
      "cellulose" = "104974-001-011_S1_L001"
    ))) %>%
    mutate_all(as.integer)),
  keep.order = TRUE,
  sets.x.label = "Genes per filter",
  mainbar.y.label = "Intersections for tRNA Qiagen",
  order.by = "freq"
)

# mRNA Illumina
UpSetR::upset(
  as.data.frame(gene_tpm_pk_filtered_binary %>%
    select((comparison_mrna_illumina %>% select(sample) %>% pull())) %>%
    rename(all_of(c(
      "no filter" = "104974-001-012_S1_L001",
      "cent/pip" = "104974-001-013_S1_L001",
      "plasmodipur" = "104974-001-014_S1_L001",
      "PMACS" = "104974-001-015_S1_L001",
      "cellulose" = "104974-001-016_S1_L001"
    ))) %>%
    mutate_all(as.integer)),
  keep.order = TRUE,
  sets.x.label = "Genes per filter",
  mainbar.y.label = "Intersections for mRNA Illumina",
  order.by = "freq"
)
```

## Sample-to-sample distances {.tabset}

Which samples are most similar to each other in terms of gene expression profile? Use Euclidean distance between samples on `rlog` transformed gene counts (filtered to genes with a minimum of 10 reads per sample).

```{r}
heatmap_of_distances <- function(transformed_data, title) {
  
  sampleDists <- dist(t(assay(transformed_data)))
  sampleDistMatrix <- as.matrix(sampleDists)
  rownames(sampleDistMatrix) <- transformed_data$sample_short
  colnames(sampleDistMatrix) <- NULL
  
    
  # add grouping columns
  anno <- as.data.frame(colData(transformed_data)[, c("kit_name", "filter_name")]) %>%
    mutate(kit_name = as.factor(kit_name)) %>%
    mutate(filter_name = as.factor(filter_name)) %>%
    rename_with(~ str_to_title(str_replace(.x, "_", " ")))
  
  # set names of samples
  colnames(sampleDistMatrix) <- (transformed_data)$sample_short
  rownames(anno) <- (transformed_data)$sample_short
  
  # Adapted from: https://slowkow.com/notes/pheatmap-tutorial/
  mat_colors <- list(
    `Kit Name` = hcl.colors(length(unique(anno$`Kit Name`)), palette = "plasma", alpha = 1),
    `Filter Name` = palette.colors(length(unique(anno$`Filter Name`)), palette = "Okabe-Ito", alpha = 1)
  )
  names(mat_colors$`Kit Name`) <- unique(anno$`Kit Name`)
  names(mat_colors$`Filter Name`) <- unique(anno$`Filter Name`)
  
  # create heatmap
  p <- pheatmap(sampleDistMatrix,
    annotation_col = anno,
    annotation_colors = mat_colors,
    clustering_distance_rows = sampleDists,
    clustering_distance_cols = sampleDists,
    color = viridis(
      n = 256, alpha = 1,
      begin = 0, end = 1, option = "viridis"
    ),
    # main = "Sample-to-sample Euclidian distance based on all genes",
    main = title,
    scale = "none",
    show_rownames = TRUE,
    show_colnames=FALSE
  ) +
    theme(plot.title = element_markdown())
  
  return(p)
}
```

### All genes (rlog)

```{r sample-distance-rlog}
heatmap_of_distances(rld_filtered, str_wrap("Sample-to-sample Euclidian distance based on all genes - rlog", 40))
```

<!-- TODO check scaling options https://adairama.wordpress.com/2022/09/10/beware-of-using-the-scale-parameter-in-pheatmap-or-scaled-data-in-complexheatmap/ -->

<!-- ```{r} -->
<!-- sampleDists <- dist(t(assay(rld_filtered))) -->
<!-- sampleDistMatrix <- as.matrix(sampleDists) -->
<!-- rownames(sampleDistMatrix) <- rld_filtered$sample_name -->
<!-- colnames(sampleDistMatrix) <- NULL -->
<!-- pheatmap(sampleDistMatrix, -->
<!--   clustering_distance_rows = sampleDists, -->
<!--   clustering_distance_cols = sampleDists, -->
<!--   color = viridis( -->
<!--     n = 256, alpha = 1, -->
<!--     begin = 0, end = 1, option = "viridis" -->
<!--   ), -->
<!--   main = "Sample-to-sample Euclidian distance based on all genes", -->
<!--   scale="row" -->
<!-- ) -->
<!-- ``` -->

### All genes (vst)

The VST transform leads to slightly different results: the two large blocks remains the same, but the positioning of samples 5, 11 and 16 differs. Overall, since the samples are so similar overall, slight deviations in the clustering of the distance matrix are not too surprising.

```{r sample-distance-vst}
heatmap_of_distances(vsd_filtered, str_wrap("Sample-to-sample Euclidian distance based on all genes - vst", 40))
```

### *P. knowlesi* only (rlog)

```{r sample-distance-pk-rlog, warning=FALSE}
heatmap_of_distances(rld_filtered[rowData(rld_filtered)$species == "pk", ], str_wrap("Sample-to-sample Euclidian distance based on *P. knowlesi* genes - rlog before subset", 40))
```

### *P. knowlesi* only (vst)

```{r sample-distance-pk-vst, warning=FALSE}
heatmap_of_distances(vsd_filtered[rowData(vsd_filtered)$species == "pk", ], str_wrap("Sample-to-sample Euclidian distance based on P. knowlesi genes - vst before subset", 40))
```

### *P. knowlesi* only (rlog) - transform after subset

```{r sample-distance-pk-rlog-transform, warning=FALSE}
heatmap_of_distances(rld_pk_filtered, str_wrap("Sample-to-sample Euclidian distance based on *P. knowlesi* genes - rlog after subset", 40))
```

### *P. knowlesi* only (vst) - transform after subset

```{r sample-distance-pk-vst-transform, warning=FALSE}
heatmap_of_distances(vsd_pk_filtered, str_wrap("Sample-to-sample Euclidian distance based on *P. knowlesi* genes - vst after subset", 40))
```

### *P. knowlesi* protein-coding only (rlog)

```{r sample-distance-pk-coding-rlog, warning=FALSE}
heatmap_of_distances(rld_filtered[rowData(rld_filtered)$species == "pk" & rowData(rld_filtered)$gene_biotype == "protein_coding", ], str_wrap("Sample-to-sample Euclidian distance based on *P. knowlesi** protein-coding genes", 40))
```

### *P. knowlesi* protein-coding only (vst)

```{r sample-distance-pk-coding-vst, warning=FALSE}
heatmap_of_distances(vsd_filtered[rowData(vsd_filtered)$species == "pk" & rowData(vsd_filtered)$gene_biotype == "protein_coding", ], str_wrap("Sample-to-sample Euclidian distance based on *P. knowlesi* protein-coding genes", 40))
```

### *P. knowlesi* protein-coding only (rlog) - transform after subset

```{r sample-distance-pk-coding-rlog-transform, warning=FALSE}
heatmap_of_distances(rld_pk_filtered_coding, str_wrap("Sample-to-sample Euclidian distance based on *P. knowlesi* protein-coding genes - rlog after subset", 40))
```

### *P. knowlesi* protein-coding only (vst) - transform after subset

```{r sample-distance-pk-coding-vst-transform, warning=FALSE}
heatmap_of_distances(vsd_pk_filtered_coding, str_wrap("Sample-to-sample Euclidian distance based on *P. knowlesi* protein-coding genes - vst after subset", 40))
```

## Heatmap of count matrix {.tabset}

<!-- **TODO: try quantile breaks for better separation between values instead, see https://slowkow.com/notes/pheatmap-tutorial/**  -->

```{r heatmap-count-function}
heatmap_of_counts <- function(raw_data, transformed_data, top_type = "mean", top_n = 20, center = TRUE) {
  # select n most highly expressed/variable genes
  if (top_type == "mean") {
    select <- order(rowMeans(counts(raw_data, normalized = FALSE)),
                    decreasing = TRUE
                    )[1:top_n]
  } else if (top_type == "median") {
    select <- order(rowMedians(counts(raw_data, normalized = FALSE)),
                    decreasing = TRUE
                    )[1:top_n]
  } else if (top_type == "var") {
    select <- head(order(rowVars(assay(transformed_data)), decreasing = TRUE), top_n)
  }
  
  # select transformed data and center if requested
  mat <- assay(transformed_data)[select, ]
  
  if (center == TRUE){
    mat <- mat - rowMeans(mat)
  }
  
  # add grouping columns
  anno <- as.data.frame(colData(transformed_data)[, c("kit_name", "filter_name")]) %>%
  mutate(kit_name = as.factor(kit_name)) %>%
  mutate(filter_name = as.factor(filter_name)) %>%
  rename_with(~ str_to_title(str_replace(.x, "_", " ")))
  
  # set names of samples
  colnames(mat) <- (transformed_data)$sample_short
  rownames(anno) <- (transformed_data)$sample_short
  
  # Adapted from: https://slowkow.com/notes/pheatmap-tutorial/
  mat_colors <- list(
    `Kit Name` = hcl.colors(length(unique(anno$`Kit Name`)), palette = "plasma", alpha = 1),
    `Filter Name` = palette.colors(length(unique(anno$`Filter Name`)), palette = "Okabe-Ito", alpha = 1)
  )
  names(mat_colors$`Kit Name`) <- unique(anno$`Kit Name`)
  names(mat_colors$`Filter Name`) <- unique(anno$`Filter Name`)

  # create heatmap
  pheatmap(mat,
    annotation_col = anno,
    annotation_colors = mat_colors,
    color = viridis(n = 256, alpha = 1, begin = 0, end = 1, option = "viridis"),
    # cluster_rows = FALSE,
    # cluster_cols = FALSE,
    show_rownames = FALSE,
  )
  
  # create table of genes
  gene_annotation %>% filter(gene_id %in% rownames(mat))
}
```

### All genes - vst - top expressed (mean)

This plot clusters the top n=20 most highly expressed genes (mean) centred on the gene's mean expression level across all samples using the VST data.

Several of the top expressed genes are globin and rRNA genes (and most reads are human).

```{r heatmap-counts-all-genes-vst-top-mean}
heatmap_of_counts(gse_filtered, vsd_filtered, "mean", 20, TRUE)
```

### All genes - rlog - top expressed (mean)

This plot cluster the top n=20 most highly expressed genes (mean) centred on the gene's mean expression level across all samples using the rlog transformed data.

Several of the top expressed genes are globin and rRNA genes (and most reads are human).

```{r heatmap-counts-all-genes-rlog-top-mean}
heatmap_of_counts(gse_filtered, rld_filtered, "mean", 20, TRUE)
```

### All genes - vst - top expressed (median)

This plot clusters the top n=20 most highly expressed genes (median) centred on the gene's mean expression level across all samples using the VST data.

Several of the top expressed genes are globin and rRNA genes (and most reads are human).

```{r heatmap-counts-all-genes-vst-top-median}
heatmap_of_counts(gse_filtered, vsd_filtered, "median", 20, TRUE)
```

### All genes - rlog - top expressed (median)

This plot cluster the top n=20 most highly expressed genes (median) centred on the gene's mean expression level across all samples using the rlog transformed data.

Several of the top expressed genes are globin and rRNA genes (and most reads are human).

```{r heatmap-counts-all-genes-rld-top-median}
heatmap_of_counts(gse_filtered, rld_filtered, "median", 20, TRUE)
```

### All genes - vst - top variable

This plot clusters the top n=20 most variable genes centred on the gene's mean expression level across all samples using the VST data.

Several of the top variable genes are human/Pk rRNA genes (and most reads are human).

```{r heatmap-counts-all-genes-vst-top-var}
heatmap_of_counts(gse_filtered, vsd_filtered, "var", 20, TRUE)
```

### All genes - rlog - top variable

This plot cluster the top n=20 most variable genes centred on the gene's mean expression level across all samples using the rlog transformed data.

Several of the top variable genes are human genes (but globin and rRNA are rare).

```{r heatmap-counts-all-genes-rld-top-var}
heatmap_of_counts(gse_filtered, rld_filtered, "var", 20, TRUE)
```

### Pk genes protein coding - vst - top expressed (mean)

This plot clusters the top n=20 most highly expressed genes centred on the gene's mean expression level across all samples using the VST data.

The choice of transformation only has a minor effect on this subset of top-expressed genes.

```{r heatmap-counts-pk-coding-vst-top-mean}
heatmap_of_counts(gse_filtered[rowData(gse_filtered)$species == "pk" & rowData(gse_filtered)$gene_biotype == "protein_coding"],
                  vsd_filtered[rowData(vsd_filtered)$species == "pk" & rowData(vsd_filtered)$gene_biotype == "protein_coding"], 
                  "mean", 20, TRUE)
```

### Pk genes protein coding - rlog - top expressed (mean)

This plot cluster the top n=20 most highly expressed genes centred on the gene's mean expression level across all samples using the rlog transformed data.

The choice of transformation only has a minor effect on this subset of top-expressed genes.

```{r heatmap-counts-pk-coding-rlog-top-mean}
heatmap_of_counts(gse_filtered[rowData(gse_filtered)$species == "pk" & rowData(gse_filtered)$gene_biotype == "protein_coding"],
                  rld_filtered[rowData(rld_filtered)$species == "pk" & rowData(rld_filtered)$gene_biotype == "protein_coding"], 
                  "mean", 20, TRUE)
```

### Pk genes protein coding - vst - top variable

This plot clusters the top n=20 most variable genes centred on the gene's mean expression level across all samples using the VST data.

The choice of transformation only has a minor effect on this subset of top-variable genes.

```{r heatmap-counts-pk-coding-vst-top-var}
heatmap_of_counts(gse_filtered[rowData(gse_filtered)$species == "pk" & rowData(gse_filtered)$gene_biotype == "protein_coding"],
                  vsd_filtered[rowData(vsd_filtered)$species == "pk" & rowData(vsd_filtered)$gene_biotype == "protein_coding"], 
                  "var", 20, TRUE)
```

### Pk genes protein coding - rlog - top variable

This plot cluster the top n=20 most variable genes centred on the gene's mean expression level across all samples using the rlog transformed data.

The choice of transformation only has a minor effect on this subset of top-variable genes.

```{r heatmap-counts-pk-coding-rlog-top-var}
heatmap_of_counts(gse_filtered[rowData(gse_filtered)$species == "pk" & rowData(gse_filtered)$gene_biotype == "protein_coding"],
                  rld_filtered[rowData(rld_filtered)$species == "pk" & rowData(rld_filtered)$gene_biotype == "protein_coding"], 
                  "var", 20, TRUE)
```

### Pk genes protein coding - vst - top expressed genes (mean) - transform after subsetting

This plot cluster the top n=20 most highly expressed Pk protein coding genes (centred on the gene's mean expression level) across all samples using the vst transformed data, without clustering. The data was transformed after subsetting.

```{r heatmap-counts-pk-coding-vst-top-mean-transform}
heatmap_of_counts(gse_pk_filtered_coding,
                  vsd_pk_filtered_coding, 
                  "mean", 20, TRUE)
```

### Pk genes protein coding - rlog - top expressed genes (mean) - transform after subsetting

This plot cluster the top n=20 most highly expressed Pk protein coding genes (centred on the gene's mean expression level) across all samples using the rlog transformed data, without clustering. The data was transformed after subsetting.

```{r heatmap-counts-pk-coding-rlog-top-mean-transform}
heatmap_of_counts(gse_pk_filtered_coding,
                  rld_pk_filtered_coding, 
                  "mean", 20, TRUE)
```

### Pk genes protein coding - vst - top variable - transform after subsetting

This plot cluster the top n=20 most highly variable Pk protein coding genes across all samples using the vst transformed data, without clustering. The data was transformed after subsetting.

```{r heatmap-counts-pk-coding-vst-top-var-transform}
heatmap_of_counts(gse_pk_filtered_coding,
                  vsd_pk_filtered_coding, 
                  "var", 20, TRUE)
```

### Pk genes protein coding - rlog - top variable - transform after subsetting

This plot cluster the top n=20 most highly variable Pk protein coding genes across all samples using the rlog transformed data, without clustering. The data was transformed after subsetting.

```{r heatmap-counts-pk-coding-rlog-top-var-transform}
heatmap_of_counts(gse_pk_filtered_coding,
                  rld_pk_filtered_coding, 
                  "var", 20, TRUE)
```

### Pk genes - vst - top expressed genes

This plot cluster the top n=20 most highly expressed Pk genes (centred on the gene's mean expression level) across all samples using the rlog transformed data, without clustering.

Several of the top expressed genes in the previous plot are rRNA genes, but not all of them.

```{r heatmap-counts-pk-all-vst-top-mean}
heatmap_of_counts(gse_filtered[rowData(gse_filtered)$species == "pk"],
                  vsd_filtered[rowData(gse_filtered)$species == "pk"],
                  "mean", 20, TRUE)
```

### Pk genes - rlog - top expressed genes

This plot cluster the top n=20 most highly expressed Pk genes (centred on the gene's mean expression level) across all samples using the rlog transformed data, without clustering.

```{r heatmap-counts-pk-all-rlog-top-mean}
heatmap_of_counts(gse_filtered[rowData(gse_filtered)$species == "pk"],
                  vsd_filtered[rowData(gse_filtered)$species == "pk"], 
                  "mean", 20, TRUE)
```

### Pk genes - vst - top variable

This plot cluster the top n=20 most highly variable Pk genes across all samples using the rlog transformed data, without clustering.

```{r heatmap-counts-pk-all-vst-top-var}
heatmap_of_counts(gse_filtered[rowData(gse_filtered)$species == "pk"],
                  vsd_filtered[rowData(gse_filtered)$species == "pk"],
                  "var", 20, TRUE)
```

### Pk genes - rlog - top variable

This plot cluster the top n=20 most highly variable Pk genes across all samples using the rlog transformed data, without clustering.

```{r heatmap-counts-pk-all-rlog-top-var}
heatmap_of_counts(gse_filtered[rowData(gse_filtered)$species == "pk"],
                  vsd_filtered[rowData(gse_filtered)$species == "pk"], 
                  "var", 20, TRUE)
```

### Pk genes - vst - top expressed genes - transform after subsetting

This plot cluster the top n=20 most highly expressed Pk genes (centred on the gene's mean expression level) across all samples using the rlog transformed data, without clustering. The data was transformed after subsetting.

Several of the top expressed genes in the previous plot are rRNA genes, but not all of them.

```{r heatmap-counts-pk-all-vst-top-mean-transform}
heatmap_of_counts(gse_pk_filtered,
                  vsd_pk_filtered,
                  "mean", 20, TRUE)
```

### Pk genes - rlog - top expressed genes - transform after subsetting

This plot cluster the top n=20 most highly expressed Pk genes (centred on the gene's mean expression level) across all samples using the rlog transformed data, without clustering. The data was transformed after subsetting.

Several of the top expressed genes in the previous plot are rRNA genes, but not all of them.

```{r heatmap-counts-pk-all-rlog-top-mean-transform}
heatmap_of_counts(gse_pk_filtered,
                  rld_pk_filtered, 
                  "mean", 20, TRUE)
```

### Pk genes - vst - top variable - transform after subsetting

This plot cluster the top n=20 most highly variable Pk genes across all samples using the rlog transformed data, without clustering. The data was transformed after subsetting.

```{r heatmap-counts-pk-all-vst-top-var-transform}
heatmap_of_counts(gse_pk_filtered,
                  vsd_pk_filtered,
                  "var", 20, TRUE)
```

### Pk genes - rlog - top variable - transform after subsetting

This plot cluster the top n=20 most highly variable Pk genes across all samples using the rlog transformed data, without clustering. The data was transformed after subsetting.

```{r heatmap-counts-pk-all-rlog-top-var-transform}
heatmap_of_counts(gse_pk_filtered,
                  rld_pk_filtered, 
                  "var", 20, TRUE)
```

## Expression count correlation plots

### Protein-coding Pk genes - TPM

The following figure plots the (log-transformed) TPM of each protein-coding *P. knowlesi* gene (filtered on a minimum count of 10 for at least one sample) for all pairwise combinations of samples. 

```{r, fig.width=10, fig.height=10}
mat <- round(
          cor(
            gene_tpm_filtered %>%
              filter(species == "pk") %>%
              filter(gene_biotype == "protein_coding") %>% 
              select(`104974-001-001_S1_L001`:`104974-001-016_S1_L001`) %>% 
              rename_with(~samples %>% pull(sample_name) %>% as.character(), everything())
            ),
          1)
corr <- mat

ggcorrplot(corr,
           type = "lower",
           method = "circle",
           hc.order = FALSE,
           outline.color = "white",
           p.mat = cor_pmat(mat),
           lab = TRUE,
           legend.title	= "Pearson correlation of TPM values") +
  scale_fill_gradientn(colors = viridis(256, option = 'D'), limits = c(-1,1),
                       name = "Pearson correlation")
```

```{r correlation-plot, message=FALSE, cache=TRUE, fig.height=18, out.width="100%"}
correlation_plot <- ggpairs(
  gene_tpm_filtered %>%
    filter(species == "pk") %>%
    filter(gene_biotype == "protein_coding"),
  columns = 3:18,
  columnLabels = samples$sample_short,
  upper = list(continuous = wrap("cor", size = 5)),
) +
  theme(
    axis.text = element_text(size = 9),
    axis.title = element_text(size = 9),
    strip.text.x = element_text(size = 11),
    strip.text.y = element_text(size = 10),
    axis.text.x = element_text(size = 6),
    axis.text.y = element_text(size = 9)
    )
  # scale_y_discrete()

correlation_plot
```

```{r correlation-plot-log, message=FALSE, cache=TRUE, fig.height=18, out.width="100%"}
correlation_plot_log <- ggpairs(
  gene_tpm_filtered %>%
    filter(species == "pk") %>%
    filter(gene_biotype == "protein_coding"),
  columns = 3:18, # `104974-001-001_S1_L001`:`104974-001-002_S1_L001`,
  lower = list(continuous = "smooth"),
  upper = list(continuous = wrap("cor", size = 5)),
  columnLabels = samples$sample_short
) +
  theme(
    axis.text = element_text(size = 9),
    axis.title = element_text(size = 9),
    strip.text.x = element_text(size = 11),
    strip.text.y = element_text(size = 10),
    axis.text.x = element_text(size = 6),
    axis.text.y = element_text(size = 9)
    ) +
  scale_x_continuous(trans = "log1p") +
  scale_y_continuous(trans = "log1p")

correlation_plot_log
```

## PCA plots {.tabset}

All PCA plots use the top 500 most variable genes and are filtered to only genes with a minimum count of 10 reads for at least one sample, unless otherwise stated.

### All genes (rlog)

```{r}
pcaData <- plotPCA(rld_filtered, intgroup = c("kit_name", "filter_name"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit_name, shape = filter_name)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(
    title = expression(paste("PCA of rlog-transformed counts of human and ", italic("P. knowlesi"), " genes")),
    color = "Library preparation kit", shape = "Filter method"
  ) +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

### All genes (vst)

```{r}
pcaData <- plotPCA(vsd_filtered, intgroup = c("kit_name", "filter_name"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit_name, shape = filter_name)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(
    title = expression(paste("PCA of VST counts of human and ", italic("P. knowlesi"), " genes")),
    color = "Library preparation kit", shape = "Filter method"
  ) +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

### Pk genes (rlog)

```{r}
pcaData <- plotPCA(rld_filtered[rowData(rld_filtered)$species == "pk", ], intgroup = c("kit_name", "filter_name"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit_name, shape = filter_name)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(
    title = expression(paste("PCA of rlog-transformed counts of ", italic("P. knowlesi"), " genes")),
    color = "Library preparation kit", shape = "Depletion method"
  ) +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

### Pk genes (vst)

```{r}
pcaData <- plotPCA(vsd_filtered[rowData(vsd_filtered)$species == "pk", ], intgroup = c("kit_name", "filter_name"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit_name, shape = filter_name)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(
    title = expression(paste("PCA of VST counts of ", italic("P. knowlesi"), " genes")),
    color = "Library preparation kit", shape = "Depletion method"
  ) +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

### Pk genes protein-coding (rlog)

```{r}
pcaData <- plotPCA(rld_filtered[rowData(rld_filtered)$species == "pk" & rowData(rld_filtered)$gene_biotype == "protein_coding"], intgroup = c("kit_name", "filter_name"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit_name, shape = filter_name)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(
    title = expression(paste("PCA of rlog-transformed counts of ", italic("P. knowlesi"), " genes")),
    color = "Library preparation kit", shape = "Depletion method"
  ) +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

### Pk genes protein-coding (vst)

```{r}
pcaData <- plotPCA(vsd_filtered[rowData(vsd_filtered)$species == "pk" & rowData(vsd_filtered)$gene_biotype == "protein_coding"], intgroup = c("kit_name", "filter_name"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit_name, shape = filter_name)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(
    title = expression(paste("PCA of VST counts of ", italic("P. knowlesi"), " genes")),
    color = "Library preparation kit", shape = "Depletion method"
  ) +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

### Pk genes (rlog) - transform after subset

```{r}
pcaData <- plotPCA(rld_pk_filtered, intgroup = c("kit_name", "filter_name"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit_name, shape = filter_name)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(
    title = expression(paste("PCA of rlog-transformed counts of ", italic("P. knowlesi"), " genes")),
    color = "Library preparation kit", shape = "Depletion method"
  ) +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

### Pk genes (vst) - transform after subset

```{r}
pcaData <- plotPCA(vsd_pk_filtered, intgroup = c("kit_name", "filter_name"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit_name, shape = filter_name)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(
    title = expression(paste("PCA of VST counts of ", italic("P. knowlesi"), " genes")),
    color = "Library preparation kit", shape = "Depletion method"
  ) +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

### Pk genes protein-coding (rlog)

```{r}
protein_gene_filter <- intersect(rownames(rowData(rld_filtered)), gene_annotation %>% filter(gene_biotype == "protein_coding" & species == "pk") %>% select(gene_id) %>% pull())
pcaData <- plotPCA(rld_filtered[protein_gene_filter, ], intgroup = c("kit_name", "filter_name"), returnData = TRUE, ntop = 500)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit_name, shape = filter_name)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(
    title = expression(paste("PCA of rlog-transformed counts of ", italic("P. knowlesi"), " genes")),
    color = "Library preparation kit", shape = "Depletion method"
  ) +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

### Pk genes protein-coding (vst)

```{r}
protein_gene_filter <- intersect(rownames(rowData(vsd_filtered)), gene_annotation %>% filter(gene_biotype == "protein_coding" & species == "pk") %>% select(gene_id) %>% pull())
pcaData <- plotPCA(vsd_filtered[protein_gene_filter, ], intgroup = c("kit_name", "filter_name"), returnData = TRUE, ntop = 500)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit_name, shape = filter_name)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(
    title = expression(paste("PCA of VST counts of ", italic("P. knowlesi"), " genes")),
    color = "Library preparation kit", shape = "Depletion method"
  ) +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

### Pk genes protein-coding (rlog) - transform after subset

```{r}
pcaData <- plotPCA(rld_pk_filtered_coding, intgroup = c("kit_name", "filter_name"), returnData = TRUE, ntop = 500)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit_name, shape = filter_name)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(
    title = expression(paste("PCA of rlog counts of ", italic("P. knowlesi"), " protein coding genes")),
    color = "Library preparation kit", shape = "Depletion method"
  ) +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

### Pk genes protein-coding (vst) - transform after subset

```{r}
pcaData <- plotPCA(vsd_pk_filtered_coding, intgroup = c("kit_name", "filter_name"), returnData = TRUE, ntop = 500)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = kit_name, shape = filter_name)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(
    title = expression(paste("PCA of VST counts of ", italic("P. knowlesi"), " protein coding genes")),
    color = "Library preparation kit", shape = "Depletion method"
  ) +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()
```

# Figures for publication

## Sum of gene counts {.tabset}

### Ordered by library - depletion - filter

```{r combined-figure-ordered-alternative, message = FALSE, fig.width=10}
samples_order <- samples %>% 
  arrange(factor(samples$kit, levels = c("mRNA_illumina", "mRNA_qiagen_FSglobH", "mRNA_qiagen_FSglob", "tRNA_qiagen_FSglobH")),
          factor(samples$WBC_depletion, levels = c("nofilter", "centpip", "plasmo", "pmacs", "cellulose"))) %>% 
  select(sample_name)

p2_a <- gene_counts_txi_filtered_long %>%
  left_join(samples, by = "sample") %>%
  group_by(sample_name, species) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads)) %>% 
  # arrange(factor(sample_name, levels = samples_order %>% pull())) %>% 
  mutate(sample_name = factor(sample_name, levels = samples_order %>% pull())) %>% 
  ggplot(aes(x = sum_reads, y = sample_name, fill = species)) +
  geom_bar(position = "stack", stat = "identity") +
  scale_fill_viridis(
    option = "viridis", discrete = T,
    # limits = rev,
    # labels = c("human", "human globin", "human rRNA", expression(italic("P. knowlesi")), expression(italic("P. knowlesi rRNA")))
  ) +
  scale_y_discrete(limits = rev) +
  scale_x_continuous(labels = scales::label_number(scale_cut = cut_short_scale())) +
  labs(y = "Sample", x = "Sum of all gene counts", fill = "RNA type") +
  theme_classic() +
  ggtitle("A")

# print p2a to show y labels
p2_a
p2_a <- p2_a + theme(axis.title.y = element_blank(), axis.text.y = element_blank())

p2_b <- gene_tpm_filtered_long %>%
  left_join(samples, by = "sample") %>%
  group_by(sample_name, gene_type) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads)) %>%
  mutate(sample_name = factor(sample_name, levels = samples_order %>% pull())) %>% 
  ggplot(aes(x = sum_reads, y = sample_name, fill = gene_type)) +
  geom_bar(position = "stack", stat = "identity") +
  scale_fill_manual(
    values = c(`human` = "#440154FF", 
               `human_globin` = "#3B528BFF", 
               `human_rRNA` = "#21908CFF",
               `pk_rRNA` = "#5DC863FF",
               `pk` = "#FDE725FF"),
      labels = c("human", "human globin", "human rRNA", expression(italic("P. knowlesi")),
                 expression(italic("P. knowlesi rRNA")))
    ) +
  # scale_fill_viridis(
  #   option = "viridis", discrete = T,
  #   # limits = rev,
  #   labels = c("human", "human globin", "human rRNA", expression(italic("P. knowlesi")),
  #              expression(italic("P. knowlesi rRNA")))
  # ) +
  scale_y_discrete(limits = rev) +
  scale_x_continuous(labels = scales::label_number(scale_cut = cut_short_scale())) +
  labs(y = "Sample", x = "Sum of TPM counts", fill = "RNA type") +
  theme_classic() +
  theme(axis.title.y = element_blank(), axis.text.y = element_blank()) +
  ggtitle("B")

p2_c <- gene_counts_txi_filtered_long %>%
  left_join(samples, by = "sample") %>%
  filter(species == "pk" & gene_biotype == "protein_coding") %>%
  group_by(sample_name,) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads)) %>%
  mutate(sample_name = factor(sample_name, levels = samples_order %>% pull())) %>% 
  ggplot(aes(x = sum_reads, y = sample_name)) +
  geom_bar(position = "stack", stat = "identity", fill = "#FDE725FF") +
  # scale_fill_manual(viridis::viridis(5)[4:5]) +
  # scale_fill_manual(
  #   values = c(pk = "#5DC863FF", pk_rRNA = "#FDE725FF"),
  #   # scale_fill_viridis(
  #   #   option = "viridis", discrete = T,
  #   #   # limits = rev,
  #   labels = c(expression(italic("P. knowlesi")), expression(italic("P. knowlesi rRNA")))
  # ) +
  scale_y_discrete(limits = rev) +
  scale_x_continuous(labels = scales::label_number(scale_cut = cut_short_scale())) +
  labs(y = "Sample", x = str_wrap("Sum of *P. knowlesi* protein-coding gene counts", 10), fill = "RNA type") +
  theme_classic() +
  theme(axis.title.y = element_blank(), axis.text.y = element_blank(), axis.title.x = ggtext::element_markdown()) +
  ggtitle("C")

fig <- ggarrange(
  p2_a, p2_b, p2_c,
  ncol = 3,
  legend = "bottom",
  common.legend = TRUE,
  legend.grob = get_legend(p2_b)
  # widths = c(2,1,2)
)
fig

ggsave(filename = here("results/figures/gene_counts.svg"), 
      plot = fig, width=10, height=6)
```

### Ordered by sample number (subtypes)

```{r combined-figure, message = FALSE}
p2_a <- gene_counts_txi_filtered_long %>%
  left_join(samples, by = "sample") %>%
  group_by(sample_name, gene_type) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads)) %>%
  ggplot(aes(y = sum_reads, x = sample_name, fill = gene_type)) +
  geom_bar(position = "stack", stat = "identity") +
  scale_fill_viridis(
    option = "viridis", discrete = T,
    # limits = rev,
    labels = c("human", "human globin", "human rRNA", expression(italic("P. knowlesi")), expression(italic("P. knowlesi rRNA")))
  ) +
  coord_flip() +
  scale_x_discrete(limits = rev) +
  scale_y_continuous(labels = scales::label_number(scale_cut = cut_short_scale())) +
  labs(x = "Sample", y = "Sum of gene counts", fill = "RNA type") +
  theme_classic() +
  ggtitle("A")

# print p2a to show y labels
p2_a
p2_a <- p2_a + theme(axis.title.y = element_blank(), axis.text.y = element_blank())


p2_b <- gene_tpm_filtered_long %>%
  left_join(samples, by = "sample") %>%
  group_by(sample_name, gene_type) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads)) %>%
  ggplot(aes(y = sum_reads, x = sample_name, fill = gene_type)) +
  geom_bar(position = "stack", stat = "identity") +
  scale_fill_viridis(
    option = "viridis", discrete = T,
    # limits = rev,
    labels = c("human", "human globin", "human rRNA", expression(italic("P. knowlesi")), expression(italic("P. knowlesi rRNA")))
  ) +
  coord_flip() +
  scale_x_discrete(limits = rev) +
  scale_y_continuous(labels = scales::label_number(scale_cut = cut_short_scale())) +
  labs(x = "Sample", y = "Sum of TPM counts", fill = "RNA type") +
  theme_classic() +
  theme(axis.title.y = element_blank(), axis.text.y = element_blank()) +
  ggtitle("B")

p2_c <- gene_counts_txi_filtered_long %>%
  left_join(samples, by = "sample") %>%
  filter(species == "pk") %>%
  group_by(sample_name, gene_type) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads)) %>%
  ggplot(aes(y = sum_reads, x = sample_name, fill = gene_type)) +
  geom_bar(position = "stack", stat = "identity") +
  # scale_fill_manual(viridis::viridis(5)[4:5]) +
  scale_fill_manual(
    values = c(pk = "#5DC863FF", pk_rRNA = "#FDE725FF"),
    # scale_fill_viridis(
    #   option = "viridis", discrete = T,
    #   # limits = rev,
    labels = c(expression(italic("P. knowlesi")), expression(italic("P. knowlesi rRNA")))
  ) +
  coord_flip() +
  scale_x_discrete(limits = rev) +
  scale_y_continuous(labels = scales::label_number(scale_cut = cut_short_scale())) +
  labs(x = "Sample", y = expression(paste("Sum of ", italic("P. knowlesi"), " gene counts")), fill = "RNA type") +
  theme_classic() +
  theme(axis.title.y = element_blank(), axis.text.y = element_blank()) +
  ggtitle("C")

fig <- ggarrange(
  p2_a, p2_b, p2_c,
  ncol = 3,
  common.legend = TRUE, legend = "bottom"
)
fig
```

### Ordered by library - depletion - filter (subtypes)

```{r combined-figure-ordered, message = FALSE}
samples_order <- samples %>% 
  arrange(factor(samples$kit, levels = c("mRNA_illumina", "mRNA_qiagen_FSglobH", "mRNA_qiagen_FSglob", "tRNA_qiagen_FSglobH")),
          factor(samples$WBC_depletion, levels = c("nofilter", "centpip", "plasmo", "pmacs", "cellulose"))) %>% 
  select(sample_name)

p2_a <- gene_counts_txi_filtered_long %>%
  left_join(samples, by = "sample") %>%
  group_by(sample_name, gene_type) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads)) %>% 
  # arrange(factor(sample_name, levels = samples_order %>% pull())) %>% 
  mutate(sample_name = factor(sample_name, levels = samples_order %>% pull())) %>% 
  ggplot(aes(x = sum_reads, y = sample_name, fill = gene_type)) +
  geom_bar(position = "stack", stat = "identity") +
  scale_fill_viridis(
    option = "viridis", discrete = T,
    # limits = rev,
    labels = c("human", "human globin", "human rRNA", expression(italic("P. knowlesi")), expression(italic("P. knowlesi rRNA")))
  ) +
  scale_y_discrete(limits = rev) +
  scale_x_continuous(labels = scales::label_number(scale_cut = cut_short_scale())) +
  labs(y = "Sample", x = "Sum of gene counts", fill = "RNA type") +
  theme_classic() +
  ggtitle("A")

# print p2a to show y labels
p2_a
p2_a <- p2_a + theme(axis.title.y = element_blank(), axis.text.y = element_blank())

p2_b <- gene_tpm_filtered_long %>%
  left_join(samples, by = "sample") %>%
  group_by(sample_name, gene_type) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads)) %>%
  mutate(sample_name = factor(sample_name, levels = samples_order %>% pull())) %>% 
  ggplot(aes(x = sum_reads, y = sample_name, fill = gene_type)) +
  geom_bar(position = "stack", stat = "identity") +
  scale_fill_viridis(
    option = "viridis", discrete = T,
    # limits = rev,
    labels = c("human", "human globin", "human rRNA", expression(italic("P. knowlesi")), expression(italic("P. knowlesi rRNA")))
  ) +
  scale_y_discrete(limits = rev) +
  scale_x_continuous(labels = scales::label_number(scale_cut = cut_short_scale())) +
  labs(y = "Sample", x = "Sum of TPM counts", fill = "RNA type") +
  theme_classic() +
  theme(axis.title.y = element_blank(), axis.text.y = element_blank()) +
  ggtitle("B")

p2_c <- gene_counts_txi_filtered_long %>%
  left_join(samples, by = "sample") %>%
  filter(species == "pk") %>%
  group_by(sample_name, gene_type) %>%
  summarise(sum_reads = sum(counts, na.rm = TRUE)) %>%
  mutate(per = 100 * sum_reads / sum(sum_reads)) %>%
  mutate(sample_name = factor(sample_name, levels = samples_order %>% pull())) %>% 
  ggplot(aes(x = sum_reads, y = sample_name, fill = gene_type)) +
  geom_bar(position = "stack", stat = "identity") +
  # scale_fill_manual(viridis::viridis(5)[4:5]) +
  scale_fill_manual(
    values = c(pk = "#5DC863FF", pk_rRNA = "#FDE725FF"),
    # scale_fill_viridis(
    #   option = "viridis", discrete = T,
    #   # limits = rev,
    labels = c(expression(italic("P. knowlesi")), expression(italic("P. knowlesi rRNA")))
  ) +
  scale_y_discrete(limits = rev) +
  scale_x_continuous(labels = scales::label_number(scale_cut = cut_short_scale())) +
  labs(y = "Sample", x = expression(paste("Sum of ", italic("P. knowlesi"), " gene counts")), fill = "RNA type") +
  theme_classic() +
  theme(axis.title.y = element_blank(), axis.text.y = element_blank()) +
  ggtitle("C")

fig <- ggarrange(
  p2_a, p2_b, p2_c,
  ncol = 3,
  common.legend = TRUE, legend = "bottom"
)
fig
```

## Heatmaps of distances {.tabset}

```{r}
heatmap_of_distances <- function(transformed_data, title) {
  
  sampleDists <- dist(t(assay(transformed_data)))
  sampleDistMatrix <- as.matrix(sampleDists)
  rownames(sampleDistMatrix) <- transformed_data$sample_short
  colnames(sampleDistMatrix) <- NULL
  
    
  # add grouping columns
  anno <- as.data.frame(colData(transformed_data)[, c("kit_name", "filter_name")]) %>%
    mutate(kit_name = as.factor(kit_name)) %>%
    mutate(filter_name = as.factor(filter_name)) %>%
    rename_with(~ str_to_title(str_replace(.x, "_", " ")))
  
  # set names of samples
  colnames(sampleDistMatrix) <- (transformed_data)$sample_short
  rownames(anno) <- (transformed_data)$sample_short
  
  # Adapted from: https://slowkow.com/notes/pheatmap-tutorial/
  mat_colors <- list(
    `Kit Name` = hcl.colors(length(unique(anno$`Kit Name`)), palette = "plasma", alpha = 1),
    `Filter Name` = palette.colors(length(unique(anno$`Filter Name`)), palette = "Okabe-Ito", alpha = 1)
  )
  names(mat_colors$`Kit Name`) <- unique(anno$`Kit Name`)
  names(mat_colors$`Filter Name`) <- unique(anno$`Filter Name`)
  
  # create heatmap
  p <- pheatmap(sampleDistMatrix,
    annotation_col = anno,
    annotation_colors = mat_colors,
    clustering_distance_rows = sampleDists,
    clustering_distance_cols = sampleDists,
    color = viridis(
      n = 256, alpha = 1,
      begin = 0, end = 1, option = "viridis"
    ),
    # main = "Sample-to-sample Euclidian distance based on all genes",
    # main = title,
    scale = "none",
    show_rownames = TRUE,
    show_colnames=FALSE
  ) 
  
  return(p)
}
```

### Pk protein coding genes (rlog before subset)

```{r message=FALSE}
p <- heatmap_of_distances(rld_filtered[rowData(rld_filtered)$species == "pk" & rowData(rld_filtered)$gene_biotype == "protein_coding", ], str_wrap("Sample-to-sample Euclidian distance based on *P. knowlesi** protein-coding genes", 40))
ggsave(filename = here("results/figures/genes_pk_coding_sample_dist_rlog.svg"), 
      plot = p)
p
```

### Pk protein coding genes (rlog after subset)

```{r message=FALSE}
p <- heatmap_of_distances(rld_pk_filtered_coding, str_wrap("Sample-to-sample Euclidian distance based on *P. knowlesi* protein-coding genes - rlog after subset", 40))
ggsave(filename = here("results/figures/genes_pk_coding_sample_dist_rlog_transform_after_subset.svg"), 
      plot = p)
p
```

## Heatmaps of count matrix {.tabset}

```{r}
heatmap_of_counts <- function(raw_data, transformed_data, top_type = "mean", top_n = 20, center = TRUE) {
  # select n most highly expressed/variable genes
  if (top_type == "mean") {
    select <- order(rowMeans(counts(raw_data, normalized = FALSE)),
                    decreasing = TRUE
                    )[1:top_n]
  } else if (top_type == "median") {
    select <- order(rowMedians(counts(raw_data, normalized = FALSE)),
                    decreasing = TRUE
                    )[1:top_n]
  } else if (top_type == "var") {
    select <- head(order(rowVars(assay(transformed_data)), decreasing = TRUE), top_n)
  }
  
  # select transformed data and center if requested
  mat <- assay(transformed_data)[select, ]
  
  if (center == TRUE){
    mat <- mat - rowMeans(mat)
  }
  
  # add grouping columns
  anno <- as.data.frame(colData(transformed_data)[, c("kit_name", "filter_name")]) %>%
  mutate(kit_name = as.factor(kit_name)) %>%
  mutate(filter_name = as.factor(filter_name)) %>%
  rename_with(~ str_to_title(str_replace(.x, "_", " ")))
  
  # set names of samples
  colnames(mat) <- (transformed_data)$sample_short
  rownames(anno) <- (transformed_data)$sample_short
  
  # Adapted from: https://slowkow.com/notes/pheatmap-tutorial/
  mat_colors <- list(
    `Kit Name` = hcl.colors(length(unique(anno$`Kit Name`)), palette = "plasma", alpha = 1),
    `Filter Name` = palette.colors(length(unique(anno$`Filter Name`)), palette = "Okabe-Ito", alpha = 1)
  )
  names(mat_colors$`Kit Name`) <- unique(anno$`Kit Name`)
  names(mat_colors$`Filter Name`) <- unique(anno$`Filter Name`)

  # create heatmap
  pheatmap(mat,
    annotation_col = anno,
    annotation_colors = mat_colors,
    color = viridis(n = 256, alpha = 1, begin = 0, end = 1, option = "viridis"),
    # cluster_rows = FALSE,
    # cluster_cols = FALSE,
    show_rownames = FALSE,
  )
}
```

### Pk protein coding genes (top variable, rlog before subset)

```{r message=FALSE}
p <- heatmap_of_counts(gse_filtered[rowData(gse_filtered)$species == "pk" & rowData(gse_filtered)$gene_biotype == "protein_coding"],
                  rld_filtered[rowData(rld_filtered)$species == "pk" & rowData(rld_filtered)$gene_biotype == "protein_coding"], 
                  "var", 20, TRUE)
ggsave(filename = here("results/figures/genes_pk_coding_top_var_rlog.svg"), 
      plot = p)
p
```

### Pk protein coding genes (top variable, rlog after subset)

```{r message=FALSE}
p <- heatmap_of_counts(gse_pk_filtered_coding,
                  rld_pk_filtered_coding, 
                  "var", 20, TRUE)
ggsave(filename = here("results/figures/genes_pk_coding_top_var_rlog_transform_after_subset.svg"), 
      plot = p)
p
```

### Pk all genes (top variable, rlog before subset)

```{r message=FALSE}
p <- heatmap_of_counts(gse_filtered[rowData(gse_filtered)$species == "pk"],
                  vsd_filtered[rowData(gse_filtered)$species == "pk"], 
                  "var", 20, TRUE)
ggsave(filename = here("results/figures/genes_pk_all_top_var_rlog.svg"), 
      plot = p)
p
```

### Pk all genes (top variable, rlog after subset)

```{r message=FALSE}
p <- heatmap_of_counts(gse_pk_filtered,
                  rld_pk_filtered, 
                  "var", 20, TRUE)
ggsave(filename = here("results/figures/genes_pk_all_top_var_rlog_transform_after_subset.svg"), 
      plot = p)
p
```

## PCA plots

### Pk genes protein-coding (rlog) - transform after subset

```{r message=FALSE}
pcaData <- plotPCA(rld_pk_filtered_coding, intgroup = c("kit_name", "filter_name"), returnData = TRUE, ntop = 500)
percentVar <- round(100 * attr(pcaData, "percentVar"))
p <- ggplot(pcaData, aes(PC1, PC2, color = kit_name, shape = filter_name)) +
  geom_point(size = 4) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  labs(
    title = expression(paste("PCA of rlog counts of ", italic("P. knowlesi"), " protein coding genes")),
    color = "Library preparation kit", shape = "Depletion method"
  ) +
  coord_fixed() +
  scale_color_viridis(discrete = TRUE) +
  theme_bw()

ggsave(filename = here("results/figures/pca_pk_coding_rlog_transform_after_subset.svg"), 
      plot = p)
p
```

## Gene recovery venn diagrams

See previous section for in-line plots; the following code chunk only exports the relevant files.

```{r, results='hide'}
draw_venn <- function(set_list, names, title, output) {
  VennDiagram::venn.diagram(
    x = set_list,
    category.names = names,
    
    # Circles
    lwd = 2,
    col = hcl.colors(3, palette = "viridis", alpha = 0.9),
    fill = hcl.colors(3, palette = "viridis", alpha = 0.4),

    # Numbers
    cex = 1.4,
    fontface = "bold",
    fontfamily = "sans",

    # Set names
    cat.cex = 1.1,
    cat.fontface = "bold",
    cat.default.pos = "outer",
    cat.pos = c(-15, 15, 125),
    cat.dist = c(0.055, 0.055, 0.09),
    cat.fontfamily = "sans",
    rotation = 1,

    # title
    main = title,
    main.cex = 1.4,
    main.fontface = "bold",
    main.fontfamily = "sans",

    # output features
    # height = 1400,
    # width = 1400,
    units = "px",
    resolution = 600,
    # compression = "lzw",
    filename = output,
    output = TRUE,
    disable.logging = TRUE,
    imagetype = "png"
  )
}

p <- draw_venn(
  gene_tpm_filtered_binary_list[comparison_centpip$sample],
  # c("mRNA Qiagen FS globin", "total RNA Qiagen FS globin/rRNA", "mRNA Illumina"),
  c("Qiagen mRNA FSglob", "Qiagen total RNA FSglobH", "Illumina mRNA"),
  "Cent/pip",
  here("results/figures/venn-centpip.png")
)

draw_venn(
  gene_tpm_filtered_binary_list[comparison_plasmo$sample],
  c("Qiagen mRNA FSglob", "Qiagen total RNA FSglobH", "Illumina mRNA"),
  "Plasmo",
  here("results/figures/venn-plasmo.png")
)

draw_venn(
  gene_tpm_filtered_binary_list[comparison_pmacs$sample],
  c("Qiagen mRNA FSglob", "Qiagen total RNA FSglobH", "Illumina mRNA"),
  "PMACS",
  here("results/figures/venn-pmacs.png")
)

draw_venn(
  gene_tpm_filtered_binary_list[comparison_cellulose$sample],
  c("Qiagen mRNA FSglob", "Qiagen total RNA FSglobH", "Illumina mRNA"),
  "Cellulose",
  here("results/figures/venn-cellulose.png")
)

draw_venn(
  gene_tpm_filtered_binary_list[comparison_nofilter$sample],
  c("Qiagen mRNA FSglob", "Qiagen total RNA FSglobH", "Illumina mRNA"),
  "Cellulose",
  here("results/figures/venn-nofilter.png")
)

draw_venn(
  gene_tpm_pk_filtered_binary_list[comparison_centpip$sample],
  c("Qiagen mRNA FSglob", "Qiagen total RNA FSglobH", "Illumina mRNA"),
  "Cent/pip",
  here("results/figures/venn-centpip_pk.png")
)

draw_venn(
  gene_tpm_pk_filtered_binary_list[comparison_plasmo$sample],
  c("Qiagen mRNA FSglob", "Qiagen total RNA FSglobH", "Illumina mRNA"),
  "Plasmo",
  here("results/figures/venn-plasmo_pk.png")
)

draw_venn(
  gene_tpm_pk_filtered_binary_list[comparison_pmacs$sample],
  c("Qiagen mRNA FSglob", "Qiagen total RNA FSglobH", "Illumina mRNA"),
  "PMACS",
  here("results/figures/venn-pmacs_pk.png")
)

draw_venn(
  gene_tpm_pk_filtered_binary_list[comparison_cellulose$sample],
  c("Qiagen mRNA FSglob", "Qiagen total RNA FSglobH", "Illumina mRNA"),
  "Cellulose",
  here("results/figures/venn-cellulose_pk.png")
)

draw_venn(
  gene_tpm_pk_filtered_binary_list[comparison_nofilter$sample],
  c("Qiagen mRNA FSglob", "Qiagen total RNA FSglobH", "Illumina mRNA"),
  "No filter",
  here("results/figures/venn-nofilter_pk.png")
)
```

## Gene expression count correlation plots {.tabset}

### Simple correlation matrix

```{r, fig.width=10, fig.height=10}
mat <- round(
          cor(
            gene_tpm_filtered %>%
              filter(species == "pk") %>%
              filter(gene_biotype == "protein_coding") %>% 
              select(`104974-001-001_S1_L001`:`104974-001-016_S1_L001`) %>% 
              rename_with(~samples %>% pull(sample_name) %>% as.character(), everything())
            ),
          1)
corr <- mat

p <- ggcorrplot(corr,
           type = "lower",
           method = "circle",
           hc.order = FALSE,
           outline.color = "white",
           p.mat = cor_pmat(mat),
           lab = TRUE,
           legend.title	= "Pearson correlation of TPM values") +
  scale_fill_gradientn(colors = viridis(256, option = 'D'), limits = c(-1,1), name = "Pearson correlation")
ggsave(filename = here("results/figures/pk_coding_tpm_correlation.svg"), 
      plot = p,
      width = 10, height = 10)
p
```

Ordered by mRNA - total RNA library preparation kit:

```{r}
mat <- round(
          cor(
            gene_tpm_filtered %>%
              filter(species == "pk") %>%
              filter(gene_biotype == "protein_coding") %>% 
              select(`104974-001-001_S1_L001`:`104974-001-016_S1_L001`) %>% 
              rename_with(~samples %>% pull(sample_name) %>% as.character(), everything()) %>% 
              relocate(samples_order %>% pull())
            ),
          1)
corr <- mat

p <- ggcorrplot(corr,
           type = "lower",
           method = "circle",
           hc.order = FALSE,
           outline.color = "white",
           p.mat = cor_pmat(mat),
           lab = TRUE,
           legend.title	= "Pearson correlation of TPM values") +
  scale_fill_gradientn(colors = viridis(256, option = 'D'), limits = c(-1,1), name = "Pearson correlation")
ggsave(filename = here("results/figures/pk_coding_tpm_correlation_ordered.svg"), 
      plot = p,
      width = 10, height = 10)
p
```

### Correlation pairs plots

Pk protein coding genes only:

```{r correlation-plot-export, message=FALSE, cache=TRUE, fig.height=18, out.width="100%"}
correlation_plot

ggsave(filename = here("results/figures/correlation_protein_coding_genes_pk.png"), 
      plot = correlation_plot,
      height = 4000, width = 6000, units = "px")
ggsave(filename = here("results/figures/correlation_protein_coding_genes_pk.svg"),
      plot = correlation_plot,
      height = 4000, width = 6000, units = "px")

# order by tRNA - mRNA

samples_order <- samples %>% 
  arrange(factor(samples$kit,
                 levels = c("mRNA_illumina", "mRNA_qiagen_FSglobH", "mRNA_qiagen_FSglob", "tRNA_qiagen_FSglobH")),
          factor(samples$WBC_depletion,
                 levels = c("nofilter", "centpip", "plasmo", "pmacs", "cellulose")))

correlation_plot_ordered <- ggpairs(
  gene_tpm_filtered %>%
    filter(species == "pk") %>%
    filter(gene_biotype == "protein_coding") %>% 
    relocate(
      samples_order %>% select(sample) %>% pull()
      ),
  columns = 1:16,
  # columnLabels = samples_order %>% select(sample_short) %>% pull(),
  columnLabels = samples_order %>% select(sample_short) %>% mutate(sample_short = factor(sample_short, levels = sample_short)) %>% pull(),
  upper = list(continuous = wrap("cor", size = 5)),
) +
  theme(
    axis.text = element_text(size = 9),
    axis.title = element_text(size = 9),
    strip.text.x = element_text(size = 11),
    strip.text.y = element_text(size = 10),
    axis.text.x = element_text(size = 6),
    axis.text.y = element_text(size = 9)
    )
  # scale_y_discrete()

correlation_plot_ordered

ggsave(filename = here("results/figures/correlation_protein_coding_genes_pk_ordered.png"), 
      plot = correlation_plot_ordered,
      height = 4000, width = 6000, units = "px")
ggsave(filename = here("results/figures/correlation_protein_coding_genes_pk_ordered.svg"),
      plot = correlation_plot_ordered,
      height = 4000, width = 6000, units = "px")

```

All Pk genes:

```{r correlation-plot-log-export, message=FALSE, cache=TRUE, fig.height=18, out.width="100%"}
correlation_plot_log

ggsave(here("results/figures/correlation_protein_coding_genes_logscaled.png"), 
       plot = correlation_plot_log,
       height = 4000, width = 6000, units = "px")
ggsave(here("results/figures/correlation_protein_coding_genes_logscaled.svg"), 
       plot = correlation_plot_log,
       height = 4000, width = 6000, units = "px")

# order by tRNA-mRNA

correlation_plot_log_ordered <- ggpairs(
  gene_tpm_filtered %>%
    filter(species == "pk") %>%
    filter(gene_biotype == "protein_coding") %>% 
    relocate(
      samples_order %>% select(sample) %>% pull()
    ),
  columns = 1:16, # `104974-001-001_S1_L001`:`104974-001-002_S1_L001`,
  lower = list(continuous = "smooth"),
  upper = list(continuous = wrap("cor", size = 5)),
  columnLabels = samples_order %>% select(sample_short) %>% mutate(sample_short = factor(sample_short, levels = sample_short)) %>% pull(),,
) +
  theme(
    axis.text = element_text(size = 9),
    axis.title = element_text(size = 9),
    strip.text.x = element_text(size = 11),
    strip.text.y = element_text(size = 10),
    axis.text.x = element_text(size = 6),
    axis.text.y = element_text(size = 9)
    ) +
  scale_x_continuous(trans = "log1p") +
  scale_y_continuous(trans = "log1p")

correlation_plot_log

ggsave(here("results/figures/correlation_protein_coding_genes_logscaled_ordered.png"), 
       plot = correlation_plot_log_ordered,
       height = 4000, width = 6000, units = "px")
ggsave(here("results/figures/correlation_protein_coding_genes_logscaled_ordered.svg"), 
       plot = correlation_plot_log_ordered,
       height = 4000, width = 6000, units = "px")
```
