---
title: "WBC filtration RNASeq TPM"
author: "Eline Kattenberg"
date: "01/09/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library("dplyr")
library("tidyr")
library("ggplot2")
library("data.table")
library('tibble')
library("RColorBrewer")
library("ggfortify")
library("viridis")
library("GGally")

setwd("C:/Users/ekattenberg/Documents/Werk_LOCAL/WBC filtration")
```

## SAMPLE INFORMATION

loading the sample information table
```{r}
info<-read.csv("WBCfiltration_RNAseq_samplemap_stats.csv", header = TRUE)
```

## LOADING READS AND COUNTS FROM TPM TABLE
combine countdata for each sample from separate files

1. make the list of files

```{r}

ff <- list.files( path = "C:/Users/ekattenberg/Documents/Werk_LOCAL/WBC filtration/TPM_coordsorted/TPM_rawoutput", pattern = "*_TPMCalculator_genes.out", full.names = TRUE )
```

extra the data and collect as separate dataframes as lists in the object counts.files 

```{r}
counts.files <- lapply(ff, read.table, fill = TRUE, header = TRUE )
```

example of first count file for the 1st sample
```{r}
head(counts.files[[1]])
```
extract the reads column from all the counts files
```{r}
reads <- as.data.frame( sapply( counts.files, function(x) x[ ,6])) #6 for reads
```

extract the counts column from all the counts files
```{r}
TPMcounts <- as.data.frame( sapply( counts.files, function(x) x[ ,7])) #7 for TPM counts
```

extract the sample names
```{r}
ff <- gsub( "_TPMCalculator_genes[.]out", "", ff )
ff <- gsub( "C[:]/Users/ekattenberg/Documents/Werk_LOCAL/WBC filtration/TPM_coordsorted/TPM_rawoutput/104974-001-", "", ff )
```

load the gene info table
```{r}
geneID<-read.csv("gene_IDs_with_underscore.csv")
```

combine to make a count table with TPM counts
```{r}
colnames(TPMcounts) <- ff
row.names(TPMcounts) <- geneID$Gene_Id
colnames(TPMcounts)<- info$Libprep_filtration
```

combine to make a reads table
```{r}
colnames(reads) <- ff
row.names(reads) <- geneID$Gene_Id
colnames(reads)<- info$Libprep_filtration
```

Add table with gene info to count table
```{r}
DB_TPMcounts<-cbind(geneID, TPMcounts)
DB_reads<-cbind(geneID, reads)
```

transform tables from wide (one column for each sample) to long format (one column with a sample name variable, one column with all the counts/reads of all samples)

```{r}
DB_TPMcounts_long <- gather(DB_TPMcounts,key = "sample",  value = "TPMcounts", mRNA_qiagen_FSglob_nofilter:mRNA_illumina_cellulose, factor_key=TRUE)
head(DB_TPMcounts_long)

```

```{r}
DB_reads_long <- gather(DB_reads,key = "sample",  value = "reads", mRNA_qiagen_FSglob_nofilter:mRNA_illumina_cellulose, factor_key=TRUE)
head(DB_reads_long)
```

## Plot percentage of reads rRNA, globin and other by species for each library

Make a barplot of % reads by species and type of genes (rRNA, globin, other)
```{r}
r <- ggplot(DB_reads_long, aes(x= reads, y= sample, fill = species_type))
r + geom_bar(position="fill", stat="identity") + scale_fill_viridis(option = "cividis", discrete = T) 
```

Make new table first to get the percentages
```{r}
db_reads_by_sample_speciestype<-DB_reads_long %>%
  group_by(sample, species_type) %>%
  summarise(sum_reads = sum(reads, na.rm=TRUE))%>%
    mutate(per =  100 *sum_reads/sum(sum_reads))  
db_reads_by_sample_speciestype
```
```{r eval=FALSE, include=FALSE}
write.csv(db_reads_by_sample_speciestype, file = 'Reads_by_library_speciestype.csv')
```


```{r}
r <- ggplot(db_reads_by_sample_speciestype, aes(y= sum_reads, x= sample, fill = species_type))
r + geom_bar(position="fill", stat="identity") + scale_fill_viridis(option = "cividis", discrete = T) +coord_flip()
```

```{r}
r <- ggplot(db_reads_by_sample_speciestype, aes(y= sum_reads, x= sample, fill = species_type))
r + geom_bar(position="dodge", stat="identity") + scale_fill_viridis(option = "cividis", discrete = T) +coord_flip()
```

```{r}
r <- ggplot(db_reads_by_sample_speciestype, aes(y= sum_reads, x= sample, fill = species_type))
r + geom_bar(stat="identity") + scale_fill_viridis(option = "cividis", discrete = T) +coord_flip()
```


Make a barplot of % TPM counts by species and type of genes (rRNA, globin, other)
```{r}
t <- ggplot(DB_TPMcounts_long, aes(x= TPMcounts, y= sample, fill = species_type))
t + geom_bar(position="fill", stat="identity") + scale_fill_viridis(option = "magma", discrete = T) 
```

Make new table first to get the percentages
```{r}
db_counts_by_sample_speciestype<-DB_TPMcounts_long %>%
  group_by(sample, species_type) %>%
  summarise(sum_counts = sum(TPMcounts, na.rm=TRUE))%>%
    mutate(per =  100 *sum_counts/sum(sum_counts))  
db_counts_by_sample_speciestype
```

```{r eval=FALSE, include=FALSE}
write.csv(db_counts_by_sample_speciestype, file = 'Counts_by_library_speciestype.csv')
```

```{r}
r <- ggplot(db_counts_by_sample_speciestype, aes(y= sum_counts, x= sample, fill = species_type))
r + geom_bar(position="fill", stat="identity") + scale_fill_viridis(option = "magma", discrete = T) +coord_flip()
```

```{r}
r <- ggplot(db_counts_by_sample_speciestype, aes(y= sum_counts, x= sample, fill = species_type))
r + geom_bar(position="dodge", stat="identity") + scale_fill_viridis(option = "magma", discrete = T) +coord_flip()
```

```{r}
r <- ggplot(db_counts_by_sample_speciestype, aes(y= sum_counts, x= sample, fill = species_type))
r + geom_bar(stat="identity") + scale_fill_viridis(option = "magma", discrete = T) +coord_flip()
```


Make a barplot of amount of reads for all genes 
```{r}
db_reads_by_sample_species<-DB_reads_long %>%
  group_by(sample, Species) %>%
  summarise(sum_reads = sum(reads, na.rm=TRUE))%>%
    mutate(per =  100 *sum_reads/sum(sum_reads))  
db_reads_by_sample_species

r <- ggplot(db_reads_by_sample_species, aes(y= sum_reads, x= sample, fill = Species))
r + geom_bar(position="fill", stat="identity") + scale_fill_viridis(option = "cividis", discrete = T) +coord_flip()

t1 <- ggplot(db_reads_by_sample_species, aes(x= sum_reads, y= sample, fill = Species))
t1 + geom_bar(position="dodge", stat="identity" ) + scale_fill_viridis(option = "cividis", discrete = T)

t2 <- ggplot(db_reads_by_sample_species, aes(x= sum_reads, y= sample, fill = Species))
t2 + geom_bar(stat="identity" ) + scale_fill_viridis(option = "cividis", discrete = T)
```
old plot (incorrect? => no looks the same as 3rd one from previous chunk. You can extract the reads and percentages from the summary table there)
```{r}
t1 <- ggplot(DB_reads_long, aes(x= reads, y= sample,fill = Species))
t1 + geom_bar(stat="identity")
```


Make a barplot of amount of TPM counts for Pk specific genes 
```{r}
DB_Pk <- filter(DB_TPMcounts_long, species_type == "Pk")
t1 <- ggplot(DB_Pk, aes(x= TPMcounts, y= sample))
t1 + geom_bar(stat="identity")
```
 with log axis
```{r}
t1 <- ggplot(DB_Pk, aes(x= TPMcounts, y= sample))
t1 + geom_bar(stat="identity") + scale_x_log10()
```
Using the new table and to double check the axes
```{r}
db_PK_genecounts<-filter(db_counts_by_sample_speciestype, species_type == "Pk")
db_PK_genecounts

t1 <- ggplot(db_PK_genecounts, aes(x= sum_counts, y= sample))
t1 + geom_bar(position="dodge", stat="identity" ) + scale_fill_viridis(option = "viridis", discrete = T)

t2 <- ggplot(db_PK_genecounts, aes(x= sum_counts, y= sample))
t2 + geom_bar(stat="identity" ) + scale_fill_viridis(option = "viridis", discrete = T)
```





Make a barplot of amount of TPM counts for human specific genes 
```{r}
DB_Hum <- filter(DB_TPMcounts_long, species_type == "Hum")
t2 <- ggplot(DB_Hum, aes(x= TPMcounts, y= sample))
t2 + geom_bar(stat="identity")
```

Make a barplot of amount of TPM counts for Pk and human specific genes 
```{r}
DB_HumPK <- filter(DB_TPMcounts_long, (species_type == "Hum" | species_type == "Pk"))
t2 <- ggplot(DB_HumPK, aes(x= TPMcounts, y= sample, fill = species_type ))
t2 + geom_bar(position="dodge",stat="identity") +
    scale_fill_viridis(discrete = T) 
```

log scale
```{r}
t3 <- ggplot(DB_HumPK, aes(x= TPMcounts, y= sample, fill = species_type ))
t3 + geom_bar(position="dodge",stat="identity") +
    scale_fill_viridis(discrete = T)  + scale_x_log10()
```
all genes by making a database so you can extract numbers and percentages
```{r}
contractedHum_PK_genecounts<-DB_HumPK %>%
  group_by(sample, Species) %>%
  summarise(sum_counts = sum(TPMcounts, na.rm=TRUE))%>%
    mutate(per =  100 *sum_counts/sum(sum_counts))  
contractedHum_PK_genecounts

r <- ggplot(contractedHum_PK_genecounts, aes(y= sum_counts, x= sample, fill = Species))
r + geom_bar(position="fill", stat="identity") + scale_fill_viridis(option = "viridis", discrete = T) +coord_flip()

t1 <- ggplot(contractedHum_PK_genecounts, aes(x= sum_counts, y= sample, fill = Species))
t1 + geom_bar(position="dodge", stat="identity" ) + scale_fill_viridis(option = "viridis", discrete = T)

t2 <- ggplot(contractedHum_PK_genecounts, aes(x= sum_counts, y= sample, fill = Species))
t2 + geom_bar(stat="identity" ) + scale_fill_viridis(option = "viridis", discrete = T)
```
```{r eval=FALSE, include=FALSE}
write.csv(contractedHum_PK_genecounts, file = 'Counts_by_library_protein_coding_only.csv')
```

Make a barplot of amount of reads for Pk and human specific genes 
```{r}
DB_HumPKr <- filter(DB_reads_long, (species_type == "Hum" | species_type == "Pk"))
t2 <- ggplot(DB_HumPKr, aes(y= reads, x= sample, fill = species_type ))
t2 + geom_bar(position="dodge",stat="identity") +
    scale_fill_viridis(discrete = T) +
  coord_flip()
```
all genes by making a database so you can extract numbers and percentages
```{r}
contractedHum_PK_genereads<-DB_HumPKr %>%
  group_by(sample, Species) %>%
  summarise(sum_reads = sum(reads, na.rm=TRUE))%>%
    mutate(per =  100 *sum_reads/sum(sum_reads))  
contractedHum_PK_genereads

r <- ggplot(contractedHum_PK_genereads, aes(y= sum_reads, x= sample, fill = Species))
r + geom_bar(position="fill", stat="identity") + scale_fill_viridis(option = "viridis", discrete = T) +coord_flip()

t1 <- ggplot(contractedHum_PK_genereads, aes(x= sum_reads, y= sample, fill = Species))
t1 + geom_bar(position="dodge", stat="identity" ) + scale_fill_viridis(option = "viridis", discrete = T)

t2 <- ggplot(contractedHum_PK_genereads, aes(x= sum_reads, y= sample, fill = Species))
t2 + geom_bar(stat="identity" ) + scale_fill_viridis(option = "viridis", discrete = T)
```
```{r eval=FALSE, include=FALSE}
write.csv(contractedHum_PK_genereads, file = 'Reads_by_library_protein_coding_only.csv')
```


Make a barplot of amount of reads for Pk genes 
```{r}
DB_Pkr <- filter(DB_reads_long, species_type == "Pk")
t1r <- ggplot(DB_Pkr, aes(y= reads, x= sample))
t1r + geom_bar(stat="identity") +
  coord_flip()
```
Using the new table and to double check the axes
```{r}
db_PK_genereads<-filter(db_reads_by_sample_speciestype, species_type == "Pk")
db_PK_genereads

t2 <- ggplot(db_PK_genereads, aes(x= sum_reads, y= sample))
t2 + geom_bar(stat="identity" ) + scale_fill_viridis(option = "viridis", discrete = T)
```


Now focus more on the Pk reads

load the more specific gene type list
```{r}
genetypePK<-read.csv("Gene_function_list_Pk.csv", header = TRUE)
```

link it to the Pk (any type) DB
```{r}
DB_Pkall <- filter(DB_TPMcounts_long, Species == "Pk")
DB_Pk2<-left_join(DB_Pkall, genetypePK, by = "Gene_Id" )

```

```{r}
t <- ggplot(DB_Pk2, aes(x= TPMcounts, y= sample, fill = gene_type))
t + geom_bar(position="fill", stat="identity") + scale_fill_viridis(option = "magma", discrete = T) 
```

```{r}
DB_Pkallr <- filter(DB_reads_long, Species == "Pk")
DB_Pk2r<-left_join(DB_Pkallr, genetypePK, by = "Gene_Id" )
```


```{r}
t2 <- ggplot(DB_Pk2r, aes(x= reads, y= sample, fill = gene_type ))
t2 + geom_bar(position="dodge",stat="identity") +
    scale_fill_viridis(discrete = T) 
```
```{r}
t3 <- ggplot(DB_Pk2r, aes(x= reads, y= sample, fill = gene_type))
t3 + geom_bar(position="fill", stat="identity") + scale_fill_viridis(option = "magma", discrete = T) 
```

# Correlation plots

Now we need to use the wide dataset and make the between-sample correlation plots


First filter Pk protein coding genes only
```{r}
DB_Pk_wide <- filter(DB_TPMcounts, Species == "Pk")
DB_Pk_wide<-left_join(DB_Pk_wide, genetypePK, by = "Gene_Id" )
DB_Pk_wide2 <- filter(DB_Pk_wide, gene_type == "protein_coding_gene")
```

```{r}
#ggpairs(flea, columns = 2:4, ggplot2::aes(colour=species)) 
co_genes<-ggpairs(DB_Pk_wide2, columns =9:24, lower = list(
    continuous = "smooth"), upper = list(continuous = wrap("cor", size = 2))) 
co_genes + theme_bw() + scale_x_continuous(trans="log1p") + scale_y_continuous(trans="log1p")
```

```{r}
#ggpairs(flea, columns = 2:4, ggplot2::aes(colour=species)) 
co_all<-ggpairs(DB_Pk_wide, columns =9:24, ggplot2::aes(colour=gene_type), lower = list(
    continuous = "smooth"), upper = list(continuous = wrap("cor", size = 1))) 
co_all + theme_bw() + scale_x_continuous(trans="log1p") + scale_y_continuous(trans="log1p")
```

Select on the cellulose WBC depleted samples only (to make i)

```{r}
co_all_CELL<-ggpairs(DB_Pk_wide, columns =c(13,19,24), ggplot2::aes(colour=gene_type), lower = list(
    continuous = "smooth"), upper = list(continuous = wrap("cor", size = 2))) 
co_all_CELL + theme_bw() + scale_x_continuous(trans="log1p") + scale_y_continuous(trans="log1p")
```

protein coding genes only
```{r}
co_all_CELLg<-ggpairs(DB_Pk_wide2, columns =c(13,19,24), lower = list(
    continuous = "smooth")) 
co_all_CELLg + theme_bw() + scale_x_continuous(trans="log1p") + scale_y_continuous(trans="log1p")
```
## macs
```{r}
co_all_pmacs<-ggpairs(DB_Pk_wide, columns =c(12,18,23), ggplot2::aes(colour=gene_type), lower = list(
    continuous = "smooth")) 
co_all_pmacs + theme_bw() + scale_x_continuous(trans="log1p") + scale_y_continuous(trans="log1p")
```

```{r}
co_all_pmacsg<-ggpairs(DB_Pk_wide2, columns =c(12,18,23), lower = list(
    continuous = "smooth")) 
co_all_pmacsg + theme_bw() + scale_x_continuous(trans="log1p") + scale_y_continuous(trans="log1p")
```

## plasmodipur
```{r}
co_all_plasmo<-ggpairs(DB_Pk_wide, columns =c(11,17,22), ggplot2::aes(colour=gene_type), lower = list(
    continuous = "smooth")) 
co_all_plasmo + theme_bw() + scale_x_continuous(trans="log1p") + scale_y_continuous(trans="log1p")
```


```{r}
co_all_plasmog<-ggpairs(DB_Pk_wide2, columns =c(11,17,22), lower = list(
    continuous = "smooth")) 
co_all_plasmog + theme_bw() + scale_x_continuous(trans="log1p") + scale_y_continuous(trans="log1p")
```

## centpip
```{r}
co_all_centpip<-ggpairs(DB_Pk_wide, columns =c(10,16,21), ggplot2::aes(colour=gene_type), lower = list(
    continuous = "smooth")) 
co_all_centpip + theme_bw() + scale_x_continuous(trans="log1p") + scale_y_continuous(trans="log1p")
```

```{r}
co_all_centpipg<-ggpairs(DB_Pk_wide2, columns =c(10,16,21), lower = list(
    continuous = "smooth")) 
co_all_centpipg + theme_bw() + scale_x_continuous(trans="log1p") + scale_y_continuous(trans="log1p")
```

## no filter
```{r}
co_all_nofilt<-ggpairs(DB_Pk_wide, columns =c(9,14,15,20), ggplot2::aes(colour=gene_type), lower = list(
    continuous = "smooth")) 
co_all_nofilt + theme_bw() + scale_x_continuous(trans="log1p") + scale_y_continuous(trans="log1p")
```

```{r}
co_all_nofiltg<-ggpairs(DB_Pk_wide2, columns =c(9,14,15,20), lower = list(
    continuous = "smooth")) 
co_all_nofiltg + theme_bw() + scale_x_continuous(trans="log1p") + scale_y_continuous(trans="log1p")
```

# correlation plots by library type

## mRNA qiagen
```{r}
co_all_mRNAqiagen<-ggpairs(DB_Pk_wide, columns =9:14, ggplot2::aes(colour=gene_type), lower = list(
    continuous = "smooth"), upper = list(continuous = wrap("cor", size = 2))) 
co_all_mRNAqiagen + theme_bw() + scale_x_continuous(trans="log1p") + scale_y_continuous(trans="log1p")
```

```{r}
co_gen_mRNAqiagen<-ggpairs(DB_Pk_wide2, columns =9:14, lower = list(
    continuous = "smooth")) 
co_gen_mRNAqiagen + theme_bw() + scale_x_continuous(trans="log1p") + scale_y_continuous(trans="log1p")
```


## tRNA qiagen
```{r}
co_all_tRNAqiagen<-ggpairs(DB_Pk_wide, columns =15:19, ggplot2::aes(colour=gene_type), lower = list(
    continuous = "smooth"), upper = list(continuous = wrap("cor", size = 2))) 
co_all_tRNAqiagen + theme_bw() + scale_x_continuous(trans="log1p") + scale_y_continuous(trans="log1p")
```

```{r}
co_gen_tRNAqiagen<-ggpairs(DB_Pk_wide2, columns =15:19, lower = list(
    continuous = "smooth")) 
co_gen_tRNAqiagen + theme_bw() + scale_x_continuous(trans="log1p") + scale_y_continuous(trans="log1p")
```

## mRNA illumina
```{r}
co_all_mRNAillum<-ggpairs(DB_Pk_wide, columns =20:24, ggplot2::aes(colour=gene_type), lower = list(
    continuous = "smooth"), upper = list(continuous = wrap("cor", size = 2))) 
co_all_mRNAillum + theme_bw() + scale_x_continuous(trans="log1p") + scale_y_continuous(trans="log1p")
```

```{r}
co_gen_mRNAillum<-ggpairs(DB_Pk_wide2, columns =20:24, lower = list(
    continuous = "smooth")) 
co_gen_mRNAillum + theme_bw() + scale_x_continuous(trans="log1p") + scale_y_continuous(trans="log1p")
```

## All samples but change the order per sample
```{r}
co_genes2<-ggpairs(DB_Pk_wide2, columns = c(9,14,15,20,10,16,21,11,17,22,12,18,23,13,19,24), lower = list(
    continuous = "smooth"), upper = list(continuous = wrap("cor", size = 2))) 
co_genes2 + theme_bw() + scale_x_continuous(trans="log1p") + scale_y_continuous(trans="log1p") + theme(strip.text = element_text(size = 5))
```

```{r}
co_all2<-ggpairs(DB_Pk_wide, columns = c(9,14,15,20,10,16,21,11,17,22,12,18,23,13,19,24), ggplot2::aes(colour=gene_type), lower = list(
    continuous = "smooth"), upper = list(continuous = wrap("cor", size = 2))) 
co_all2 + theme_bw() + scale_x_continuous(trans="log1p") + scale_y_continuous(trans="log1p") + theme(strip.text = element_text(size = 5))
```


